---
title: Genome-wide analysis of heat shock response 
author: 
  - name: Peter Humburg
    affiliation: 1
address:
  - code: 1
    address: Wellcome Trust Centre for Human Genetics, University of Oxford, Roosevelt Dr., Oxford, OX3 7BN, UK
date: `r format(Sys.time(), "%a %d %b %Y")`
---

```{r, setup, include=FALSE}
library(gdata)
library(affy)
library(vsn)
library(limma)
library(scatterplot3d)
library(sparcl)
library(sva)
library(ggplot2)
library(dplyr)
library(tidyr)
library(readr)
library(illuminaHumanv3.db)
library(knitr)
library(pander)

opts_chunk$set(autodep=TRUE)
opts_chunk$set(tidy=TRUE)
opts_chunk$set(echo=FALSE)
opts_chunk$set(cache=TRUE)
panderOptions("digits", 3)
panderOptions("table.split.table", 160)
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")
options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")
```


```{r, functions, include=FALSE}
pcaPlot <- function(data, col, pch, ...){
        pca <- prcomp(t(data), scale=TRUE)
        scores <- as.data.frame(pca$x)
        plot(scores$PC1, scores$PC2, col=col, pch=pch, ...)
}

reshapeData <- function(data, value){
	ans <- reshape2::melt(data, id.vars="probe",
		value.name=value, variable.name="sample")
	splitLabel <- strsplit(as.character(ans$sample), "_", fixed=TRUE)
	ans$sample <- sapply(splitLabel, "[", 1)
	ans$treatment <- sapply(splitLabel, function(x) x[length(x)])
	as.tbl(ans[c("sample", "treatment", "probe", value)])
}

ibsClust <- function(genome, ...){
	dst <- spread(dplyr::select(genome, IID1, IID2, DST), IID2, DST)
	size <- nrow(dst) + 1
	labels <- c(as.character(dst$IID1[1]), names(dst)[-1])
	dst <- t(1 - as.matrix(dst[-1]))
	dst <- dst[!is.na(dst)]
	attr(dst, "Size") <- size
	attr(dst, "Labels") <- labels
	attr(dst, "Diag") <- FALSE
	attr(dst, "Upper") <- FALSE
	class(dst) <- "dist"
	
	hclust(dst, ...)
}

figRef <- local({
    tag <- numeric()
    created <- logical()
    used <- logical()
    function(label, caption, prefix = options("figcap.prefix"), 
        sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight")) {
        i <- which(names(tag) == label)
        if (length(i) == 0) {
            i <- length(tag) + 1
            tag <<- c(tag, i)
            names(tag)[length(tag)] <<- label
            used <<- c(used, FALSE)
            names(used)[length(used)] <<- label
            created <<- c(created, FALSE)
            names(created)[length(created)] <<- label
        }
        if (!missing(caption)) {
            created[label] <<- TRUE
            paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
                " ", caption)
        } else {
            used[label] <<- TRUE
            paste(prefix, tag[label])
        }
    }
})

tabRef <- local({
    tag <- numeric()
    created <- logical()
    used <- logical()
    function(label, caption, prefix = options("tabcap.prefix"), 
        sep = options("tabcap.sep"), prefix.highlight = options("tabcap.prefix.highlight")) {
        i <- which(names(tag) == label)
        if (length(i) == 0) {
            i <- length(tag) + 1
            tag <<- c(tag, i)
            names(tag)[length(tag)] <<- label
            used <<- c(used, FALSE)
            names(used)[length(used)] <<- label
            created <<- c(created, FALSE)
            names(created)[length(created)] <<- label
        }
        if (!missing(caption)) {
            created[label] <<- TRUE
            paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
                " ", caption)
        } else {
            used[label] <<- TRUE
            paste(prefix, tag[label])
        }
    }
})
``` 
```{r, formats, include=FALSE}
formats <- read.dcf(file="default.pandoc", fields="t")[,1]
formats <- formats[!is.na(formats) & !grepl("html", formats)]
latexIdx <- which(formats == "latex")
if(length(latexIdx)){
	formats[latexIdx] <- "pdf"
}
url <- paste("heatshock_analysis", formats, sep=".")
```

<aside class="download">

# Downloads
## Report
`r paste0("<span title=\"Download report as ", formats, "\">[", formats, "](", url, ")</span>", collapse=", ")`

## Log files

```{r linkLogs, cache=FALSE, results="asis"}
logFiles <- sapply(logs, linkLog, "markdown")
desc <- sapply(logs, "[[", "desc")
cat(paste0("<span title=\"", desc, "\">", logFiles, "</span>", collapse=",\n"))
```
</aside>


# Introduction
The heat shock response is a highly conserved mechanism responsible for the 
maintenance of cellular function under stress. Originally identified as part of the 
response to increased environmental temperatures [@Ritossa62] heat shock proteins (HSPs) 
have since been implicated  in susceptability to infection [@Kee08], cancer risk [@Sfar10] 
and drug response [@Martin04]. Despite its obvious significance many details of the
heat shock response still remain uncertain.

Here we study the genome-wide changes to gene expression induced by heat shock in
60 HapMap cell lines from Yoruban individuals [@Gibbs10] to identify genes and pathways
involved in the human heat shock response. To further elucidate underlying mechanisms
genetic variants modulating gene expression are mapped across all relevant genes.

# Material and Methods

## Quality Control

```{r, loadExpression}
rawIntensity <- readr::read_tsv("data/TPS00590_FinalReport_12Oct09.txt", skip=8, progress=FALSE)

sig.col<-seq(3, 722, 6)
det.col<-seq(8, 722, 6)

sig.data.raw <- rawIntensity[,sig.col]
det.data.raw <- rawIntensity[,det.col]

names(sig.data.raw) <- gsub(":AVG_Signal", "", names(sig.data.raw))
names(det.data.raw) <- gsub(":Detection Pval", "", names(det.data.raw))

info<-readr::read_tsv("data/YRI_sample_list_file_03NOV09.txt", 
		col_types=list("Date of Harvest"=col_date("%d.%m.%y"), 
				"Date of RNA extraction"=col_date("%d.%m.%y")))

## remove empty lines at end of file
info<-info[1:120,] 
factors <- which(sapply(info, is.factor))
for(i in factors){
	info[[i]] <- factor(as.character(info[[i]]))
}


newNames <- paste(info[["Cell Line"]], info$Treatment, sep="_")
names(newNames)<- info[["ARRAY CODE"]]

## typo in sample ID
names(sig.data.raw)[13]<- "NM-75"
names(det.data.raw)[13]<- "NM-75"

names(sig.data.raw) <-gsub("-", "_", fixed=TRUE, names(sig.data.raw))
names(det.data.raw) <-gsub("-", "_", fixed=TRUE, names(det.data.raw)) #differing name format
names(sig.data.raw) <- newNames[names(sig.data.raw)]
names(det.data.raw) <- newNames[names(det.data.raw)]

sig.data.raw <- cbind(probe=rawIntensity$ProbeID, sig.data.raw)
det.data.raw <- cbind(probe=rawIntensity$ProbeID, det.data.raw)

sig.data.log<- cbind(probe=rawIntensity$ProbeID, log2(sig.data.raw[-1]))
```
Probe intensities for resting and stimulated cells were imported into R for further
processing together with associated meta data.

```{r excludeSetup}
exclude <- list()
exclude$sample <- character()
exclude$probe <- character()
```

```{r longData}
sig.data.long <- reshapeData(sig.data.log, "intensity")
det.data.long <- reshapeData(det.data.raw, "p.value")
```

Annotations for all probes were obtained via the *illuminaHumanv3.db* Bioconductor
package [@Barbosa10]. Only probes that are considered to are of perfect or good
quality were considered for analysis. Additionally, all probes mapping to more than
one genomic location or to a location that contains a known SNP were excluded.

```{r probeQual}
annot <- illuminaHumanv3fullReannotation()
annot <- subset(illuminaHumanv3fullReannotation(), 
		!grepl("Good|Perfect", ProbeQuality) | !is.na(OverlappingSNP) | 
		!is.na(SecondMatches) | !is.na(OtherGenomicMatches) | !is.na(RepeatMask))
exclude$probe <- union(exclude$probe, annot$ArrayAddress)
```


Probes were required to exhibit significant signal (detection p-value < 0.05) 
in at least seven samples and samples with less than 30% of the remaining probes 
providing significant signal were excluded (together with the paired sample).
Samples showing exceptionally low variation in probe intensities (standard deviation
of the log intensities of all retained probes below 0.7) were also removed.

```{r lowDetProbe}
detected <- tally(group_by(subset(det.data.long, p.value < 0.05), probe))
exclude$probe <- union(exclude$probe, filter(detected, n < 7)$probe)
```

```{r filterProbe}
sig.data.long <- filter(sig.data.long, !probe %in% exclude$probe)
det.data.long <- filter(det.data.long, !probe %in% exclude$probe)
```

```{r lowDetSample}
missingSample <- tally(group_by(filter(det.data.long, p.value > 0.05), sample, treatment))
missingSample$n <- missingSample$n/count(det.data.long, sample, treatment)$n
exclude$sample <- union(exclude$sample, filter(missingSample, n >= 0.7)$sample)
```

```{r lowSD}
sampleSD <- summarise(group_by(filter(sig.data.long, is.finite(intensity)), 
				sample, treatment), sd(intensity))
exclude$sample <- union(exclude$sample, filter(sampleSD, `sd(intensity)` < 0.7)$sample)
```

```{r filterSample}
sig.data.long <- filter(sig.data.long, !sample %in% exclude$sample)
det.data.long <- filter(det.data.long, !sample %in% exclude$sample)
```

After filtering `r length(unique(sig.data.long$probe))` of `r nrow(sig.data.raw)`
probes (`r format(length(unique(sig.data.long$probe))/nrow(sig.data.raw)*100, digits=2, nsmall=1)`%)
and `r length(unique(sig.data.long$sample))` samples remain. See `r figRef("densityPlot")`
for the distributions of intensities for the remaining probes in each sample. 

```{r densityPlot, cache=FALSE, fig.cap=figRef("densityPlot", "Distribution of probe intensities.")}
ggplot(sig.data.long, aes(x=intensity, colour=sample, linetype=treatment)) + 
		geom_density() + theme_bw() + guides(colour="none")
```

```{r pcaPrepare}
flt <- filter(sig.data.long, !is.finite(intensity))$probe
sig.data.filter <- full_join(spread(filter(sig.data.long, 
					treatment == "Basal" & !probe %in% flt), 
					sample, intensity)[-1],
			spread(filter(sig.data.long, treatment == "Stim" & !probe %in% flt), 
					sample, intensity)[-1],
			by="probe")
names(sig.data.filter) <- sub(".x", "_Basal", names(sig.data.filter), fixed=TRUE)
names(sig.data.filter) <- sub(".y", "_Stim", names(sig.data.filter), fixed=TRUE)

sig.pca <- prcomp(t(sig.data.filter[-1]), scale=TRUE)
sig.pc <- as.data.frame(sig.pca$x)
sig.pc$sample <- rownames(sig.pc)
sig.pc <- dplyr::select(as.tbl(sig.pc), sample, PC1, PC2)
info <- mutate(info, sample=paste(`Cell Line`, `Treatment`, sep="_"))
sig.pc <- inner_join(sig.pc, info, by="sample")

cluster <- dplyr::filter(sig.pc, `Date cRNA Amplified` != "CE_29Sep09")
clusterBound <- chull(cluster$PC1, cluster$PC2) 
```

```{r pcaPlot, cache=FALSE, fig.cap=figRef("pcaPlot", "PCA plot of samples. Colours correspond to different BeadChips while shapes indicate different amplification dates. Samples from the two amplifications in October 2009 are clustered together (highlighted by the grey area).")}
ggplot(sig.pc, aes(x=PC1, y=PC2)) + 
     geom_polygon(data=cluster[clusterBound, ], fill="lightgrey", colour="grey") + 
     geom_point(aes( colour=as.character(Sentrix_ID), shape=`Date cRNA Amplified`), size=3) + 
     theme_bw() + scale_colour_discrete("BeadChip") + scale_shape_discrete("RNA Amplification")
```

A principle component analysis of the gene expression for the remaining samples indicates 
a clustering of samples from two RNA amplification batches in October 2009
(`r figRef("pcaPlot")`). This batch effect is noted for later correction during the analysis.

```{r sampleList, cache=FALSE}
samples <- sub("_.", "", sub("GM", "NA", unique(sig.data.long$sample)))
sampleInfo <- readr::read_delim("data/hapmap_YRI_r23a_filtered.fam", delim=" ", 
		col_names=c("Family", "Individual", "Father", "Mother", "Sex", "Phenotype"))
keepSamples <- dplyr::filter(dplyr::select(sampleInfo, Family, Individual), Individual %in% samples)
write.table(keepSamples, file="tmp/sample.lst", quote=FALSE, col.names=FALSE, row.names=FALSE)
```

```{r plinkSNPs, engine="bash", cache=FALSE}
plink --bfile data/hapmap_YRI_r23a_filtered --out tmp/hapmap_yri.snp_flt --keep tmp/sample.lst --maf 0.1 --geno 0.1 --hwe 1e-4 --autosome --make-bed > log/hapmap_yri.snp_flt.log
```

```{r snpFilter, cache.extra=list(file.info("data/hapmap_YRI_r23a_filtered.bim")$mtime, file.info("tmp/hapmap_yri.bim")$mtime)}
totalSNPs <- R.utils::countLines("data/hapmap_YRI_r23a_filtered.bim")
remainSNPs <- R.utils::countLines("tmp/hapmap_yri.snp_flt.bim")
```

Genotype data provided by the HapMap project [@Gibbs10] was processed with Plink [@Chang15]
to restrict the data to autosomes and remove SNPs with low genotyping rate and 
and those with a minor allele frequency of less than 10% in the sample. This results
in the exclusion of `r totalSNPs - remainSNPs` of 
`r totalSNPs` SNPs (`r format((totalSNPs - remainSNPs)/totalSNPs*100, nsmall=2, digits=2)`%);

```{r plinkSample, engine="bash", cache=FALSE}
plink --bfile tmp/hapmap_yri.snp_flt --out tmp/hapmap_yri.smpl_flt --mind 0.1 --neighbour 1 3 --genome gz nudge --cluster --mds-plot 4 --make-bed > log/hapmap_yri.smpl_flt.log
```

```{r nnIBS, cache=FALSE}
nnIBS <- readr::read_table("tmp/hapmap_yri.smpl_flt.nearest")
```

The IBS nearest neighbour caculation shows that the majority of samples are quite
similar with regard to the extend of genetic differences between them while a few
show markedly increased sharing of alleles (see `r figRef("nnIBS")` and `r tabRef("nnIBS")`). 
This is consistent with a population of largely unrelated individuals including a few 
individuals that are more closely related than expected. One individual (NA18913) also shows
evidence of reduced similarity with the second and third nearest neighbour.

```{r nnIBSplot, cache=FALSE, fig.cap=figRef("nnIBS", "Similarity between individuals. For each individual the similarity, in terms of IBS, to all other individuals is determined and individuals are ranked acording to this measure. From this a Z-score is computed for for each of the three closest neighbours (see the Plink [documentation](http://pngu.mgh.harvard.edu/~purcell/plink/strat.shtml#outlier) for details). Note the spikes corresponding to individuals with positive Z-scores indicating increased similarity of their genomes compared to the rest of the cohort.")}
ggplot(nnIBS, aes(x=Z, colour=as.factor(NN))) + geom_density(aes(group=NN)) + theme_bw()
```

```{r nnIBStable, cache=FALSE}
offset <- mean(filter(nnIBS, Z < 0 & NN == 1)$Z)
nnSuspect <- filter(nnIBS, (abs(Z - offset) > 2 & NN == 1) | abs(Z) > 2)
nnSuspect <- nnSuspect[order(abs(nnSuspect$Z), decreasing=TRUE), ]
set.caption(tabRef("nnIBS", 
		"Pairs of Yoruban HapMap samples with evidence of decreased or increased genetic similarity (absolute IBS-based nearest neighbour Z-score > 2)."))
pander(nnSuspect)
```

```{r loadIBD}
ibdReport <- read.table("tmp/hapmap_yri.smpl_flt.genome.gz", header=TRUE)
ibdSD <- sd(ibdReport$PI_HAT)
ibdHigh <- dplyr::filter(ibdReport, PI_HAT > 2*ibdSD)
```

A similar picture emerges if we consider the estimated proportion of IBD for all
sample pairs with `r nrow(ibdHigh)` pairs showing evidence of higher than expected
relatedness (see `r tabRef("ibdHigh")` and `r figRef("ibsDendro")` for details).

```{r selectExlusions}
pairs <- rbind(as.matrix(dplyr::select(ibdHigh, IID1, IID2)), 
		as.matrix(dplyr::select(nnSuspect, IID, IID2)))
exSample <- character()
while(length(pairs)){
	ids <- sort(table(pairs), decreasing=TRUE)
	exSample <- c(exSample, names(ids[1]))
	pairs <- pairs[!apply(pairs, 1, `%in%`, x=names(ids)[1]), , drop=FALSE]
}
exclude$sample <- union(exclude$sample, exSample)
```

To ensure that at most one sample of each suspect pair are included in the analysis
the samples `r pander(exSample)` are excluded.

```{r ibdHighTab}
set.caption(tabRef("ibdHigh", paste("Sample pairs with estimated proportion of IBD exceeding",
		format(2*ibdSD, nsmall=2, digits=2))))
pander(dplyr::select(ibdHigh, IID1, IID2, Z0, Z1, Z2, PI_HAT, DST, RATIO))
```

```{r ibsDendroPlot, cache=FALSE, fig.cap=figRef("ibsDendro", "Dendrogram of individuals included in study. Distances are based on identity by state. Three pairs show clear indications of increased relatedness.")}
plot(ibsClust(ibdReport), main="", xlab="IBS", sub="")
```

# Results

# Appendix {-}
## Custom functions used
```{r, ref.label="functions", eval=FALSE, echo=TRUE}
```

## Session Info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References {-}
