---
title: Genome-wide analysis of heat shock response 
author: 
  - name: Peter Humburg
    affiliation: 1
address:
  - code: 1
    address: Wellcome Trust Centre for Human Genetics, University of Oxford, Roosevelt Dr., Oxford, OX3 7BN, UK
date: `r format(Sys.time(), "%a %d %b %Y")`
---

```{r, setup, include=FALSE}
library(gdata)
library(affy)
library(vsn)
library(limma)
library(scatterplot3d)
library(sparcl)
library(sva)
library(ggplot2)
library(dplyr)

opts_chunk$set(autodep=TRUE)
opts_chunk$set(tidy=TRUE)
opts_chunk$set(echo=FALSE)
opts_chunk$set(cache=TRUE)
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")
options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")
```


```{r, functions, include=FALSE}
pcaPlot <- function(data, col, pch, ...){
        pca <- prcomp(t(data))
        scores <- as.data.frame(pca$x)
        plot(scores$PC1, scores$PC2, col=col, pch=pch, ...)
}

figRef <- local({
    tag <- numeric()
    created <- logical()
    used <- logical()
    function(label, caption, prefix = options("figcap.prefix"), 
        sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight")) {
        i <- which(names(tag) == label)
        if (length(i) == 0) {
            i <- length(tag) + 1
            tag <<- c(tag, i)
            names(tag)[length(tag)] <<- label
            used <<- c(used, FALSE)
            names(used)[length(used)] <<- label
            created <<- c(created, FALSE)
            names(created)[length(created)] <<- label
        }
        if (!missing(caption)) {
            created[label] <<- TRUE
            paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
                " ", caption)
        } else {
            used[label] <<- TRUE
            paste(prefix, tag[label])
        }
    }
})

tabRef <- local({
    tag <- numeric()
    created <- logical()
    used <- logical()
    function(label, caption, prefix = options("tabcap.prefix"), 
        sep = options("tabcap.sep"), prefix.highlight = options("tabcap.prefix.highlight")) {
        i <- which(names(tag) == label)
        if (length(i) == 0) {
            i <- length(tag) + 1
            tag <<- c(tag, i)
            names(tag)[length(tag)] <<- label
            used <<- c(used, FALSE)
            names(used)[length(used)] <<- label
            created <<- c(created, FALSE)
            names(created)[length(created)] <<- label
        }
        if (!missing(caption)) {
            created[label] <<- TRUE
            paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
                " ", caption)
        } else {
            used[label] <<- TRUE
            paste(prefix, tag[label])
        }
    }
})
``` 
```{r, formats, include=FALSE}
formats <- read.dcf(file="default.pandoc", fields="t")[,1]
formats <- formats[!is.na(formats) & !grepl("html", formats)]
latexIdx <- which(formats == "latex")
if(length(latexIdx)){
	formats[latexIdx] <- "pdf"
}
url <- paste("heatshock_analysis", formats, sep=".")
```

<div id="download">
<p>
`r paste0("[", formats, "](", url, ")", collapse=", ")`
</p>
</div>

# Introduction
The heat shock response is a highly conserved mechanism responsible for the 
maintenance of cellular function under stress. Originally identified as part of the 
response to increased environmental temperatures [@Ritossa62] heat shock proteins (HSPs) 
have since been implicated  in susceptibility to infection [@Kee08], cancer risk [@Sfar10] 
and drug response [@Martin04]. Despite its obvious significance many details of the
heat shock response still remain uncertain.

Here we study the genome-wide changes to gene expression induced by heat shock in
60 HapMap cell lines from Yoruban individuals [@Gibbs10] to identify genes and pathways
involved in the human heat shock response. To further elucidate underlying mechanisms
genetic variants modulating gene expression are mapped across all relevant genes.

# Material and Methods

## Quality Control

```{r, loadExpression}
## probe intensities
rawIntensity <- read.delim("data/TPS00590_FinalReport_12Oct09.txt", skip=8)

sig.col<-seq(3, 722, 6)
det.col<-seq(8, 722, 6)

sig.data.raw <- rawIntensity[,sig.col]
det.data.raw <- rawIntensity[,det.col]

row.names(sig.data.raw) <- rawIntensity$ProbeID
row.names(det.data.raw) <- rawIntensity$ProbeID

colnames(sig.data.raw) <- gsub(".AVG_Signal", "", colnames(sig.data.raw))
colnames(det.data.raw) <- gsub(".Detection.Pval", "", colnames(det.data.raw))

## sample information
n<-read.delim("data/YRI_sample_list_file_03NOV09.txt")

## remove empty lines at end of file
n<-n[1:120,] 
factors <- which(sapply(n, is.factor))
for(i in factors){
	n[[i]] <- factor(as.character(n[[i]]))
}


newNames <- paste(n[["Cell.Line"]], n$Treatment, sep="_")
names(newNames)<- n$ARRAY.CODE

## typo in sample ID
names(sig.data.raw)[13]<- "NM_75"
names(det.data.raw)[13]<- "NM_75"

names(sig.data.raw) <-gsub(".", "_", fixed=TRUE, names(sig.data.raw))
names(det.data.raw) <-gsub(".", "_", fixed=TRUE, names(det.data.raw)) #differing name format
names(sig.data.raw) <- newNames[names(sig.data.raw)]
names(det.data.raw) <- newNames[names(det.data.raw)]

sig.data.log<-log2(sig.data.raw)
```
Probe intensities for resting and stimulated cells were imported into R for further
processing.

```{r excludeSetup}
exclude <- list()
exclude$sample <- character()
exclude$probe <- character()
```

```{r longData}
sig.data.long <- reshape2::melt(sig.data.log, value.name="intensity")
splitLabel <- strsplit(as.character(sig.data.long$variable), "_")
sig.data.long$variable <- sapply(splitLabel, function(x) paste(x[-length(x)], collapse="_"))
sig.data.long$treatment <- sapply(splitLabel, function(x) x[length(x)])
sig.data.long$probe <- rep(rownames(sig.data.log), ncol(sig.data.log))
names(sig.data.long)[1] <- "sample"
sig.data.long <- as.tbl(sig.data.long[c("sample", "treatment", "probe", "intensity")])

det.data.long <- reshape2::melt(det.data.raw, value.name="p.value")
splitLabel <- strsplit(as.character(det.data.long$variable), "_")
det.data.long$variable <- sapply(splitLabel, function(x) paste(x[-length(x)], collapse="_"))
det.data.long$treatment <- sapply(splitLabel, function(x) x[length(x)])
det.data.long$probe <- rep(rownames(det.data.raw), ncol(det.data.raw))
names(det.data.long)[1] <- "sample"
det.data.long <- as.tbl(det.data.long[c("sample", "treatment", "probe", "p.value")])
```

Probes were required to exhibit significant signal (detection p-value < 0.05) 
in at least seven samples and samples with less than 20% of probes providing
significant signal were excluded (together with the paired sample).

```{r lowDetProbe}
detected <- tally(group_by(subset(det.data.long, p.value < 0.05), probe))
exclude$probe <- union(exclude$probe, filter(detected, n < 7)$probe)
sig.data.long <- filter(sig.data.long, !probe %in% exclude$probe)
det.data.long <- filter(det.data.long, !probe %in% exclude$probe)
```

```{r lowDetSample}
detectedSample <- tally(group_by(subset(det.data.long, p.value < 0.05), sample))
detectedSample$n <- detectedSample$n/nrow(det.data.raw)
exclude$sample <- union(exclude$sample, filter(detectedSample, n < 0.25)$sample)
sig.data.long <- filter(sig.data.long, !sample %in% exclude$sample)
det.data.long <- filter(det.data.long, !sample %in% exclude$sample)
```



# Results
## Quality Control


```{r densityPlot, fig.cap=figRef("densityPlot", "Distribution of probe intensities.")}
ggplot(sig.data.long, aes(x=intensity, colour=sample, linetype=treatment)) + geom_density() + theme_bw()
```


# Appendix {-}
## Custom functions used
```{r, ref.label="functions", eval=FALSE, echo=TRUE}
```

## Session Info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References {-}
