---
title: Genome-wide analysis of heat shock response 
author: 
  - name: Peter Humburg
    affiliation: 1
address:
  - code: 1
    address: Wellcome Trust Centre for Human Genetics, University of Oxford, Roosevelt Dr., Oxford, OX3 7BN, UK
date: `r format(Sys.time(), "%a %d %b %Y")`
---

```{r, setup, include=FALSE}
library(gdata)
library(affy)
library(vsn)
library(limma)
library(scatterplot3d)
library(sparcl)
library(sva)
library(ggplot2)
library(dplyr)
library(tidyr)
library(readr)
library(illuminaHumanv3.db)
library(knitr)

opts_chunk$set(autodep=TRUE)
opts_chunk$set(tidy=TRUE)
opts_chunk$set(echo=FALSE)
opts_chunk$set(cache=TRUE)
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")
options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")
```


```{r, functions, include=FALSE}
pcaPlot <- function(data, col, pch, ...){
        pca <- prcomp(t(data), scale=TRUE)
        scores <- as.data.frame(pca$x)
        plot(scores$PC1, scores$PC2, col=col, pch=pch, ...)
}

figRef <- local({
    tag <- numeric()
    created <- logical()
    used <- logical()
    function(label, caption, prefix = options("figcap.prefix"), 
        sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight")) {
        i <- which(names(tag) == label)
        if (length(i) == 0) {
            i <- length(tag) + 1
            tag <<- c(tag, i)
            names(tag)[length(tag)] <<- label
            used <<- c(used, FALSE)
            names(used)[length(used)] <<- label
            created <<- c(created, FALSE)
            names(created)[length(created)] <<- label
        }
        if (!missing(caption)) {
            created[label] <<- TRUE
            paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
                " ", caption)
        } else {
            used[label] <<- TRUE
            paste(prefix, tag[label])
        }
    }
})

tabRef <- local({
    tag <- numeric()
    created <- logical()
    used <- logical()
    function(label, caption, prefix = options("tabcap.prefix"), 
        sep = options("tabcap.sep"), prefix.highlight = options("tabcap.prefix.highlight")) {
        i <- which(names(tag) == label)
        if (length(i) == 0) {
            i <- length(tag) + 1
            tag <<- c(tag, i)
            names(tag)[length(tag)] <<- label
            used <<- c(used, FALSE)
            names(used)[length(used)] <<- label
            created <<- c(created, FALSE)
            names(created)[length(created)] <<- label
        }
        if (!missing(caption)) {
            created[label] <<- TRUE
            paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
                " ", caption)
        } else {
            used[label] <<- TRUE
            paste(prefix, tag[label])
        }
    }
})
``` 
```{r, formats, include=FALSE}
formats <- read.dcf(file="default.pandoc", fields="t")[,1]
formats <- formats[!is.na(formats) & !grepl("html", formats)]
latexIdx <- which(formats == "latex")
if(length(latexIdx)){
	formats[latexIdx] <- "pdf"
}
url <- paste("heatshock_analysis", formats, sep=".")
```

<div id="download">
<p>
`r paste0("[", formats, "](", url, ")", collapse=", ")`
</p>
</div>

# Introduction
The heat shock response is a highly conserved mechanism responsible for the 
maintenance of cellular function under stress. Originally identified as part of the 
response to increased environmental temperatures [@Ritossa62] heat shock proteins (HSPs) 
have since been implicated  in susceptibility to infection [@Kee08], cancer risk [@Sfar10] 
and drug response [@Martin04]. Despite its obvious significance many details of the
heat shock response still remain uncertain.

Here we study the genome-wide changes to gene expression induced by heat shock in
60 HapMap cell lines from Yoruban individuals [@Gibbs10] to identify genes and pathways
involved in the human heat shock response. To further elucidate underlying mechanisms
genetic variants modulating gene expression are mapped across all relevant genes.

# Material and Methods

## Quality Control

```{r, loadExpression}
rawIntensity <- readr::read_tsv("data/TPS00590_FinalReport_12Oct09.txt", skip=8, progress=FALSE)

sig.col<-seq(3, 722, 6)
det.col<-seq(8, 722, 6)

sig.data.raw <- rawIntensity[,sig.col]
det.data.raw <- rawIntensity[,det.col]

row.names(sig.data.raw) <- rawIntensity$ProbeID
row.names(det.data.raw) <- rawIntensity$ProbeID

colnames(sig.data.raw) <- gsub(":AVG_Signal", "", colnames(sig.data.raw))
colnames(det.data.raw) <- gsub(":Detection Pval", "", colnames(det.data.raw))

info<-readr::read_tsv("data/YRI_sample_list_file_03NOV09.txt", 
		col_types=list("Date of Harvest"=col_date("%d.%m.%y"), 
				"Date of RNA extraction"=col_date("%d.%m.%y")))

## remove empty lines at end of file
info<-info[1:120,] 
factors <- which(sapply(info, is.factor))
for(i in factors){
	info[[i]] <- factor(as.character(info[[i]]))
}


newNames <- paste(info[["Cell Line"]], info$Treatment, sep="_")
names(newNames)<- info[["ARRAY CODE"]]

## typo in sample ID
names(sig.data.raw)[13]<- "NM-75"
names(det.data.raw)[13]<- "NM-75"

names(sig.data.raw) <-gsub("-", "_", fixed=TRUE, names(sig.data.raw))
names(det.data.raw) <-gsub("-", "_", fixed=TRUE, names(det.data.raw)) #differing name format
names(sig.data.raw) <- newNames[names(sig.data.raw)]
names(det.data.raw) <- newNames[names(det.data.raw)]

sig.data.log<-log2(sig.data.raw)
```
Probe intensities for resting and stimulated cells were imported into R for further
processing together with associated meta data.

```{r excludeSetup}
exclude <- list()
exclude$sample <- character()
exclude$probe <- character()
```

```{r longData}
sig.data.long <- reshape2::melt(sig.data.log, value.name="intensity")
splitLabel <- strsplit(as.character(sig.data.long$variable), "_")
sig.data.long$variable <- sapply(splitLabel, function(x) paste(x[-length(x)], collapse="_"))
sig.data.long$treatment <- sapply(splitLabel, function(x) x[length(x)])
sig.data.long$probe <- rep(rownames(sig.data.log), ncol(sig.data.log))
names(sig.data.long)[1] <- "sample"
sig.data.long <- as.tbl(sig.data.long[c("sample", "treatment", "probe", "intensity")])

det.data.long <- reshape2::melt(det.data.raw, value.name="p.value")
splitLabel <- strsplit(as.character(det.data.long$variable), "_")
det.data.long$variable <- sapply(splitLabel, function(x) paste(x[-length(x)], collapse="_"))
det.data.long$treatment <- sapply(splitLabel, function(x) x[length(x)])
det.data.long$probe <- rep(rownames(det.data.raw), ncol(det.data.raw))
names(det.data.long)[1] <- "sample"
det.data.long <- as.tbl(det.data.long[c("sample", "treatment", "probe", "p.value")])
```

Annotations for all probes were obtained via the *illuminaHumanv3.db* Bioconductor
package [@Barbosa10]. Only probes that are considered to are of perfect or good
quality were considered for analysis. Additionally, all probes mapping to more than
one genomic location or to a location that contains a known SNP were excluded.

```{r probeQual}
annot <- illuminaHumanv3fullReannotation()
annot <- subset(illuminaHumanv3fullReannotation(), 
		!grepl("Good|Perfect", ProbeQuality) | !is.na(OverlappingSNP) | 
		!is.na(SecondMatches) | !is.na(OtherGenomicMatches) | !is.na(RepeatMask))
exclude$probe <- union(exclude$probe, annot$ArrayAddress)
```


Probes were required to exhibit significant signal (detection p-value < 0.05) 
in at least seven samples and samples with less than 30% of the remaining probes 
providing significant signal were excluded (together with the paired sample).
Samples showing exceptionally low variation in probe intensities (standard deviation
of the log intensities of all retained probes below 0.7) were also removed.

```{r lowDetProbe}
detected <- tally(group_by(subset(det.data.long, p.value < 0.05), probe))
exclude$probe <- union(exclude$probe, filter(detected, n < 7)$probe)
```

```{r filterProbe}
sig.data.long <- filter(sig.data.long, !probe %in% exclude$probe)
det.data.long <- filter(det.data.long, !probe %in% exclude$probe)
```

```{r lowDetSample}
missingSample <- tally(group_by(filter(det.data.long, p.value > 0.05), sample, treatment))
missingSample$n <- missingSample$n/count(det.data.long, sample, treatment)$n
exclude$sample <- union(exclude$sample, filter(missingSample, n >= 0.7)$sample)
```

```{r lowSD}
sampleSD <- summarise(group_by(filter(sig.data.long, is.finite(intensity)), 
				sample, treatment), sd(intensity))
exclude$sample <- union(exclude$sample, filter(sampleSD, `sd(intensity)` < 0.7)$sample)
```

```{r filterSample}
sig.data.long <- filter(sig.data.long, !sample %in% exclude$sample)
det.data.long <- filter(det.data.long, !sample %in% exclude$sample)
```

After filtering `r length(unique(sig.data.long$probe))` of `r nrow(sig.data.raw)`
probes (`r format(length(unique(sig.data.long$probe))/nrow(sig.data.raw)*100, digits=2, nsmall=1)`%)
and `r length(unique(sig.data.long$sample))` samples remain. See `r figRef("densityPlot")`
for the distributions of intensities for the remaining probes in each sample. 

```{r densityPlot, fig.cap=figRef("densityPlot", "Distribution of probe intensities.")}
ggplot(sig.data.long, aes(x=intensity, colour=sample, linetype=treatment)) + 
		geom_density() + theme_bw() + guides(colour="none")
```

```{r pcaPrepare}
flt <- filter(sig.data.long, !is.finite(intensity))$probe
sig.data.filter <- full_join(spread(filter(sig.data.long, 
					treatment == "Basal" & !probe %in% flt), 
					sample, intensity)[-1],
			spread(filter(sig.data.long, treatment == "Stim" & !probe %in% flt), 
					sample, intensity)[-1],
			by="probe")
names(sig.data.filter) <- sub(".x", "_Basal", names(sig.data.filter), fixed=TRUE)
names(sig.data.filter) <- sub(".y", "_Stim", names(sig.data.filter), fixed=TRUE)

sig.pca <- prcomp(t(sig.data.filter[-1]), scale=TRUE)
sig.pc <- as.data.frame(sig.pca$x)
sig.pc$sample <- rownames(sig.pc)
sig.pc <- select(as.tbl(sig.pc), sample, PC1, PC2)
sig.pc <- inner_join(sig.pc, info, by="sample")
```

```{r pcaPlot, fig.cap=figRef("pcaPlot", "PCA plot of samples. Colours correspond to different BeadChips while shapes indicate different amplification dates. Samples from the two amplifications in October 2009 are clustered together.")}
ggplot(sig.pc, aes(x=PC1, y=PC2, colour=as.character(Sentrix_ID), shape=`Date cRNA Amplified`)) + 
	geom_point(size=3) + theme_bw() + scale_colour_discrete("BeadChip") + scale_shape_discrete("RNA Amplification")

```

A principle component analysis of the gene expression for the remaining samples indicates 
a clustering of samples from two RNA amplification batches in October 2009
(`r figRef("pcaPlot")`). This batch effect is noted for later correction during the analysis.

```{r sampleList}
samples <- sub("_.", "", sub("GM", "NA", unique(sig.data.long$sample)))
sampleInfo <- readr::read_delim("data/hapmap_YRI_r23a_filtered.fam", delim=" ", 
		col_names=c("Family", "Individual", "Father", "Mother", "Sex", "Phenotype"))
keepSamples <- filter(select(sampleInfo, Family, Individual), Individual %in% samples)
write.table(keepSamples, file="data/sample.lst", quote=FALSE, col.names=FALSE, row.names=FALSE)
```

Genotype data provided by the HapMap project [@Gibbs10] was processed with Plink [@Purcell07]
to restrict the data to autosomes and remove SNPs with low genotyping rate and 
and those with a minor allele frequency of less than 10% in the sample.


```{r plinkSNPs, engine="bash"}
plink --bfile data/hapmap_YRI_r23a_filtered --out hapmap_yri --keep data/sample.lst --maf 0.1 --geno 0.1 --split-x b37 no-fail --autosome-xy --make-bed --silent
```

# Results

# Appendix {-}
## Custom functions used
```{r, ref.label="functions", eval=FALSE, echo=TRUE}
```

## Session Info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References {-}
