---
title: Genome-wide analysis of heat shock response 
author: 
  - name: Peter Humburg
    affiliation: Wellcome Trust Centre for Human Genetics, University of Oxford, Roosevelt Dr., Oxford, OX3 7BN, UK
    email: peter.humburg@well.ox.ac.uk
  - name: Wanseon Lee
    affiliation: Wellcome Trust Centre for Human Genetics, University of Oxford, Roosevelt Dr., Oxford, OX3 7BN, UK
    email: wanseon.lee@well.ox.ac.uk
date: "`r format(Sys.time(), '%a %d %b %Y')`"
output:
  pdf_document:
    includes:
      in_header:
        - include/captions.tex
        - include/encoding.tex
    toc: true
    fig_caption: true
    highlight: tango
    template: include/report.latex
    pandoc_args: ["--filter=/analysis/include/hidden-print.py"]
  knitrBootstrap::bootstrap_document:
    includes:
     in_header:  include/report.css
    title: Genome-wide analysis of heat shock response
    highlight: github 
    theme.chooser: true
    highlight.chooser: true
    toc: true
    fig_caption: true
    pandoc_args: ["--filter=/analysis/include/go-link.py"]
bibliography: include/references.bib
geometry: margin=2cm
documentclass: report
classoption: a4paper
---

```{r, setup, include=FALSE, cache=FALSE}
library(MASS)
library(proto)
library(gdata)
library(affy)
library(vsn)
library(limma)
library(scatterplot3d)
library(sva)
library(gplots)
library(RColorBrewer)
library(ggplot2)
library(dplyr)
library(tidyr)
library(readr)
library(stringr)
library(illuminaHumanv3.db)
library(knitr)
library(pander)
library(reshape2)
library(ggdendro)
library(topGO)
library(SNPlocs.Hsapiens.dbSNP142.GRCh37)
library(BSgenome.Hsapiens.UCSC.hg19)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(org.Hs.eg.db)
library(plink2R)
library(MatrixEQTL)
library(plsdepot)
library(MultiPhen)
library(RISmed)
library(biomaRt)
library(PWMEnrich)
library(PWMEnrich.Hsapiens.background)

opts_chunk$set(autodep=TRUE)
opts_chunk$set(tidy=TRUE)
opts_chunk$set(echo=TRUE)
opts_chunk$set(cache=TRUE)
opts_chunk$set(fig.width=8, fig.height=8, dpi=300)
opts_chunk$set(collapse=TRUE, hold=TRUE)
opts_chunk$set(bootstrap.type="button")
opts_chunk$set(bootstrap.show.code=FALSE,
		bootstrap.show.output=FALSE,
		bootstrap.show.message=FALSE,
		bootstrap.show.warning=FALSE,
		bootstrap.show.error=TRUE)
panderOptions("digits", 3)
panderOptions("table.split.table", 160)
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")
options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")
```


```{r, functions, include=FALSE}
reshapeData <- function(data, value){
	ans <- reshape2::melt(data, id.vars="probe",
		value.name=value, variable.name="sample")
	splitLabel <- strsplit(as.character(ans$sample), "_", fixed=TRUE)
	ans$sample <- sapply(splitLabel, "[", 1)
	ans$treatment <- sapply(splitLabel, function(x) x[length(x)])
	as.tbl(ans[c("sample", "treatment", "probe", value)])
}

ibsClust <- function(genome, ...){
	dst <- spread(dplyr::select(genome, IID1, IID2, DST), IID2, DST)
	size <- nrow(dst) + 1
	labels <- c(as.character(dst$IID1[1]), names(dst)[-1])
	dst <- t(1 - as.matrix(dst[-1]))
	dst <- dst[!is.na(dst)]
	attr(dst, "Size") <- size
	attr(dst, "Labels") <- labels
	attr(dst, "Diag") <- FALSE
	attr(dst, "Upper") <- FALSE
	class(dst) <- "dist"
	
	hclust(dst, ...)
}

pcaClusterPlot <- function(data, i, j, fillColour, legend=TRUE){
    data <- data[, c("sample", paste0("PC", i), paste0("PC", j), "Sentrix_ID", "cluster")]
    names(data) <- c("sample", "x", "y", "Sentrix_ID", "cluster")
    clusters <- by(data, data$cluster, function(x) x)
    fig <- ggplot(data, aes(x=x, y=y))
    for(cl in clusters){
        ch <- chull(cl$x, cl$y)
        fig <- fig + geom_polygon(data=cl[ch,], fill=fillColour[cl$cluster[1]], alpha=0.7)
    }
    fig <- fig + geom_point(aes(colour=as.character(Sentrix_ID)), size=3) +
            theme_bw() + scale_colour_discrete("BeadChip") + xlab(paste0("PC", i)) + ylab(paste0("PC", j))
    if(!legend) fig <- fig + theme(legend.position="none")
    fig
}

adjustLabel <- function(x, y, centre){
	just <- c(h=0, v=0)
	if(x < centre[1]){
		just["h"] <- just["h"] + 0.9
	} 
	if(y < centre[2]){
		just["v"] <- just["v"] + 1
	}
	just
}

countBlocks <- function(geno, pval, pcut=0.05, rcut=0.8){
  count <- 0
  selected <- which(pval < pcut)
  if(length(selected)){
    if(length(selected) == 1){
      count <- 1
    } else{
      if(names(pval)[1] %in% colnames(geno)) geno <- geno[, names(pval)[selected]]
      else geno <- t(geno[names(pval)[selected], ])
      r2 <-cor(geno, use="pairwise", method="spearman")^2
      remain <- 1:ncol(r2)
      while(length(remain)){
        i <- remain[1]
        count <- count + 1
        remain <- intersect(which(r2[i, ] < rcut), remain)
      }
    }
  }
  count
}

## Create a scatter plot with one panel per group (by default BeadChip)
## Data points (by default samples in PC space) are assumed to be from a 
## bivariate t distribution. Confidence ellipses are drawn based on this
## (at a 95% level by default) and points outside these ellipses are marked 
## as outliers. 
outlierPlot <- function(data, x="PC1", y="PC2", group="Sentrix_ID", 
			shape="Treatment", id="Cell Line", level=0.95, ...){
	ellipses <- by(data[, c(x, y)], data[[group]], 
			function(d) tryCatch(car::dataEllipse(as.matrix(d), 
						levels=level, draw=FALSE, robust=TRUE), 
                    error=function(e) cbind(NA, NA)))
	data[[group]] <- as.character(data[[group]])
	
	names(data)[which(names(data) == x)] <- make.names(x)
	x <- make.names(x)
	names(data)[which(names(data) == y)] <- make.names(y)
	y <- make.names(y)
	names(data)[which(names(data) == group)] <- make.names(group)
	group <- make.names(group)
	names(data)[which(names(data) == shape)] <- make.names(shape)
	shape <- make.names(shape)
	names(data)[which(names(data) == id)] <- make.names(id)
	id <- make.names(id)
	
	data$outlier <- logical(nrow(data))
	for(i in 1:nrow(data)){
		data$outlier[i] <- !sp::point.in.polygon(data[i, x], data[i, y], 
				ellipses[[as.character(data[i, group])]][,1], 
				ellipses[[as.character(data[i, group])]][,2])
	}
	data$colour <- as.character(data[[group]])
	data$colour[data$outlier] <- NA
	centre <- by(data[, c(x, y)], data[[group]], colMeans)
	data$h <- if(!is.null(list(...)$hjust)) list(...)$hjust else 0
	data$v <- if(!is.null(list(...)$vjust)) list(...)$vjust else 0
	labelJust <- mapply(function(x, y, g, c) adjustLabel(x, y, c[[g]]), 
			data[[x]], data[[y]], data[[group]], MoreArgs=list(centre))
	data$h <- ifelse(labelJust["h", ] > 0, -data$h, data$h)
	data$h <- data$h + labelJust["h", ]
	data$v <- ifelse(labelJust["v", ] > 0, -data$v, data$v)
	data$v <- data$v + labelJust["v", ]			
	ggplot(data, aes_string(x=x, y=y, colour="colour")) +
			geom_point(aes_string(shape=shape), size=3) + 
			stat_ellipse(level=level) + facet_wrap(eval(parse(text=paste("~", group))), ncol=3) +
			geom_text(data=dplyr::filter(data, outlier), 
					aes_string(label=id, hjust="h", vjust="v"), size=4) +
			theme_bw() + scale_colour_discrete(guide="none")
}

plsPlot <- function(plsComp, genotype){
  treatGrp <- by(dplyr::select(plsComp, t1, t2), 
                 list(plsComp$treatment, plsComp$sex), colMeans)
  grps <- do.call(expand.grid, dimnames(treatGrp))
    grps <- dplyr::filter(grps, !sapply(treatGrp, is.null))
    names(grps) <- c("treatment", "sex")
  treatAvg <- cbind(do.call(rbind, treatGrp), grps)
  treatAvg$sample <- treatAvg$sex
  
  if(!missing(genotype)){
    plsComp[[genotype]] <- as.factor(plsComp[[genotype]])
    grpMean <- by(dplyr::select(plsComp, t1, t2), list(plsComp[[genotype]], 
                                                       plsComp$treatment, plsComp$sex), colMeans)
    grps <- do.call(expand.grid, dimnames(grpMean))
    grps <- dplyr::filter(grps, !sapply(grpMean, is.null))
    names(grps) <- c(genotype, "treatment", "sex")
    genoAvg <- cbind(do.call(rbind, grpMean), grps) 
    genoAvg$sample <- paste(genoAvg[[genotype]], genoAvg[["sex"]], sep="_")
  }
  
  fig <- ggplot(plsComp, aes(x=t1, y=t2, group=sample, shape=sex))
  if(missing(genotype)){
    fig <- fig + geom_point(data=treatAvg, aes(colour=treatment), size=5)
  } else {
      fig <- fig + geom_point(data=treatAvg, size=5, colour="black")
      fig <- fig + geom_point(data=genoAvg, aes_string(colour=genotype), size=5)
      fig <- fig + geom_line(data=genoAvg, aes_string(colour=genotype),  line.width=2,
              arrow=grid::arrow(type="closed", length=grid::unit(0.15, "inches"), angle=20))
  }
  fig <- fig + geom_line(data=treatAvg, colour="black",  line.width=2,
              arrow=grid::arrow(type="closed", length=grid::unit(0.15, "inches"), angle=20))
  if(missing(genotype)){
    fig <- fig + geom_point(aes(colour=treatment), size=3, alpha=0.6)
    fig <- fig + scale_colour_manual(values=c(Basal="blue", Stim="red"))
  } else {
    fig <- fig + geom_point(aes_string(colour=genotype), size=3, alpha=0.6)
    fig <- fig + scale_colour_manual(values=c("0"="#3E79F7", "1"="#FA2AA4", "2"="#FFB72B"))
  }
  fig <- fig + geom_line(colour="grey", alpha=0.6,
                       arrow=grid::arrow(type="closed", length=grid::unit(0.15, "inches"), angle=20))
  
  fig <- fig + theme_bw() + xlab("Component 1") + ylab("Component 2")
  fig
}


# Multiple plot function
#
# ggplot objects can be passed in ..., or to plotlist (as a list of ggplot objects)
# - cols:   Number of columns in layout
# - layout: A matrix specifying the layout. If present, 'cols' is ignored.
#
# If the layout is something like matrix(c(1,2,3,3), nrow=2, byrow=TRUE),
# then plot 1 will go in the upper left, 2 will go in the upper right, and
# 3 will go all the way across the bottom.
#
# author: Winston Chang ()
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)

  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)

  numPlots = length(plots)

  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                    ncol = cols, nrow = ceiling(numPlots/cols))
  }

 if (numPlots==1) {
    print(plots[[1]])

  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))

    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))

      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}

filterSample <- function(sig, sig.raw, det, excl){
  sig <- filter(sig, !sample %in% excl)
  sig.raw <- filter(sig.raw, !sample %in% excl)
  det <- filter(det, !sample %in% excl)
  
  list(sig=sig, sig.raw=sig.raw, det=det)
}

updateSampleList <- function(samples, sampleFile){
  samples <- sub("_.", "", unique(samples))
  sampleInfo <- readr::read_delim(sampleFile, delim=" ", 
                                  col_names=c("Family", "Individual", "Father", "Mother", "Sex", "Phenotype"))
  keepSamples <- dplyr::filter(dplyr::select(sampleInfo, Family, Individual), Individual %in% samples)
  write.table(keepSamples, file="tmp/sample.lst", quote=FALSE, col.names=FALSE, row.names=FALSE)
  invisible(keepSamples)
}

formatQTL <- function (me, limmaTable) {
  eQTL <- me$all$eqtls
  eQTL <- dplyr::left_join(eQTL, dplyr::select(limmaTable, NuID, SymbolReannotated, logFC), by=c(gene="NuID"))
  eQTL <- dplyr::select(eQTL, SNP, SymbolReannotated, NuID, beta:FDR, AveExpr, logFC)
  eQTL <- dplyr::rename(eQTL, Gene="SymbolReannotated")
  eQTL
}

promoterScan <- function(seqs, pwms, group){
  res <- motifEnrichment(seqs, pwms)
  report <- lapply(1:length(seqs), function(x, i) sequenceReport(x, i)@d, x=res)
  report <- do.call(rbind, report)
  cbind(report,  EntrezReannotated=names(seqs), expr=group)
}

figRef <- local({
    tag <- numeric()
    created <- logical()
    used <- logical()
    function(label, caption, prefix = options("figcap.prefix"), 
        sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight")) {
        i <- which(names(tag) == label)
        if (length(i) == 0) {
            i <- length(tag) + 1
            tag <<- c(tag, i)
            names(tag)[length(tag)] <<- label
            used <<- c(used, FALSE)
            names(used)[length(used)] <<- label
            created <<- c(created, FALSE)
            names(created)[length(created)] <<- label
        }
        if (!missing(caption)) {
            created[label] <<- TRUE
            paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
                " ", caption)
        } else {
            used[label] <<- TRUE
            paste(prefix, tag[label])
        }
    }
})

tabRef <- local({
    tag <- numeric()
    created <- logical()
    used <- logical()
    function(label, caption, prefix = options("tabcap.prefix"), 
        sep = options("tabcap.sep"), prefix.highlight = options("tabcap.prefix.highlight")) {
        i <- which(names(tag) == label)
        if (length(i) == 0) {
            i <- length(tag) + 1
            tag <<- c(tag, i)
            names(tag)[length(tag)] <<- label
            used <<- c(used, FALSE)
            names(used)[length(used)] <<- label
            created <<- c(created, FALSE)
            names(created)[length(created)] <<- label
        }
        if (!missing(caption)) {
            created[label] <<- TRUE
            paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
                " ", caption)
        } else {
            used[label] <<- TRUE
            paste(prefix, tag[label])
        }
    }
})
```

<div class="col-md-3 hidden-print hidden-xs col-sm-6 well pull-right downloads">

### Downloads {-}
#### Report
<span title="Download report as PDF">[PDF](heatshock_analysis.pdf)</span>

#### Data

```{r linkData, echo=FALSE, cache=FALSE, results="asis"}
dataFiles <- sapply(data, linkFile, "data", "markdown")
desc <- sapply(data, "[[", "desc")
cat(paste0("<span title=\"", desc, "\">", dataFiles, "</span>", collapse="<br>"))
```

#### Results

```{r linkResults, echo=FALSE, cache=FALSE, results="asis"}
resultFiles <- sapply(results, linkFile, "results", "markdown")
desc <- sapply(data, "[[", "desc")
cat(paste0("<span title=\"", desc, "\">", resultFiles, "</span>", collapse="<br>"))
```


#### Log files

```{r linkLogs, echo=FALSE, cache=FALSE, results="asis"}
logFiles <- sapply(logs, linkLog, "markdown")
desc <- sapply(logs, "[[", "desc")
cat(paste0("<span title=\"", desc, "\">", logFiles, "</span>", collapse="<br>"))
```
</div>

# Introduction
The heat shock response is a highly conserved mechanism responsible for the 
maintenance of cellular function under stress. Originally identified as part of the 
response to increased environmental temperatures [@Ritossa62] heat shock proteins (HSPs) 
have since been implicated  in susceptibility to infection [@Kee08], cancer risk [@Sfar10] 
and drug response [@Martin04]. Despite its obvious significance many details of the
heat shock response still remain uncertain.

Here we study the genome-wide changes to gene expression induced by heat shock in
60 HapMap cell lines from Yoruba individuals [@Gibbs10] to identify genes and pathways
involved in the human heat shock response. To further elucidate underlying mechanisms
genetic variants modulating gene expression are mapped across all relevant genes.

# Methods

## Pre-processing and Quality Control

### Raw gene expression QC

```{r loadExpression, results="hide"}
rawIntensity <- readr::read_tsv("data/TPS00590_FinalReport_12Oct09.txt", skip=8, progress=FALSE)

sig.col<-seq(3, 722, 6)
det.col<-seq(8, 722, 6)

sig.data.raw <- rawIntensity[,sig.col]
det.data.raw <- rawIntensity[,det.col]

names(sig.data.raw) <- gsub(":AVG_Signal", "", names(sig.data.raw))
names(det.data.raw) <- gsub(":Detection Pval", "", names(det.data.raw))

info<-readr::read_tsv("data/YRI_sample_list_file_03NOV09.txt", 
		col_types=list("Date of Harvest"=col_date("%d.%m.%y"), 
				"Date of RNA extraction"=col_date("%d.%m.%y")))

## remove empty lines at end of file
info<-info[1:120,] 
factors <- which(sapply(info, is.factor))
for(i in factors){
	info[[i]] <- factor(as.character(info[[i]]))
}
info[["Cell Line"]] <- sub("^GM", "NA", info[["Cell Line"]])
info[["Cell Line"]] <- sub("_.$", "", info[["Cell Line"]])
info <- mutate(info, sample=paste(`Cell Line`, `Treatment`, sep="_"))

newNames <- paste(info[["Cell Line"]], info$Treatment, sep="_")
names(newNames)<- info[["ARRAY CODE"]]

## typo in sample ID
names(sig.data.raw)[13]<- "NM-75"
names(det.data.raw)[13]<- "NM-75"

names(sig.data.raw) <-gsub("-", "_", fixed=TRUE, names(sig.data.raw))
names(det.data.raw) <-gsub("-", "_", fixed=TRUE, names(det.data.raw)) #differing name format
names(sig.data.raw) <- newNames[names(sig.data.raw)]
names(det.data.raw) <- newNames[names(det.data.raw)]

sig.data.raw <- cbind(probe=rawIntensity$ProbeID, sig.data.raw)
det.data.raw <- cbind(probe=rawIntensity$ProbeID, det.data.raw)

sig.data.log<- cbind(probe=rawIntensity$ProbeID, log2(sig.data.raw[-1]))
```
Probe intensities for resting and stimulated cells were imported into R for further
processing together with associated meta data.

```{r excludeSetup}
exclude <- list()
exclude$sample <- character()
exclude$probe <- character()
```

```{r longData}
sig.data.long <- reshapeData(sig.data.log, "intensity")
sig.data.raw.long <- reshapeData(sig.data.raw, "intensity")
det.data.long <- reshapeData(det.data.raw, "p.value")
```

Annotations for all probes were obtained via the *illuminaHumanv3.db* Bioconductor
package [@Barbosa10]. Only probes that are considered to be of perfect or good
quality were taken forward for analysis. Additionally, all probes mapping to more than
one genomic location or to a location that contains a known SNP were excluded.

```{r probeQual}
annot <- illuminaHumanv3fullReannotation() 
exclude$probe <- union(exclude$probe, dplyr::filter(annot, 
		!grepl("Good|Perfect", ProbeQuality) | !is.na(OverlappingSNP) | 
		!is.na(RepeatMask))$ArrayAddress)
annot <- dplyr::filter(annot, !ArrayAddress %in% exclude$probe)

txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
probePos <- strsplit(dplyr::filter(annot, !is.na(GenomicLocation))$GenomicLocation, "[:, ]")
probePos <- data.frame(ArrayAddress=dplyr::filter(annot, !is.na(GenomicLocation))$ArrayAddress,
                       chrom=sapply(probePos, "[[", 1), 
                       start=as.numeric(sapply(probePos, "[[", 2)),
                       end=as.numeric(sapply(probePos, "[[", 3)), 
                       strand=sapply(probePos, "[[", 4))
probePos <- makeGRangesFromDataFrame(probePos, keep.extra.columns=TRUE)
tx <- transcriptsByOverlaps(txdb, probePos)
hits <- findOverlaps(probePos, tx)
probeTx <- data.frame(ArrayAddress=probePos$ArrayAddress[queryHits(hits)], tx_id=tx$tx_id[subjectHits(hits)] )
tx2gene <- transcripts(txdb, vals=list(tx_id=unique(probeTx$tx_id)), columns=c("tx_id", "gene_id"))
tx2gene <- dplyr::select(as.data.frame(tx2gene), tx_id, gene_id)
probeGene <- dplyr::left_join(probeTx, tx2gene, by="tx_id")
probeGene <- unique(dplyr::select(probeGene, ArrayAddress, gene_id))
probeGene$gene_id <- as.character(probeGene$gene_id)

excl <- table(probeGene$ArrayAddress)
exclude$probe <- union(exclude$probe, names(excl)[excl > 1])
annot <- dplyr::filter(annot, !ArrayAddress %in% exclude$probe)

annot <- dplyr::left_join(annot, probeGene, by="ArrayAddress")
annot$EntrezReannoated <- ifelse(is.na(annot$gene_id), annot$EntrezReannotated, annot$gene_id)
annot <- dplyr::select(annot, -gene_id)

eg <- mappedkeys(org.Hs.egSYMBOL)
symbols <- as.list(org.Hs.egSYMBOL[eg])
symbols <- data.frame(entrez=names(symbols), symbol=unlist(symbols), stringsAsFactors=FALSE)
annot <- dplyr::left_join(annot, symbols, by=c(EntrezReannotated="entrez"))
annot$SymbolReannotated <- ifelse(is.na(annot$symbol), annot$SymbolReannotated, annot$symbol)
annot <- dplyr::select(annot, -symbol)
```

```{r thresholds}
det.pval <- 0.01
det.n <- 10
sample.missing <- 0.7
sd.min <- 0.8
```

Probes were required to exhibit significant signal (detection p-value < `r det.pval`) 
in at least `r det.n` samples and samples with less than `r (1-sample.missing)*100`% 
of the remaining probes providing significant signal were excluded 
(together with the paired sample). Samples showing exceptionally low variation in 
probe intensities (standard deviation of the log intensities of all retained probes 
below `r sd.min`) were also removed.

```{r lowDetProbe}
detected <- tally(group_by(subset(det.data.long, p.value < det.pval), probe))
exclude$probe <- union(exclude$probe, filter(detected, n < det.n)$probe)
```

```{r filterProbe}
sig.data.long <- filter(sig.data.long, !probe %in% exclude$probe)
sig.data.raw.long <- filter(sig.data.raw.long, !probe %in% exclude$probe)
det.data.long <- filter(det.data.long, !probe %in% exclude$probe)
```

```{r lowDetSample}
missingSample <- tally(group_by(filter(det.data.long, p.value > det.pval), sample, treatment))
missingSample$n <- missingSample$n/count(det.data.long, sample, treatment)$n
exclude$sample <- union(exclude$sample, filter(missingSample, n >= sample.missing)$sample)
```

```{r lowSD}
sampleSD <- summarise(group_by(filter(sig.data.long, is.finite(intensity)), 
				sample, treatment), sd(intensity))
exclude$sample <- union(exclude$sample, filter(sampleSD, `sd(intensity)` < sd.min)$sample)
```

```{r filterSample}
dat <- filterSample(sig.data.long, sig.data.raw.long, det.data.long, exclude$sample)
sig.data.long <- dat$sig
sig.data.raw.long <- dat$sig.raw
det.data.long <- dat$det
```

After filtering `r length(unique(sig.data.long$probe))` of `r nrow(sig.data.raw)`
probes (`r format(length(unique(sig.data.long$probe))/nrow(sig.data.raw)*100, digits=2, nsmall=1)`%)
and `r length(unique(sig.data.long$sample))` samples remain. See `r figRef("densityPlot")`
for the distributions of intensities for the remaining probes in each sample. 

```{r densityPlot, cache=FALSE, fig.cap=figRef("densityPlot", "Distribution of probe intensities.")}
ggplot(sig.data.long, aes(x=intensity, colour=sample, linetype=treatment)) + 
		geom_density() + theme_bw() + guides(colour="none")
```

```{r pcaPrepare}
flt <- filter(sig.data.long, !is.finite(intensity))$probe
sig.data.filter <- full_join(spread(filter(sig.data.long, 
					treatment == "Basal" & !probe %in% flt), 
					sample, intensity)[-1],
			spread(filter(sig.data.long, treatment == "Stim" & !probe %in% flt), 
					sample, intensity)[-1],
			by="probe")
names(sig.data.filter) <- sub(".x", "_Basal", names(sig.data.filter), fixed=TRUE)
names(sig.data.filter) <- sub(".y", "_Stim", names(sig.data.filter), fixed=TRUE)

sig.pca <- prcomp(t(sig.data.filter[-1]), scale=TRUE)
sig.pc <- as.data.frame(sig.pca$x)
sig.pc$sample <- rownames(sig.pc)
sig.pc <- dplyr::select(as.tbl(sig.pc), sample, PC1, PC2)
sig.pc <- inner_join(sig.pc, info, by="sample")

cluster <- dplyr::filter(sig.pc, `Date cRNA Amplified` != "CE_29Sep09")
clusterBound <- chull(cluster$PC1, cluster$PC2) 
```

```{r pcaPlot, cache=FALSE, fig.cap=figRef("pcaPlot", "PCA plot of probe intensities by sample. Colours correspond to different BeadChips while shapes indicate different amplification dates. Samples from the two amplifications in October 2009 are clustered together (highlighted by the grey area).")}
ggplot(sig.pc, aes(x=PC1, y=PC2)) + 
     geom_polygon(data=cluster[clusterBound, ], fill="lightgrey", colour="grey") + 
     geom_point(aes( colour=as.character(Sentrix_ID), shape=`Date cRNA Amplified`), size=3) + 
     theme_bw() + scale_colour_discrete("BeadChip") + scale_shape_discrete("RNA Amplification")
```

A principle component analysis of the gene expression for the remaining samples indicates 
a clustering of samples from two RNA amplification batches in October 2009
(`r figRef("pcaPlot")`). This batch effect is noted for later correction during the analysis.

Further analysis of the gene expression response to heat shock revealed that four of the samples
(NA18502, NA18508, NA18522, NA18853, NA18504) exhibit patterns of expression inconsistent with those observed
in other samples, including reversal of fold change for strongly differentially expressed
genes. This suggests a potential mislabelling of these samples, which were consequently excluded.

```{r extraExclusions}
exclude$sample <- union(exclude$sample, c("NA18502", "NA18508", "NA18522", "NA18853", "NA18504"))
dat <- filterSample(sig.data.long, sig.data.raw.long, det.data.long, exclude$sample)
sig.data.long <- dat$sig
sig.data.raw.long <- dat$sig.raw
det.data.long <- dat$det
```

### Genotype QC

```{r sampleList, cache=FALSE}
keepSamples <- updateSampleList(sig.data.long$sample, "data/hapmap_YRI_r23a_filtered.fam")
```

```{r plinkSNPs, engine="bash", cache=FALSE}
plink --bfile data/hapmap_YRI_r23a_filtered --out tmp/hapmap_yri.snp_flt --keep tmp/sample.lst --maf 0.1 --geno 0.1 --hwe 1e-4 --autosome --make-bed > log/hapmap_yri.snp_flt.log
```

```{r snpFilter, cache.extra=list(file.info("data/hapmap_YRI_r23a_filtered.bim")$mtime, file.info("tmp/hapmap_yri.bim")$mtime)}
totalSNPs <- R.utils::countLines("data/hapmap_YRI_r23a_filtered.bim")
remainSNPs <- R.utils::countLines("tmp/hapmap_yri.snp_flt.bim")
```

Genotype data provided by the HapMap project [@Gibbs10] was processed with Plink [@Chang15]
to restrict the data to autosomes and remove SNPs with low genotyping rate and 
and those with a minor allele frequency of less than 10% in the sample. This results
in the exclusion of `r totalSNPs - remainSNPs` of 
`r totalSNPs` SNPs (`r format((totalSNPs - remainSNPs)/totalSNPs*100, nsmall=2, digits=2)`%);

```{r plinkSample, engine="bash", cache=FALSE}
plink --bfile tmp/hapmap_yri.snp_flt --out tmp/hapmap_yri.smpl_flt --mind 0.1 --neighbour 1 3 --genome gz nudge --make-bed > log/hapmap_yri.smpl_flt.log
```

```{r nnIBS, cache=FALSE}
nnIBS <- readr::read_table("tmp/hapmap_yri.smpl_flt.nearest")
```

The IBS nearest neighbour calculation shows that the majority of samples are quite
similar with regard to the extend of genetic differences between them while a few
show markedly increased sharing of alleles (see `r figRef("nnIBS")` and `r tabRef("nnIBS")`). 
This is consistent with a population of largely unrelated individuals including a few 
individuals that are more closely related than expected. One individual (NA18913) also shows
evidence of reduced similarity with the second and third nearest neighbour.

```{r nnIBSplot, cache=FALSE, fig.cap=figRef("nnIBS", "Similarity between individuals. For each individual the similarity, in terms of IBS, to all other individuals is determined and individuals are ranked acording to this measure. From this a Z-score is computed for for each of the three closest neighbours (see the Plink [documentation](http://pngu.mgh.harvard.edu/~purcell/plink/strat.shtml#outlier) for details). Note the spikes corresponding to individuals with positive Z-scores indicating increased similarity of their genomes compared to the rest of the cohort.")}
ggplot(nnIBS, aes(x=Z, colour=as.factor(NN))) + geom_density(aes(group=NN)) + theme_bw()
```

```{r nnIBStable, cache=FALSE}
offset <- mean(filter(nnIBS, Z < 0 & NN == 1)$Z)
nnSuspect <- filter(nnIBS, (abs(Z - offset) > 2 & NN == 1) | abs(Z) > 2)
nnSuspect <- nnSuspect[order(abs(nnSuspect$Z), decreasing=TRUE), ]
set.caption(tabRef("nnIBS", 
		"Pairs of Yoruban HapMap samples with evidence of decreased or increased genetic similarity (absolute IBS-based nearest neighbour Z-score > 2)."))
pander(nnSuspect)
```

```{r loadIBD}
ibdReport <- read.table("tmp/hapmap_yri.smpl_flt.genome.gz", header=TRUE)
ibdSD <- sd(ibdReport$PI_HAT)
ibdHigh <- dplyr::filter(ibdReport, PI_HAT > 2*ibdSD)
```

A similar picture emerges if we consider the estimated proportion of IBD for all
sample pairs with `r nrow(ibdHigh)` pairs showing evidence of higher than expected
relatedness (see `r tabRef("ibdHigh")` and `r figRef("ibsDendro")` for details).

```{r selectExlusions}
pairs <- rbind(as.matrix(dplyr::select(ibdHigh, IID1, IID2)), 
		as.matrix(dplyr::select(nnSuspect, IID, IID2)))
exSample <- character()
while(length(pairs)){
	ids <- sort(table(pairs), decreasing=TRUE)
	exSample <- c(exSample, names(ids[1]))
	pairs <- pairs[!apply(pairs, 1, `%in%`, x=names(ids)[1]), , drop=FALSE]
}
if("NA19130" %in% exSample && !"NA19192" %in% exSample){
	exSample[which(exSample == "NA19130")] <- "NA19192"
}

exclude$sample <- union(exclude$sample, exSample)
```

To ensure that at most one sample of each suspect pair are included in the analysis
the samples `r pander(exSample)` are excluded.

```{r ibdHighTab}
set.caption(tabRef("ibdHigh", paste("Sample pairs with estimated proportion of IBD exceeding",
		format(2*ibdSD, nsmall=2, digits=2))))
pander(dplyr::select(ibdHigh, IID1, IID2, PI_HAT, DST, RATIO))
```

```{r ibsDendroPlot, cache=FALSE, fig.cap=figRef("ibsDendro", "Dendrogram of individuals included in study. Distances are based on identity by state. Three pairs show clear indications of increased relatedness.")}
plot(ibsClust(ibdReport), main="", xlab="IBS", sub="")
```

```{r filterSample2}
dat <- filterSample(sig.data.long, sig.data.raw.long, det.data.long, exclude$sample)
sig.data.long <- dat$sig
sig.data.raw.long <- dat$sig.raw
det.data.long <- dat$det
```

```{r sampleList2, cache=FALSE}
keepSamples <- updateSampleList(sig.data.long$sample, "data/hapmap_YRI_r23a_filtered.fam")
```

```{r plinkMdsPlot, cache=FALSE, engine="bash"}
plink --bfile tmp/hapmap_yri.smpl_flt --out tmp/hapmap_yri.smpl_flt2 --keep tmp/sample.lst --cluster --mds-plot 4 --make-bed > log/hapmap_yri.smpl_flt.log
```

```{r mdsPlot, cache=FALSE, fig.cap=figRef("mds", "MDS plot of samples remaining after genotype QC.")}
mds <- read.table("tmp/hapmap_yri.smpl_flt2.mds", header=TRUE)
ggplot(mds, aes(x=C1, y=C2)) + geom_point(size=3) + theme_bw()
```

Plotting the first two MDS components of the remaining `r nrow(keepSamples)` samples
shows no further evidence for outliers (`r figRef("mds")`).

### Normalising gene expression estimates

```{r vsn}
sig.data.norm <- justvsn(acast(sig.data.raw.long, formula=probe ~ sample + treatment, 
		value.var="intensity"))
```

```{r selectProbes}
probeMean <- rowMeans(sig.data.norm)
probeSD <- apply(sig.data.norm, 1, sd)
sig.select <- as.tbl(as.data.frame(sig.data.norm))
sig.select <- mutate(sig.select, mean=probeMean, sd=probeSD)
sig.select <- filter(sig.select, mean > 7)
sig.select <- arrange(sig.select, desc(sd), desc(mean))
sig.select <- head(sig.select, 2000)
```

```{r vsnPCA}
sig.pca <- prcomp(t(sig.select), scale=TRUE)
sig.pc <- as.data.frame(sig.pca$x)
sig.pc$sample <- rownames(sig.pc)
sig.pc <- dplyr::select(as.tbl(sig.pc), sample, PC1, PC2, PC3, PC4)
sig.pc <- inner_join(sig.pc, info, by="sample")
```

```{r vsnPCAplot, cache=FALSE, fig.cap=figRef("vsnPCA", "PCA plot of normalised gene expression estimates by sample. Each panal is comprised of the remaining samples for a singel BeadChip. For each BeadChip a 99% confidence ellipse is shown. While the majority of samples cluster well within chips some clear ouliers are visible. Variability between chips is relatively high with chips 563403730 and 563403752 appearing distinct from the bulk of the samples.")}
ggOutlier <- outlierPlot(sig.pc, level=0.985, hjust=0.2, vjust=-0.5)
sampleCount <- table(ggOutlier$data$Sentrix_ID)
excludeChip <- unique(dplyr::filter(ggOutlier$data, Sentrix_ID %in% names(sampleCount)[sampleCount <= 4])$Sentrix_ID)
outlierSample <- unique(dplyr::filter(ggOutlier$data, outlier)$Cell.Line)
chipExclusion <- unique(dplyr::filter(ggOutlier$data, Sentrix_ID %in% excludeChip)$Cell.Line)
exclude$sample <- Reduce(union, list(exclude$sample, outlierSample, chipExclusion))
ggOutlier
```

Probe intensities were normalised with VSN [@Huber02]. A PCA plot of the initial 
VSN normalised data (`r figRef("vsnPCA")`) shows clear evidence of samples
clustering by BeadChip, as well as 
`r length(outlierSample) - length(excludeChip)` 
obvious outliers, which were removed. In addition the few remaining samples from chip
`r paste(excludeChip, collapse=", ")` 
were removed as the entire chip appears to be suspect.

```{r excludeOutliers, dependson="vsnOutliers", cache.extra=exclude$sample}
dat <- filterSample(sig.data.long, sig.data.raw.long, det.data.long, exclude$sample)
sig.data.long <- dat$sig
sig.data.raw.long <- dat$sig.raw
det.data.long <- dat$det
```

```{r exclOutlGeno, cache=FALSE}
keepSamples <- updateSampleList(sig.data.long$sample, "tmp/hapmap_yri.smpl_flt2.fam")
```

```{r plinkExclude, engine="bash", cache=FALSE}
plink --bfile tmp/hapmap_yri.smpl_flt2 --out tmp/hapmap_yri.smpl_flt3 --keep tmp/sample.lst --make-bed > log/hapmap_yri.smpl_flt3.log
```

The remaining samples were normalised separately for each BeadChip and differences 
between groups corrected with ComBat [@Johnson07;@SVA15], preserving differences due to 
heat shock stimulation.

```{r vsnBatch, cache.extra=exclude$sample}
batch <- with(subset(info, !`Cell Line` %in% exclude$sample), 
		data.frame(sample=sample, cluster=as.integer(factor(Sentrix_ID)), 
			treatment=Treatment, stringsAsFactors=FALSE))
sig.raw <- acast(sig.data.raw.long, formula=probe ~ sample + treatment, 
		value.var="intensity")
colnames(sig.raw) <- sub("_._", "_", colnames(sig.raw))
sig.raw <- sig.raw[, batch$sample]

vsnFit <- vector(mode="list", max(batch$cluster))
sig.data.norm <- vector(mode="list", max(batch$cluster))
for(i in 1:max(batch$cluster)){
	vsnFit[[i]] <- vsn2(sig.raw[, batch$cluster==i])
	sig.data.norm[[i]] <- predict(vsnFit[[i]], newdata=sig.raw[, batch$cluster==i], 
			useDataInFit = TRUE)
}
sig.data.norm <- do.call(cbind, sig.data.norm)
```

```{r vsnBatchFit, cache=FALSE, results="hide", fig.cap=figRef("vsnBatchFit", "VSN fits to grouped dataset.")}
par(mfrow=c(3,3))
mapply(meanSdPlot, vsnFit, main=paste("Cluster", 1:max(batch$cluster)), MoreArgs=list(ylim=c(0, 1)))
par(mfrow=c(1,1))
```

```{r combat}
batch <- batch[order(batch$cluster), ]
sig.data.norm <- sig.data.norm[, batch$sample]
sig.data.norm <- ComBat(sig.data.norm, batch=batch$cluster, 
		mod=model.matrix(~treatment, data=batch))
```

```{r combatPCA}
sig.pca <- prcomp(t(sig.data.norm), scale = TRUE)
sig.pc <- as.data.frame(sig.pca$x)
sig.pc$sample <- rownames(sig.pc)
sig.pc <- dplyr::select(as.tbl(sig.pc), sample, PC1, PC2, PC3, PC4)
sig.pc <- inner_join(sig.pc, info, by = "sample")
```

```{r combatPCAplot, cache=FALSE, fig.cap=figRef("combatPCA", "PCA plot of ComBat corrected gene expression.")}
ggplot(sig.pc, aes(x=PC1, y=PC2)) +  
     geom_point(aes( colour=as.character(Sentrix_ID)), size=3) + 
     theme_bw() + scale_colour_discrete("BeadChip")
```

```{r writeExpr, cache=FALSE}
exprOut <- gzfile("/analysis/tmp/heatshock_expr_norm.tab.gz", open="w")
write.table(sig.data.norm, file=exprOut, quote=FALSE)
close(exprOut)
```

```{r compressData, cache=FALSE, include=FALSE}
system("tar -czf tmp/yri_geno.tar.gz -C tmp hapmap_yri.smpl_flt3.bed hapmap_yri.smpl_flt3.bim hapmap_yri.smpl_flt3.fam")
```

## Differential expression analysis

All remaining samples were analysed for differences in gene expression between the basal
stimulated states, pairing samples from the same individual, using the *limma* 
Bioconductor package [@Ritchie15].

```{r limma}
labels <- strsplit(colnames(sig.data.norm), "_")
sample <- factor(sapply(labels, "[[", 1))
state <- factor(sapply(labels, "[[", 2), levels=c("Basal", "Stim"))
design <- model.matrix(~sample+state)
fit <- lmFit(sig.data.norm, design)
fit <- contrasts.fit(fit, coefficients="stateStim")
fit <- eBayes(fit)
limmaTable <- topTable(fit, n=length(fit$coefficients))
```

Individual probes were associated with corresponding genes using the updated
annotations provided by the *illuminaHumanv3.db* Bioconductor package [@Barbosa10].

```{r annotateTable}
limmaTable <- cbind(ArrayAddress=rownames(limmaTable), limmaTable)
limmaTable <- inner_join(dplyr::select(annot, ArrayAddress, NuID, SymbolReannotated, EntrezReannotated, GenomicLocation),
		limmaTable)
limmaTable <- limmaTable[order(limmaTable$P.Value),]
```

```{r writeDiffExpr, cache=FALSE}
write.table(limmaTable, file="/analysis/tmp/differential_expression.tab", row.names=FALSE, quote=FALSE, sep="\t")
```

## GO enrichment analysis

To further investigate the patterns of differential expression a GO enrichment analysis 
was carried out using the Bioconductor package *topGO* [@Alexa10]. Fisher's Exact Test
was used to determine enrichment separately for significantly (FDR < 0.01) up 
(`r tabRef("topGOUp")`) and down (`r tabRef("topGODown")`) regulated genes.

```{r preTopGO}
fdrCut <- 0.01
fcCut <- log2(1.2)
groupUp <- ifelse(limmaTable$adj.P.Val < fdrCut & limmaTable$logFC > fcCut, 1, 0)
groupDown <- ifelse(limmaTable$adj.P.Val < fdrCut & limmaTable$logFC < -fcCut, 1, 0)
groupAll <- ifelse(limmaTable$adj.P.Val < fdrCut & abs(limmaTable$logFC) > fcCut, 1, 0)

universeUp <- factor(groupUp)
names(universeUp) <- limmaTable$EntrezReannotated
universeUp <- universeUp[!is.na(names(universeUp))]
universeDown <- factor(groupDown)
names(universeDown) <- limmaTable$EntrezReannotated
universeDown <- universeDown[!is.na(names(universeDown))]
universeAll <- factor(groupAll)
names(universeAll) <- limmaTable$EntrezReannotated
universeAll <- universeAll[!is.na(names(universeAll))]

tgUp <- new("topGOdata", ontology="BP", allGenes=universeUp, nodeSize=10,
			description="Genes up-regulated after heat shock",
			annotationFun=annFUN.org, mapping="org.Hs.eg.db")
tgDown <- new("topGOdata", ontology="BP", allGenes=universeDown, nodeSize=10,
			description="Genes down-regulated after heat shock",
			annotationFun=annFUN.org, mapping="org.Hs.eg.db")
tgAny <- new("topGOdata", ontology="BP", allGenes=universeAll, nodeSize=10,
			description="Genes differentially expressed in response to heat shock",
			annotationFun=annFUN.org, mapping="org.Hs.eg.db")
```

```{r testTopGO}
resultUp <- runTest(tgUp, algorithm="classic", statistic="fisher")
resultDown <- runTest(tgDown, algorithm="classic", statistic="fisher")
goTabUp <- GenTable(tgUp, up = resultUp, down = resultDown, 
                     topNodes = length(score(resultUp)), orderBy = "up")
goTabDown <- GenTable(tgDown, up = resultUp, down = resultDown, 
                     topNodes = length(score(resultDown)), orderBy = "down")
```

```{r writeTopGO, include=FALSE, cache=FALSE}
write.table(goTabUp,
            file="tmp/topGO_upregulated.tab", sep="\t", quote=FALSE, row.names=FALSE)
write.table(goTabDown,
            file="tmp/topGO_downregulated.tab", sep="\t", quote=FALSE, row.names=FALSE)
system("tar -czf tmp/topGO.tar.gz -C tmp topGO_upregulated.tab topGO_downregulated.tab")
```

## Gene annotions
In addition to identifying common themes amongst differentially expressed genes the question
arises to what extend these genes have been previously associated with the heat
shock or, more generally, stress response. We use the set of genes previously
linked directly to heat shock [@Richter10] and from this create an extended set based 
on GO terms and PubMed articles linking differentially expressed genes to heat shock
response and closely related processes.

```{r loadKnown}
knownList <- list()
knownList$known <- readLines("data/heatshock_genes.lst")
```

### Gene Ontology terms
As a first step in highlighting genes 
not previously known to play a role in this context we identify all significantly
up regulated genes that lack GO annotations of obvious relevance to heat shock response.
In addition to terms related to stress response and protein folding we also include
terms related to cell death and proliferation. To account for the presence of EBV in 
these cells we ignore all genes annotated with terms related to viral infections.
Finally, any remaining genes related to regulation of gene expression are considered
to be explained by the large scale changes in gene expression that are taking place
in response to heat shock.

```{r goTerms}
sig.entrez <- dplyr::filter(limmaTable, !SymbolReannotated %in% knownList$known, 
                            adj.P.Val < fdrCut, logFC > fcCut)$EntrezReannotated
mart <- useMart("ensembl",dataset="hsapiens_gene_ensembl")
gene.data <- getBM(
  attributes=c('hgnc_symbol', 'entrezgene', 'go_id', 'name_1006', 'namespace_1003'), 
  filters = 'entrezgene', values = sig.entrez, mart = mart)
gene.data <- dplyr::filter(gene.data, namespace_1003=="biological_process")
termFilter <- c("heat", "stress", "folding", "folded", 
                "death", "apoptosis", "apoptotic", 
                "cell cycle", "mitosis", "mitotic", "proliferation", 
                "virus", "viral", "immune",
                "transcription", "translation", "gene expression")
extendedGenes <- dplyr::filter(gene.data, str_detect(name_1006, paste(termFilter, collapse="|")))
knownList$extended <- extendedGenes$hgnc_symbol
selGenes <- dplyr::filter(limmaTable, adj.P.Val < fdrCut, logFC > fcCut, 
                          !SymbolReannotated %in% knownList$extended,
                          !SymbolReannotated %in% knownList$known)
```

### PubMed
All genes not annotated with obvious GO terms were subjected to a PubMed search
to find publications that link the gene to heat shock or stress response.

```{r pubmed}
selGenes <- dplyr::filter(selGenes, SymbolReannotated != "")
queryString <- paste(unique(selGenes$SymbolReannotated), "AND (heat shock OR stress)")

query <- vector(mode="list", length(queryString))
for(i in seq(1, length(query), by=10)){
  end <- min(i + 9, length(query))
  query[i:end] <- tryCatch(lapply(queryString[i:end], EUtilsSummary),
                           error=function(e) {
                             wait(2)
                             lapply(queryString[i:end], EUtilsSummary)
                          })
}
names(query) <- unique(selGenes$SymbolReannotated)
queryEmpty <- query[sapply(query, QueryCount) == 0]
queryFound <- query[sapply(query, QueryCount) != 0]

pmids <- lapply(queryFound, function(x) PMID(EUtilsGet(x)))
knownList$extended <- union(knownList$extended, names(queryFound))
knownList$novel <- names(queryEmpty)

putativeGenes <- dplyr::filter(selGenes, SymbolReannotated %in% names(queryEmpty))
putativeGenes <- dplyr::left_join(putativeGenes, dplyr::select(gene.data, 
                                                               hgnc_symbol, go_id, name_1006), 
                                  by=c(SymbolReannotated="hgnc_symbol"))
```

```{r geneRecords, cache=FALSE}
write.table(data.frame(gene=names(queryFound), pmid=sapply(pmids, paste, collapse=",")),
            file="tmp/heatshock_pmid.txt", row.names=FALSE, quote=FALSE, sep="\t")
write.table(putativeGenes, file="tmp/heatshock_novel.tab", sep="\t", 
            quote=FALSE, row.names=FALSE)
```

### Heat shock factor binding
To further establish links between differentially expressed genes and heat shock response
we consider the evidence for binding of HSF1 and HSF2 in the promoter regions of these genes.
Using binding sites derived fro ChIP-seq data obtained from the K562 cell line
[@Vihervaara13] we annotated our list of differentially expressed genes by
cross-referencing it with the list HSF binding genes. Groups of genes corresponding
to up or down regulated genes as well as those with existing heat shock related annotations
and those without were tested for enrichment of HSF binding genes using Fisher's
exact test.

```{r loadPeaks}
peaks <- list.files(path="/analysis/data", pattern="HSF.*\\.txt", full.name=TRUE)

peaks2genes <- data.frame()
for(i in peaks){
        nm <- gsub("/analysis/data/(.*).txt", "\\1", i)
        f <- read.table(i, sep="\t", header=T)
        genes <-  paste(f[,11], collapse=",")
        genes <- unique(strsplit(genes, ",|-")[[1]])
        genes[genes==""] <- NA

        genes <- genes[!is.na(genes)]
        peaks2genes <- rbind(peaks2genes, data.frame(cond=nm, genes=genes) )
}
peaks2genes$cond2 <- gsub("^M|^U", "comb", peaks2genes$cond)
peaks2genes$treatment <- str_extract(peaks2genes$cond2, "C|HS")
peaks2genes$tf <- str_extract(peaks2genes$cond2, "HSF.")

peakSplit <- by(peaks2genes, list(peaks2genes$treatment, peaks2genes$tf), function(x) x)
peakSplit[[1]] <- subset(peakSplit[[1]], genes %in% peakSplit[[3]]$genes)
if(nrow(peakSplit[[1]])) {
  peakSplit[[1]]$tf <- "HSF1+HSF2"
  peakSplit[[1]]$cond2 <- paste0(peakSplit[[1]]$cond2, "+HSF2")
}
peakSplit[[2]] <- subset(peakSplit[[2]], genes %in% peakSplit[[4]]$genes)
if(nrow(peakSplit[[2]])){
  peakSplit[[2]]$tf <- "HSF1+HSF2"
  peakSplit[[2]]$cond2 <- paste0(peakSplit[[2]]$cond2, "+HSF2")
} 

peaks2genes <- rbind(peaks2genes, peakSplit[[1]], peakSplit[[2]])
```

```{r peakGroups}
groupUp <- unique(subset(limmaTable, limmaTable$adj.P.Val < fdrCut & limmaTable$logFC > fcCut)$SymbolReannotated)
groupDown <- unique(subset(limmaTable, limmaTable$adj.P.Val < fdrCut & limmaTable$logFC < -fcCut)$SymbolReannotated)

totalNumberOfGenes <- length(unique(limmaTable$SymbolReannotated))
NofgroupUp <- length(groupUp)
NofgroupDown <- length(groupDown)

genes_in_both <- data.frame()
hsfPval <- data.frame()
for(g in unique(peaks2genes$cond2) ){
        anno_genes <- unique( subset(peaks2genes, cond2==g)$genes )
        groupUp_in_anno_genes <- length( anno_genes[anno_genes %in% groupUp] )
        groupUp_not_in_anno_genes <- length( anno_genes[!(anno_genes %in% groupUp)] )

        groupDown_in_anno_genes <- length( anno_genes[anno_genes %in% groupDown] )
        groupDown_not_in_anno_genes <- length( anno_genes[!(anno_genes %in% groupDown)] )

        if ( length(anno_genes[anno_genes %in% groupUp] ) > 0) {
                genes_in_both <- rbind(genes_in_both, 
                                       data.frame(cond=g, exp="up", 
                                                  genes=anno_genes[anno_genes %in% groupUp] ) )
        }

        if ( length(anno_genes[anno_genes %in% groupDown] ) > 0) {
                genes_in_both <- rbind(genes_in_both, 
                                       data.frame(cond=g, exp="down", 
                                                  genes=anno_genes[anno_genes %in% groupDown] ) )
        }
        dat_up <- matrix(c(NofgroupUp, (totalNumberOfGenes - NofgroupUp), 
                           groupUp_in_anno_genes, groupUp_not_in_anno_genes), ncol=2 )
        rst_up <- fisher.test(dat_up, alternative="less")

        dat_down <- matrix(c(NofgroupDown, (totalNumberOfGenes - NofgroupDown), 
                             groupDown_in_anno_genes, groupDown_not_in_anno_genes), ncol=2 )
        rst_down <- fisher.test(dat_down, alternative="less")


        hsfPval <- rbind(hsfPval, data.frame(cond=g, up=rst_up$p.value, down=rst_down$p.value, 
                                 sigGenes_up=groupUp_in_anno_genes, sigGenes_down=groupDown_in_anno_genes) )
}


genes_in_both$heatshock_novel_genes <- ifelse( genes_in_both$genes %in% knownList$novel, 2, 
                                               ifelse(genes_in_both$genes %in% knownList$extended, 1, 0))
```

In addition to the direct evidence from the ChIP-seq data we carried out a
scan for the presence of HSF binding motifs in the promoter region 
(1200bp upstream - 300bp downstream of the TSS) of differentially expressed genes.
The scan was based on the PWM defined by SwissRegulon [@Pachkov07] and carried 
out with the Bioconductor package PWMEnrich [@PWMEnrich].

```{r prepPWMEnrich}
txdb <- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene, columns=c("gene_id", "tx_id", "tx_name"))
promoter.seqs <- getPromoterSeq(txdb, Hsapiens, upstream=1200, downstream=300)
names(promoter.seqs) <-  as.character(txdb$gene_id)
promoter.seqs <- unique(promoter.seqs)
promoter.seqs_noNA <- promoter.seqs[!is.na(names(promoter.seqs) )]

group_up <- dplyr::filter(limmaTable, adj.P.Val < fdrCut, logFC > fcCut)
group_dn <- dplyr::filter(limmaTable, adj.P.Val < fdrCut, logFC < -fcCut)
group_others <- dplyr::filter(limmaTable, !SymbolReannotated %in% group_up$SymbolReannotated,
                              !SymbolReannotated %in% group_dn$SymbolReannotated)

group_up_noNa <- subset(group_up, !is.na(EntrezReannotated))
group_dn_noNa <- subset(group_dn, !is.na(EntrezReannotated))
group_others_noNa <- subset(group_others, !is.na(EntrezReannotated))


promoter.seqs.group_up <- unique(promoter.seqs_noNA[names(promoter.seqs_noNA) %in% group_up_noNa$EntrezReannotated])
promoter.seqs.group_dn <- unique(promoter.seqs_noNA[names(promoter.seqs_noNA) %in% group_dn_noNa$EntrezReannotated])
promoter.seqs.group_others <- unique(promoter.seqs_noNA[names(promoter.seqs_noNA) %in% group_others_noNa$EntrezReannotated])
```

```{r pwmEnrich}
registerCoresPWMEnrich(4)
motif.HSF1 <- readMotifs("data/HSF1_pwm.jaspar")
genomic.acgt = getBackgroundFrequencies("hg19")
pwms.denovo = toPWM(motif.HSF1, prior=genomic.acgt)
bg.denovo = makeBackground(pwms.denovo, organism="hg19", type="logn", quick=TRUE)

r_promoter.seqs.group_up <- promoterScan(promoter.seqs.group_up, bg.denovo, "up")
r_promoter.seqs.group_dn <- promoterScan(promoter.seqs.group_dn, bg.denovo, "down")
r_promoter.seqs.group_others <- promoterScan(promoter.seqs.group_others, bg.denovo, "others")

r_promoter <- rbind(r_promoter.seqs.group_up, r_promoter.seqs.group_dn, r_promoter.seqs.group_others)
r_promoter <- dplyr::left_join(r_promoter, limmaTable, by="EntrezReannotated")

```

```{r pwmAnnotate}
r_promoter$HSFpeak <- ifelse( r_promoter$SymbolReannotated %in% peaks2genes$genes, 1, 0)
r_promoter$novel <- ifelse(r_promoter$SymbolReannotated %in% knownList$novel, 2, ifelse(r_promoter$SymbolReannotated %in% knownList$extended, 1, 0))

trx_numbers <- data.frame(table(r_promoter$SymbolReannotated))
colnames(trx_numbers) <- c("SymbolReannotated", "trx_numbers")
totalGeneNumber <- nrow(trx_numbers)

genePval <- do.call(rbind, by(r_promoter, r_promoter$SymbolReannotated, function(x){
  x[which.min(x$p.value),,drop=FALSE]
}))
genePval$pwm.FDR <- p.adjust(genePval$p.value, "BH")
r_promoter <- dplyr::left_join(r_promoter, dplyr::select(genePval, EntrezReannotated, pwm.FDR), by="EntrezReannotated")
r_promoter <- dplyr::select(r_promoter, -rank, -target, -id)
r_promoter <- dplyr::select(r_promoter, ArrayAddress, NuID, 
                            EntrezReannotated, SymbolReannotated, GenomicLocation:B,
                            expr, novel, HSFpeak, raw.score, pwm.p.value=p.value, pwm.FDR)
r_gene <- by(r_promoter, r_promoter$EntrezReannotated, function(x){
  ans <- x[1,,drop=FALSE]
  if(nrow(x) > 1){
    exprCols <- c("ArrayAddress", "NuID", "GenomicLocation", "logFC", "AveExpr", "t", "P.Value", "adj.P.Val", "B", "expr")
    pwmCols <- c("raw.score", "pwm.p.value", "pwm.FDR")
    expr <- which.min(x$adj.P.Val)
    ans[1, exprCols] <- x[expr, exprCols]
    pwm <- which.min(x[,"pwm.p.value"])
    ans[1, pwmCols] <- x[pwm, pwmCols]
  }
  ans
})
r_gene <- do.call(rbind, r_gene)
```


```{r peakOutput, cache=FALSE}
write.table(hsfPval, "tmp/hsf_genes_pvalue.txt", sep="\t", quote=FALSE, row.names=FALSE)
write.table(genes_in_both, "tmp/hsf_genes.txt" , sep="\t", quote=FALSE, row.names=FALSE)
write.table(r_promoter, "tmp/hsf_motif_genes.txt", sep="\t", quote=FALSE, row.names=FALSE)
system("tar -czf tmp/gene_annotation.tar.gz -C tmp heatshock_pmid.txt heatshock_novel.tab hsf_genes.txt hsf_genes_pvalue.txt hsf_motif_genes.txt")
```

## Global heat shock response

To assess the global difference in gene expression induced by heat shock
we carried out partial least squares regression (PLS), using the treatment state
(basal or stimulated) as a binary response variable and all gene expression
probes that passed QC as explanatory variables.

```{r pls}
names <- strsplit(colnames(sig.data.norm), "_")
treat <- factor(sapply(names, "[[", 2), levels=c("Basal", "Stim"))
plsFit <- plsreg1(t(sig.data.norm), data.frame(treatment=as.integer(treat)-1))
plsComp <- as.data.frame(plsFit$x.scores)
plsComp$sample <- sapply(names, "[[", 1)
plsComp$treatment <- treat
```

To explore to what extend this response is modulated by genetic variation we used
the length and direction of the response vector, i.e. the vector between the basal
and stimulated sample for each individual in the space spanned by the first two PLS 
components, together with the location of the basal sample in the same space as a 
multivariate phenotype. This was then tested for association with
genotypes for SNPs close to differentially expressed genes using the *MultiPhen* R
package [@OReilly12]. The GRCh37 coordinates 
for SNPs were obtained via the *SNPlocs.Hsapiens.dbSNP142.GRCh37*
Bioconductor package [@Pages14] and gene coordinates via the 
*TxDb.Hsapiens.UCSC.hg19.knownGene* package [@Carlson15].


```{r snpPosUpdate}
geno <- read_plink("tmp/hapmap_yri.smpl_flt3")
geno$bim <- dplyr::select(geno$bim, 2, 1, 4)
names(geno$bim) <- c("snp", "chrom", "pos")

geno$bim <- by(geno$bim, geno$bim$chrom, function(x){
  locs <- getSNPlocs(paste0("ch", x$chrom[1]), caching=FALSE)
  locs$RefSNP_id <- paste0("rs", locs$RefSNP_id)
  snps <- inner_join(x, locs, by=c(snp="RefSNP_id"))
  if(nrow(snps) < nrow(x)) warning("Dropped ", nrow(x) - nrow(snps), " SNP(s) from chromosome ", x$chrom[1])
  dplyr::select(snps, snp, chrom, loc)
})

geno$bim <- do.call(rbind, geno$bim)
geno$bim$chrom <- paste0("chr", geno$bim$chrom)
geno$bim <- geno$bim[order(geno$bim$chrom, geno$bim$loc),]
```

```{r cisSNPs}
cisWindow <- 1e4

exclChrom <- c("chrX", "chrY", "chrM")
selectedGenes <- dplyr::filter(limmaTable, adj.P.Val < fdrCut & abs(logFC) > fcCut &
                                 !str_detect(GenomicLocation, paste(exclChrom, collapse="|")))
selectedGenes <- dplyr::select(selectedGenes, ArrayAddress, NuID, GenomicLocation, EntrezReannotated, SymbolReannotated)

selectedGenes$GenomicLocation <- str_extract(selectedGenes$GenomicLocation, 
                                             "chr[[:digit:]]{1,2}:[[:digit:]]+:[[:digit:]]+")
selectedGenes <- dplyr::filter(selectedGenes, !is.na(GenomicLocation) & !is.na(EntrezReannotated))

txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
txTable <- select(txdb, keys=selectedGenes$EntrezReannotated, 
                  columns=c("TXID", "TXCHROM", "TXSTART", "TXEND"),
                  keytype="GENEID")
txTable <- dplyr::filter(txTable, !is.na(TXCHROM) & !str_detect(TXCHROM, "_"))
txTable <- by(txTable, txTable$GENEID, function(x) data.frame(entrez=x[["GENEID"]][1], 
                                                              chrom=x[["TXCHROM"]][1], 
                                                              start=min(x[["TXSTART"]]),
                                                              end=max(x[["TXEND"]])))
txTable <- do.call(rbind, txTable)
selectedGenes <- dplyr::right_join(selectedGenes, txTable, by=c(EntrezReannotated="entrez"))

probePos <- dplyr::select(selectedGenes, NuID, chrom, start, end)
probePos <- dplyr::arrange(probePos, chrom, start)

probePosGR <- makeGRangesFromDataFrame(probePos, ignore.strand = TRUE, keep.extra.columns = TRUE)
snpPosGR <- makeGRangesFromDataFrame(geno$bim, ignore.strand = TRUE, keep.extra.columns = TRUE, 
                                   start.field = "loc", end.field = "loc")
geneSNPs <- do.call(GRangesList, mapply(
  function(x, chr, s, e) 
    subset(x, seqnames == chr & start >= s - cisWindow & end <= e + cisWindow), 
  as.character(seqnames(probePosGR)), start(probePosGR), end(probePosGR), 
  MoreArgs=list(x=snpPosGR), SIMPLIFY=FALSE))
names(geneSNPs) <- probePos$NuID
noSNP <- probePos$NuID[sapply(geneSNPs, function(x) length(x) == 0)]
snpRange <- unique(unlist(lapply(geneSNPs, function(x) x$snp)))
```

```{r multiPheno}
stimComp <- dplyr::filter(plsComp, treatment=="Stim")
basalComp <- dplyr::filter(plsComp, treatment=="Basal")
globalPheno <- data.frame(sample=stimComp$sample,
                          basal.x=basalComp$t1,
                          basal.y=basalComp$t2,
                          distance=sqrt((stimComp$t1 - basalComp$t1)^2 + (stimComp$t2 - basalComp$t2)^2),
                          direction=atan((stimComp$t2 - basalComp$t2)/(stimComp$t1 - basalComp$t1)))
globalPheno <- dplyr::left_join(globalPheno, dplyr::select(geno$fam, V2, V5), by=c(sample="V2"))
globalPheno <- dplyr::rename(globalPheno, sex=V5)

globalPheno <- cbind(scale(dplyr::select(globalPheno, -sex, -sample)), sex=globalPheno$sex-1)
rownames(globalPheno) <- stimComp$sample

rownames(geno$bed) <- str_extract(rownames(geno$bed), "NA[[:digit:]]+$")
cisGeno <- geno$bed[rownames(globalPheno), snpRange]
assoc <- mPhen(cisGeno, globalPheno, covariates="sex")
idx <- order(assoc$Results[1, , "JointModel" , 2])
assoc$Results <- assoc$Results[,idx, , ]
assoc$maf <- assoc$maf[idx]
```

Significance of the observed associations was assessed through a permutation test to account for the
structure inherent to the data.

```{r multiPhenoPerm, results='hide'}
set.seed(54864548L)
nPerm <- 1000
bins <- sapply(apply(cisGeno, 2, table), paste, collapse="_")
names(bins) <- colnames(cisGeno)
pPerm <- matrix(NA, nrow=nPerm, ncol=length(unique(bins)))
colnames(pPerm) <- unique(bins)
binSNPs <- tapply(colnames(cisGeno), bins, function(x) x)
permGeno <- cisGeno[, sapply(binSNPs, "[", 1)]

for(i in 1:nPerm){
  permPheno <- dplyr::sample_n(as.data.frame(globalPheno), nrow(globalPheno), replace=FALSE)
  rownames(permPheno) <- rownames(globalPheno)
  assocPerm <- tryCatch( mPhen(permGeno, as.matrix(permPheno), covariates="sex"),
            error= function(e){
              permPheno <- dplyr::sample_n(as.data.frame(globalPheno), nrow(globalPheno), replace=FALSE)
              mPhen(permGeno, as.matrix(permPheno), covariates="sex")
            })
  pPerm[i, ] <- assocPerm$Results[1, ,"JointModel", 2]
}
```

To this end the observed global response phenotype for each individual and covariates used in the model
were randomly assigned to one of the observed set of genotypes `r nPerm` times and p-values for 
the joint model were computed for each permutation. From these we computed false discovery rates
by comparing observed p-values to the empirical distribution of minimum p-values from each permutation.

```{r permFDR}
minp <- apply(pPerm, 1, min)
assoc$adj.p.val <- sapply(assoc$Results[, "JointModel", 2], function(x) sum(minp < x)/length(minp))
names(assoc$adj.p.val) <- dimnames(assoc$Results)[[1]]
assoc$perm.p.val <- numeric(length(assoc$maf))
names(assoc$perm.p.val) <- dimnames(assoc$Results)[[1]]
for(i in 1:ncol(pPerm)){
  assoc$perm.p.val[binSNPs[[i]]] <- sapply(assoc$Results[binSNPs[[i]], "JointModel", 2],
                                          function(x, p) sum(p < x)/length(p), pPerm[, i])
}
```

## Response QTL analysis
Having established the association between rs7789228 and the global response to 
heat shock we sought to further investigate the influence of this SNP on the
expression of individual genes. To this end we considered all differentially 
expressed genes (FDR < `r fdrCut` and fold change > `r 2^fcCut`).

This results in a list of `r nrow(selectedGenes)` probes (targeting 
`r length(unique(selectedGenes$SymbolReannotated))` different genes).

For each probe the observed response to the stimulus (as measured by the log~2~ 
fold change) is tested for association with the genotype of rs7789228.

```{r probeFC}
selectedExpr <- sig.data.norm[selectedGenes$ArrayAddress,]
basalIdx <- grep("_Basal", colnames(selectedExpr))
stimIdx <- grep("_Stim", colnames(selectedExpr))
selectedFC <- selectedExpr[, stimIdx] - selectedExpr[, basalIdx]
colnames(selectedFC) <- str_extract(colnames(selectedFC), "NA[[:digit:]]+")
rownames(selectedFC) <- selectedGenes$NuID
```

```{r writeFcFile, cache=FALSE}
write.table(selectedFC, file="tmp/selected_probes_fc.tab", sep="\t", quote=FALSE)
```

```{r writeProbePos, cache=FALSE}
write.table(dplyr::filter(probePos, NuID %in% selectedGenes$NuID), 
            file="tmp/selected_probes_pos.tab", sep="\t", quote=FALSE, row.names=FALSE)
```

```{r writeSNPpos, cache=FALSE, results="hide"}
write.table(dplyr::filter(geno$bim, snp=="rs7789228"), file="tmp/snp_loc_GRCh37.tab", quote=FALSE, sep="\t")
```

```{r prepGeno}
rownames(geno$bed) <- str_extract(rownames(geno$bed), "NA[[:digit:]]+")
geno$bed <- geno$bed[colnames(selectedFC), geno$bim$snp]
geno$bed <- t(geno$bed)
flip <- rowSums(geno$bed, na.rm=TRUE) > ncol(geno$bed)
if(any(flip)){
  geno$bed[flip, ] <- -geno$bed[flip]+2
}
```

```{r writeGeno, cache=FALSE, results="hide"}
write.table(geno$bed["rs7789228",,drop=FALSE], file="tmp/genotypes_GRCh37.tab", quote=FALSE, sep="\t")
```

```{r fc_pc}
nPC <- 0
if(nPC > 0){
  pca <- prcomp(t(selectedFC), center=TRUE, scale = TRUE)
  pc <- pca$x
}
```

We tested for associations between genotype and heat shock response using
a linear model as implemented in Matrix-eQTL [@Shabalin12], correcting for
sex as well as the first `r nPC` principle components of the treatment response. 

```{r covariates}
geno$fam <- dplyr::filter(dplyr::select(geno$fam, 2, 5), V2 %in% colnames(geno$bed))
geno$fam <- t(geno$fam)
colnames(geno$fam) <- geno$fam[1,]
geno$fam <- geno$fam[, colnames(selectedFC)]
if(nPC > 0) geno$fam <- rbind(geno$fam, t(pc[, 1:nPC]))
```

```{r writeCovar, cache=FALSE, echo=FALSE}
write.table(geno$fam[-1, , drop=FALSE], file="tmp/covariates.tab", quote=FALSE, sep="\t")
```

```{r prepQTL, results="hide", cache=FALSE}
snps <- SlicedData$new();
snps$fileDelimiter <- "\t"
snps$fileOmitCharacters <- "NA"
snps$fileSkipRows <- 1
snps$fileSkipColumns <- 1
snps$fileSliceSize <- 2000
snps$LoadFile("tmp/genotypes_GRCh37.tab")

pheno <- SlicedData$new();
pheno$fileDelimiter <- "\t"
pheno$fileOmitCharacters <- "NA"
pheno$fileSkipRows <- 1
pheno$fileSkipColumns <- 1
pheno$fileSliceSize <- 2000
pheno$LoadFile("tmp/selected_probes_fc.tab")

covar <- SlicedData$new();
covar$fileDelimiter <- "\t"
covar$fileOmitCharacters <- "NA"
covar$fileSkipRows <- 1
covar$fileSkipColumns <- 1
covar$fileSliceSize <- 2000
covar$LoadFile("tmp/covariates.tab")
```

```{r runMatrixEQTL, results="hide", cache=FALSE}
me <- Matrix_eQTL_main(
    snps = snps,
    gene = pheno,
    cvrt = covar,
    output_file_name.cis = "tmp/reQTL_analysis",
    pvOutputThreshold = 0,
    pvOutputThreshold.cis = 1e-3,
    useModel = modelLINEAR, 
    errorCovariance = numeric(),
    snpspos = geno$bim,
    genepos = probePos,
    cisDist = cisWindow,
    verbose = FALSE,
    pvalue.hist = FALSE,
    min.pv.by.genesnp = TRUE,
    noFDRsaveMemory = FALSE)
```

```{r geneSNPs}
mainSNP <- dplyr::filter(me$cis$eqtls, !duplicated(gene))

geneP <- me$cis$min.pv.gene * sapply(names(me$cis$min.pv.gene), 
                                     function(x) length(geneSNPs[[x]]))
geneP <- sapply(geneP, min, 1)
geneP <- p.adjust(geneP, "BH")
geneP <- data.frame(NuID=names(geneP), Gene.FDR=geneP, stringsAsFactors=FALSE)
geneP <- dplyr::left_join(geneP, limmaTable, by="NuID")
geneP <- dplyr::left_join(geneP, mainSNP, by=c(NuID="gene"))
geneP <- dplyr::filter(geneP, !NuID %in% noSNP)
geneP <- dplyr::select(geneP, SymbolReannotated, snps, NuID, ArrayAddress, Gene.FDR, 
                       pvalue, beta, logFC, adj.P.Val)
geneP <- dplyr::rename(geneP, Symbol=SymbolReannotated, SNP=snps, SNP.pvalue=pvalue, 
                       expression.FDR=adj.P.Val, SNP.beta=beta)
geneP <- dplyr::arrange(geneP, Gene.FDR, desc(abs(SNP.beta)))
```

```{r prepGeneMatrixEqtl, cache=FALSE}
selGenes <- dplyr::filter(geneP, Gene.FDR < 1)
emptySel <- character()
if(nrow(selGenes)){
  for(i in 1:nrow(selGenes)){
    selSNPs <- geno$bed[geneSNPs[[selGenes$NuID[i]]]$snp, ]
    if(nrow(selSNPs)){
      write.table(selSNPs, file=paste0("tmp/genotypes_", selGenes$Symbol[i],".tab"), 
                  quote=FALSE, sep="\t")
      write.table(selectedFC[selGenes$NuID[i], , drop=FALSE], 
                  file=paste0("tmp/", selGenes$Symbol[i], "_fc.tab"), sep="\t", quote=FALSE)
      } else {
        emptySel <- c(emptySel, selGenes$NuID[i])
      }
  }
  selGenes <- dplyr::filter(selGenes, !NuID %in% emptySel)
}
```

```{r runGeneMatrixEqtl, cache=FALSE, results="hide"}
meGene <- vector(mode="list", nrow(selGenes))
names(meGene) <- selGenes$NuID
if(nrow(selGenes) > 0){
  for(i in 1:nrow(selGenes)){
    snpsSub <- SlicedData$new();
    snpsSub$fileDelimiter <- "\t"
    snpsSub$fileOmitCharacters <- "NA"
    snpsSub$fileSkipRows <- 1
    snpsSub$fileSkipColumns <- 1
    snpsSub$fileSliceSize <- 2000
    snpsSub$LoadFile(paste0("tmp/genotypes_", selGenes$Symbol[i],".tab"))
    
    pheno <- SlicedData$new();
    pheno$fileDelimiter <- "\t"
    pheno$fileOmitCharacters <- "NA"
    pheno$fileSkipRows <- 1
    pheno$fileSkipColumns <- 1
    pheno$fileSliceSize <- 2000
    pheno$LoadFile(paste0("tmp/", selGenes$Symbol[i], "_fc.tab"))
    
    meGene[[selGenes$NuID[i]]] <- Matrix_eQTL_main(
      snps = snpsSub,
      gene = pheno,
      cvrt = covar,
      output_file_name.cis = paste0("tmp/reQTL_", selGenes$Symbol[i],"analysis"),
      pvOutputThreshold = 0,
      pvOutputThreshold.cis = 1,
      useModel = modelLINEAR, 
      errorCovariance = numeric(),
      snpspos = geno$bim,
      genepos = probePos,
      cisDist = cisWindow,
      verbose = FALSE,
      pvalue.hist = FALSE,
      min.pv.by.genesnp = TRUE,
      noFDRsaveMemory = FALSE)
  }
}
geneSNPcount <- sapply(meGene, function(x) sum(x$cis$eqtls$FDR < 0.01))

```

```{r bundleReQTLinput, cache=FALSE, include=FALSE}
system("tar -czf tmp/reQTL_input.tar.gz -C tmp selected_probes_pos.tab snp_loc_GRCh37.tab selected_probes_fc.tab genotypes_GRCh37.tab covariates.tab")
```

## Analysis of eQTL in basal and stimulated cells

To further investigate potential genetic modulators of gene expression for genes that
play a role in heat shock response we carried out a cis eQTL analysis for all genes identified
as differentially expressed in response to heat shock.

```{r selectedExpr}
stimExpr <- selectedExpr[, stimIdx]
basalExpr <- selectedExpr[, basalIdx]
colnames(stimExpr) <- str_extract(colnames(stimExpr), "NA[[:digit:]]+")
colnames(basalExpr) <- str_extract(colnames(basalExpr), "NA[[:digit:]]+")
rownames(stimExpr) <- rownames(basalExpr) <- selectedGenes$NuID
```

```{r writeSelExpr, cache=FALSE}
write.table(stimExpr, file="tmp/selected_probes_stim_expr.tab", sep="\t", quote=FALSE)
write.table(basalExpr, file="tmp/selected_probes_basal_expr.tab", sep="\t", quote=FALSE)
```

```{r expr_pc}
nPC_expr <- 1
pca_stim <- prcomp(t(stimExpr), center=TRUE, scale = TRUE)
pc_stim <- pca_stim$x

pca_basal <- prcomp(t(basalExpr), center=TRUE, scale = TRUE)
pc_basal <- pca_basal$x
```

```{r writeExprCovar, cache=FALSE}
write.table(t(pc_stim)[1:nPC_expr, , drop=FALSE], file = "tmp/covar_expr_stim.tab", sep="\t", quote=FALSE)
write.table(t(pc_basal)[1:nPC_expr, , drop=FALSE], file = "tmp/covar_expr_basal.tab", sep="\t", quote=FALSE)
```


The analysis was carried out with Matrix-eQTL using an approach equivalent to the one
used for the [response QTL analysis](#response-qtl-analysis).

```{r prepEQTL, results="hide", cache=FALSE}
phenoStim <- SlicedData$new();
phenoStim$fileDelimiter <- "\t"
phenoStim$fileOmitCharacters <- "NA"
phenoStim$fileSkipRows <- 1
phenoStim$fileSkipColumns <- 1
phenoStim$fileSliceSize <- 2000
phenoStim$LoadFile("tmp/selected_probes_stim_expr.tab")

covarStim <- SlicedData$new();
covarStim$fileDelimiter <- "\t"
covarStim$fileOmitCharacters <- "NA"
covarStim$fileSkipRows <- 1
covarStim$fileSkipColumns <- 1
covarStim$fileSliceSize <- 2000
covarStim$LoadFile("tmp/covar_expr_stim.tab")

phenoBasal <- SlicedData$new();
phenoBasal$fileDelimiter <- "\t"
phenoBasal$fileOmitCharacters <- "NA"
phenoBasal$fileSkipRows <- 1
phenoBasal$fileSkipColumns <- 1
phenoBasal$fileSliceSize <- 2000
phenoBasal$LoadFile("tmp/selected_probes_basal_expr.tab")

covarBasal <- SlicedData$new();
covarBasal$fileDelimiter <- "\t"
covarBasal$fileOmitCharacters <- "NA"
covarBasal$fileSkipRows <- 1
covarBasal$fileSkipColumns <- 1
covarBasal$fileSliceSize <- 2000
covarBasal$LoadFile("tmp/covar_expr_basal.tab")
```

```{r runEQTL, cache=FALSE, results="hide"}
meStim <- Matrix_eQTL_main(
    snps = snps,
    gene = phenoStim,
    cvrt = covarStim,
    output_file_name = "tmp/eQTL_stim_analysis",
    pvOutputThreshold = 1,
    pvOutputThreshold.cis = 0,
    useModel = modelLINEAR, 
    errorCovariance = numeric(),
    verbose = FALSE,
    pvalue.hist = FALSE,
    min.pv.by.genesnp = FALSE,
    noFDRsaveMemory = FALSE)

meBasal <- Matrix_eQTL_main(
    snps = snps,
    gene = phenoBasal,
    cvrt = covarBasal,
    output_file_name = "tmp/eQTL_basal_analysis",
    pvOutputThreshold = 1,
    pvOutputThreshold.cis = 0,
    useModel = modelLINEAR, 
    errorCovariance = numeric(),
    verbose = FALSE,
    pvalue.hist = FALSE,
    min.pv.by.genesnp = FALSE,
    noFDRsaveMemory = FALSE)

stimEQTL <- formatEQTL(meStim, limmaTable)
basalEQTL <- formatEQTL(meBasal, limmaTable)
```

The genes with significant associations (FDR < 0.05) in either condition are considered
for significance in the other condition. Any gene with a significant association in 
one condition and FDR > 0.5 in the other is considered to be specific to that condition.

```{r compareGenes}
basal_extended <- dplyr::filter(basalEQTL, FDR < 0.5)$Gene
stim_extended <- dplyr::filter(stimEQTL, FDR < 0.5)$Gene

stimGenes <- dplyr::filter(stimEQTL, FDR < 0.05 & !Gene %in% basal_extended)
basalGenes <- dplyr::filter(basalEQTL, FDR < 0.05 & !Gene %in% stim_extended)
```

```{r bundleEQTLinput, cache=FALSE, include=FALSE}
system("tar -czf tmp/eQTL_input.tar.gz -C tmp selected_probes_pos.tab snp_loc_GRCh37.tab selected_probes_stim_expr.tab selected_probes_basal_expr.tab covar_expr_stim.tab covar_expr_basal.tab")
```

## Trans associations for SNPs upstream of HSF1 and HSF2

Strong associations between SNPs in the promoter region of a HSF1 homologue and the
expression of heat shock proteins are known to exist in rats [@Dumas00]. We therefore
thought to investigate whether similar effects can be observed for the differentially
expressed genes we identified. To this end we tested all genotyped SNPs that passed QC
within a window including the gene body and 1MB upstream of the transcription start site
of HSF1 and HSF2 for associations with the heat shock response (log fold change) as well
as with gene expression in basal and stimulated cells.

```{r hsfWindow}
genePos <- data.frame(chrom=c("chr8", "chr6"), start=c(145515270, 122720696), end=c(145538385, 122754264), 
                      gene=c("HSF1", "HSF2"), stringsAsFactors=FALSE)
idx <- mapply(function(x, chr, s, e, w) which(x$chrom == chr & x$loc >= s - w & x$loc <= e), 
              chr=genePos$chrom, s=genePos$start, e=genePos$end, MoreArgs=list(x=geno$bim, w=cisWindow),
              SIMPLIFY=FALSE)
idx <- Reduce(union, idx)
targetSNPs <- geno$bim[idx, ]
targetGeno <- geno$bed[targetSNPs$snp, ]
```

```{r writeHSFsnps, cache=FALSE}
write.table(targetGeno, file="tmp/genotypes_hsf.tab", sep="\t", quote=FALSE)
```

```{r runHSFtransFC, cache=FALSE, results="hide"}
snpsHSF <- SlicedData$new();
snpsHSF$fileDelimiter <- "\t"
snpsHSF$fileOmitCharacters <- "NA"
snpsHSF$fileSkipRows <- 1
snpsHSF$fileSkipColumns <- 1
snpsHSF$fileSliceSize <- 2000
snpsHSF$LoadFile(paste0("tmp/genotypes_hsf.tab"))

pheno <- SlicedData$new();
pheno$fileDelimiter <- "\t"
pheno$fileOmitCharacters <- "NA"
pheno$fileSkipRows <- 1
pheno$fileSkipColumns <- 1
pheno$fileSliceSize <- 2000
pheno$LoadFile("tmp/selected_probes_fc.tab")

meHSF <- Matrix_eQTL_main(
    snps = snpsHSF,
    gene = pheno,
    cvrt = covar,
    output_file_name = "tmp/reQTL_hsf_analysis",
    pvOutputThreshold = 1e-3,
    pvOutputThreshold.cis = 0,
    useModel = modelLINEAR, 
    errorCovariance = numeric(),
    verbose = FALSE,
    pvalue.hist = FALSE,
    min.pv.by.genesnp = TRUE,
    noFDRsaveMemory = FALSE)

hsfQTL <- formatQTL(meHSF, limmaTable)
```

```{r runHSFtransExpr, cache=FALSE, results="hide"}
nPC_hsf_expr <- 1
write.table(t(pc_stim)[1:nPC_hsf_expr, , drop=FALSE], file = "tmp/covar_hsf_expr_stim.tab", sep="\t", quote=FALSE)
write.table(t(pc_basal)[1:nPC_hsf_expr, , drop=FALSE], file = "tmp/covar_hsf_expr_basal.tab", sep="\t", quote=FALSE)

covarHSFstim <- SlicedData$new();
covarHSFstim$fileDelimiter <- "\t"
covarHSFstim$fileOmitCharacters <- "NA"
covarHSFstim$fileSkipRows <- 1
covarHSFstim$fileSkipColumns <- 1
covarHSFstim$fileSliceSize <- 2000
covarHSFstim$LoadFile("tmp/covar_hsf_expr_stim.tab")

covarHSFbasal <- SlicedData$new();
covarHSFbasal$fileDelimiter <- "\t"
covarHSFbasal$fileOmitCharacters <- "NA"
covarHSFbasal$fileSkipRows <- 1
covarHSFbasal$fileSkipColumns <- 1
covarHSFbasal$fileSliceSize <- 2000
covarHSFbasal$LoadFile("tmp/covar_hsf_expr_basal.tab")

meHSFstim <- Matrix_eQTL_main(
    snps = snpsHSF,
    gene = phenoStim,
    cvrt = covarHSFstim,
    output_file_name = "tmp/eQTL_hsf_stim_analysis",
    pvOutputThreshold = 1e-3,
    pvOutputThreshold.cis = 0,
    useModel = modelLINEAR, 
    errorCovariance = numeric(),
    verbose = FALSE,
    pvalue.hist = FALSE,
    min.pv.by.genesnp = TRUE,
    noFDRsaveMemory = FALSE)

meHSFbasal <- Matrix_eQTL_main(
    snps = snpsHSF,
    gene = phenoBasal,
    cvrt = covarHSFbasal,
    output_file_name = "tmp/eQTL_hsf_basal_analysis",
    pvOutputThreshold = 1e-3,
    pvOutputThreshold.cis = 0,
    useModel = modelLINEAR, 
    errorCovariance = numeric(),
    verbose = FALSE,
    pvalue.hist = FALSE,
    min.pv.by.genesnp = TRUE,
    noFDRsaveMemory = FALSE)

hsfBasalQTL <- formatQTL(meHSFbasal, limmaTable)
hsfStimQTL <- formatQTL(meHSFstim, limmaTable)
```



# Results
## Differential expression analysis

In total `r nrow(filter(limmaTable, adj.P.Val < fdrCut))` probes are identified
as differentially expressed at an FDR threshold of `r fdrCut` (see `r tabRef("topTable")`
for the top 20 results). Of these `r nrow(filter(limmaTable, adj.P.Val < fdrCut & logFC > 0))` 
are up and `r nrow(filter(limmaTable, adj.P.Val < fdrCut & logFC < 0))` down regulated
following heat shock treatment. Further restricting this list by excluding all probes with a 
fold change of less than `r 2^fcCut` leaves `r nrow(filter(limmaTable, adj.P.Val < fdrCut & abs(logFC) > fcCut))` 
probes with `r nrow(filter(limmaTable, adj.P.Val < fdrCut & logFC > fcCut))` and 
`r nrow(filter(limmaTable, adj.P.Val < fdrCut & logFC < -fcCut))` up and down regulated respectively. 
`r figRef("volcano")` shows a smoothed volcano plot highlighting probes of particular interest. 

```{r limmaResult, cache=FALSE}
set.caption(tabRef("topTable", "Top 20 differentially expressed probes."))
rownames(limmaTable) <- NULL
pander(head(dplyr::select(limmaTable, -ArrayAddress, -EntrezReannotated, -GenomicLocation), 20))
```

```{r limmaPlot, cache=FALSE, fig.cap=figRef("volcano", "Volcano plot of differential expression results. Probes with an adjusted p-value below 0.01 and a log fold cange of at least 0.5 are shown as yellow and red dots. Probes showing particularly strong evidence of changes in gene expression through a combination of p-value and fold change are labelled with the corresponding gene symbol.")}
fcCut2 <- log2(2)   ## label all genes with logFC greater than this
fcScale <- max(-log10(limmaTable$adj.P.Val))/max(abs(limmaTable$logFC))
limmaExtreme <- dplyr::filter(limmaTable, abs(logFC) > fcCut & adj.P.Val < fdrCut)
limmaExtreme <- mutate(limmaExtreme, distance=(fcScale*logFC)^2 + log10(adj.P.Val)^2)
limmaExtreme <- limmaExtreme[order(limmaExtreme$distance, decreasing=TRUE), ]
ggplot(limmaTable, aes(x=logFC, y=-log10(adj.P.Val))) + 
        stat_density2d(data=dplyr::filter(limmaTable, abs(logFC) <= fcCut | adj.P.Val >= fdrCut),
                      geom="tile", aes(fill=..density..^0.1), contour=FALSE) + 
        geom_point(data=limmaExtreme, size=1.5, aes(colour=distance)) +
        geom_text(data=dplyr::filter(limmaExtreme, distance > (2*fcScale)^2 & logFC >= fcCut2), 
                aes(label=SymbolReannotated), vjust=0, hjust=0, angle=30) +
        geom_text(data=dplyr::filter(limmaExtreme, distance > (2*fcScale)^2 & logFC <= -fcCut2), 
                aes(label=SymbolReannotated), vjust=0, hjust=1, angle=330) +
        geom_line(data=data.frame(x=c(-Inf, -fcCut), y=c(-log10(fdrCut), -log10(fdrCut))), 
        		aes(x=x, y=y), linetype="dotted") +
        geom_line(data=data.frame(x=c(fcCut, Inf), 
        		y=c(-log10(fdrCut), -log10(fdrCut))), aes(x=x, y=y), linetype="dotted") +
        geom_line(data=data.frame(x=c(fcCut, fcCut), y=c(-log10(fdrCut), Inf)), 
        		aes(x=x, y=y), linetype="dotted") +
        geom_line(data=data.frame(x=c(-fcCut, -fcCut), y=c(-log10(fdrCut), Inf)), 
        		aes(x=x, y=y), linetype="dotted") +
        scale_fill_gradientn(colours = colorRampPalette(c("white", blues9))(256), breaks=NULL) +
        scale_colour_gradientn(colours = colorRampPalette(c("yellow", "red"))(256), 
        		guide="none") +
        theme_bw() + xlab("log2(fold change)") + ylab("-log10(FDR)") + 
        expand_limits(x=c(min(limmaTable$logFC)-0.25, max(limmaTable$logFC)+0.25), 
            y=max(-log10(limmaTable$adj.P.Val))+2)
```

```{r heatmap, cache=FALSE, fig.cap=figRef("exprHeat", "Heatmap comparing gene expression for differentially expressed genes between basal and stimulated samples. Samples were clustered by gene  with stimulated (orange) and basal (purple) samples forming two distinct groups. Expression estimates for each gene were scaled and centred across samples. Blue cells correspond to lower and red cells to higher than average expression.")}
idx <- dplyr::arrange(dplyr::filter(limmaTable, ArrayAddress %in% rownames(selectedExpr)), logFC)$ArrayAddress
sideCols <- c(Basal="purple", Stim="orange")
treat <- str_extract(colnames(selectedExpr), "(Basal|Stim)$")
heatmap.2(selectedExpr[rev(idx),],trace = "none", Rowv=FALSE, Colv=TRUE, scale="row", dendrogram="column", col=rev(brewer.pal(11, "RdBu")), ColSideColors=sideCols[treat], key=FALSE, labRow="", labCol="", xlab="Samples", ylab="Genes")
```

## GO enrichment analysis
The GO analysis highlights several categories that are significantly enriched among
the up and down regulated genes (`r sum(score(resultUp) < 0.01)` and 
`r sum(score(resultDown) < 0.01)` categories with a p-value < 0.01 respectively).
Considering the top categories (`r tabRef("topGOUp")` and `r tabRef("topGODown")`)
it appears that genes up regulated in response to heat shock are predominantly
related to stress response (including *response to heat* [GO:0009408]) 
and protein folding (including *response to unfolded protein* [GO:0006986] and 
*protein refolding* [GO:0042026]) whereas down regulated genes
broadly relate to cell proliferation and normal lymphocyte function and 
include *positive regulation of cell cycle* [GO:0045787], *spindle assembly* [GO:0051225]
and *chemokine production* [GO:0032602].

```{r tableTopGOUp, chache=FALSE}
set.caption(tabRef("topGOUp", "Top 20 GO categories enriched for up regulated genes."))
pander(GenTable(tgUp, up=resultUp, down=resultDown, topNodes=20, orderBy="up"))
```

```{r tableTopGODown, chache=FALSE}
set.caption(tabRef("topGODown", "Top 20 GO categories enriched for down regulated genes"))
pander(GenTable(tgDown, down=resultDown, up=resultUp, topNodes=20, orderBy="down"))
```

```{r goPlot, cache=FALSE, fig.width=10, fig.height=6, fig.cap=figRef("goPlot", "GO category enrichment for differentially expressed genes.")}
tblUp <- GenTable(tgUp, up=resultUp, topNodes=length(score(resultUp)))
tblUp$up <- as.numeric(tblUp$up)
tblDown <- GenTable(tgDown, down=resultDown, topNodes=length(score(resultDown)))
tblDown$down <- as.numeric(tblDown$down)
tbl <- bind_rows(gather(tblUp, "direction", "p.value", up), gather(tblDown, "direction", "p.value", down))
ggplot(tbl, aes(x=Expected, y=Significant)) + 
    geom_point(data=dplyr::filter(tbl, p.value >= 1e-4), colour="grey") +
    geom_point(data=dplyr::filter(tbl, p.value < 1e-4), aes(fill=-log10(p.value), shape=direction), 
        size=2.5) +
    geom_abline(intercept=0, slope=1) +
    scale_shape_manual(values=c(up=24, down=25)) +
    theme_bw()
```

## Gene annotations

Of the `r length(unique(dplyr::filter(limmaTable, adj.P.Val < fdrCut & logFC > fcCut)$SymbolReannotated))` 
significantly up regulated genes 
`r sum(knownList$known %in% dplyr::filter(limmaTable, adj.P.Val < fdrCut & logFC > fcCut)$SymbolReannotated)` have been directly linked to the heat shock response and 
`r length(knownList$extended)`
are associated with GO terms that clearly relate to heat shock response or other relevant processes
or are otherwise linked to the heat shock response in the literature. This leaves 
`r length(knownList$novel)` genes with no obvious association to heat shock.

```{r peakEnrichTab, cache=FALSE}
condition <- str_replace(hsfPval$cond, "comb-C_", "basal ")
condition <- str_replace(condition, "comb-HS_", "heat shock ")
tab <- data.frame(Condition=condition, 
                  up=paste0(hsfPval$sigGenes_up, " (", format(hsfPval$up, digits=2),")"),
                  down=paste0(hsfPval$sigGenes_down, " (", format(hsfPval$down, digits=2),")"))
tab$up <- str_replace(tab$up, fixed("e+00"), "") 
tab$down <- str_replace(tab$down, fixed("e+00"), "")
tab$up <- sub("\\(([[:digit:]]\\.[[:digit:]])e(-[[:digit:]]+)\\)", "\\$(\\1 \\\\times 10^{\\2})\\$", tab$up)

set.caption(tabRef("peakEnrich", "Number (and corresponding p-values) of ChIP-seq peaks for HSF1 and HSF2 associated with significantly up and down regulated genes."))
highlightStrong <- NULL
if(any(hsfPval$up < 0.001)) highlightStrong <- cbind(which(hsfPval$up < 0.001), 2)
if(any(hsfPval$down < 0.001)) highlightStrong <- rbind(highlightStrong, cbind(which(hsfPval$down < 0.001), 3))

highlight <- NULL
if(any(hsfPval$up < 0.01 & hsfPval$up > 0.001)) 
  highlight <- cbind(which(hsfPval$up < 0.01 & hsfPval$up > 0.001), 2)
if(any(hsfPval$down < 0.01 & hsfPval$down > 0.001)) 
  highlight <- rbind(highlight, cbind(which(hsfPval$down < 0.01 & hsfPval$down > 0.001), 3))

pander(tab, emphasize.cells=highlight, emphasize.strong.cells=highlightStrong)
```

There are `r nrow(dplyr::filter(r_promoter, HSFpeak == 1, novel == 2))` up-regulated
genes without an established role and
`r nrow(dplyr::filter(r_promoter, HSFpeak == 1, novel == 1))` genes with limited 
evidence for their involvement in the heat shock response that show evidence of 
heat shock factor binding in their promoter region. Further evidence for the
binding of HSF1 and HSF2 in these regions is provided by scanning for the presence
of the HSF1 binding motif. While this provides less direct evidence than the presence
of a ChIP-seq peak and even a good match for the motif is no guarantee for the 
presence of an active binding site, it enables the detection of potential binding sites that may be
inactive in the cell type or state studied through ChIP-seq experiments. Although
overall the evidence for the presence of the HSF1 motif is relatively weak, there
is a clear increase in the average score for genes that show evidence of HSF binding
compared to those that don't (`r figRef("pwmPlot")`). 

```{r pwmPlot, cache=FALSE, fig.cap=figRef("pwmPlot", "Distribution of p-values for HSF1 motif score.")}
ggplot(dplyr::filter(r_gene, expr=="up"), 
       aes(x=factor(HSFpeak, levels=c("0", "1"), labels=c("absent", "present")), 
           y=-log10(pwm.p.value), 
           fill=factor(novel, levels=c("0", "1", "2"), labels=c("established", "limited", "novel")))) +
  geom_boxplot() + theme_bw() + xlab("HSF peak") + ylab(expression(-log[10]("p-value"))) +
  scale_fill_discrete(name="heat shock association")
```

```{r HSFtable, cache=FALSE}
hsfTab <- data.frame(expression=r_gene$expr, ChIPseq=ifelse(r_gene$HSFpeak==1, "yes", "no"),
                     motif=ifelse(r_gene$pwm.p.value < 0.05, "present", "absent"))
tab <- tidyr::spread(as.data.frame(table(hsfTab)), motif, Freq)
tab[["frequency"]] <- paste0(sprintf("%0.2f",tab$present/(tab$present + tab$absent)*100),"%")

upTest <- fisher.test(tab[tab$expression=="up", c("absent", "present")], alternative="greater")
downTest <- fisher.test(tab[tab$expression=="down", c("absent", "present")], alternative="greater")

set.caption(tabRef("HSFtable", "Summary of HSF1 motif occurance in promoters of differentially expressed genes. with and without HSF ChIP-seq peaks."))
pander(tab)
```

Considering the frequency with
wich HSF1 motifs occur (using a p-value threshold of 0.05) in the promoter regions 
of genes with and without ChIP-seq binding evidence (`r tabRef("HSFtable")`) it is apparent that the motif
is enriched amongst up-regulated genes with corresponding protein binding evidence
(Fisher's exact test p-value: `r format(upTest$p.value, digits=2)`). This enrichment
is less pronounced for down regulated genes  (`r format(downTest$p.value, digits=2)`).

Further examining up-regulated genes for evidence of HSF binding reveals that
`r nrow(dplyr::filter(r_gene, novel==2, HSFpeak==1))`
genes not previously characterised as heat shock response genes have ChIP-seq peaks
and of these
`r nrow(dplyr::filter(r_gene, novel==2, HSFpeak==1, pwm.p.value < 0.05))` also
carry the HSF1 binding motif in their promoter (`r tabRef("HSFnovel")` and 
`r tabRef("novelTable")`).

```{r HSFnovel, cache=FALSE}
hsfTab$type <- ifelse(r_gene$novel==2, "novel", ifelse(r_gene$novel==1, "limited", "established"))
hsfTab <- dplyr::filter(hsfTab, expression=="up") %>% dplyr::select(-expression)
tab <- tidyr::spread(as.data.frame(table(hsfTab)), type, Freq)

set.caption(tabRef("HSFnovel", "Summary of HSF binding evidence for the promoters of novel and established heat shock response genes."))
pander(tab, emphasize.strong.cells=matrix(c(3,4,5,5), ncol=2))
```

```{r novelTable, cache=FALSE}
set.caption(tabRef("novelTable", "Genes with newly established links to heat shock response"))
tab <- dplyr::filter(r_gene, novel==2, HSFpeak==1) %>% 
  dplyr::select(`Gene Symbol`=SymbolReannotated,
                `Fold Change`=logFC, 
                `expression FDR`=adj.P.Val,
                `Motif p-value`=pwm.p.value) %>%
  dplyr::arrange(desc(`Fold Change`), `Motif p-value`)
tab[["Fold Change"]] <- 2^tab[["Fold Change"]]
pander(tab)
```


## Global heat shock response

Considering the outcome of the partial least squares regression it is apparent that there is
a considerable change in overall gene expression in response to heat shock. Together the first
two PLS components account for `r format(sum(plsFit$R2[1:2]*100), nsmall=1, digits=2)`% 
in the variation observed due to 
heat shock and provide clear separation of the two treatment groups (`r figRef("treatPLS")`).

```{r treatPLSplot, cache=FALSE, fig.cap=figRef("treatPLS", "Samples separate clearly by treatment, showing a consistent global effect on gene expression from heat shock.")}
plsComp <- cbind(plsComp, sex=factor(globalPheno[plsComp$sample, "sex"]))
plsPlot(plsComp)
```

Of the `r length(assoc$adj.p.val)` SNPs tested for association with the global response phenotype
`r sum(assoc$adj.p.val < 0.05)` are considered significant (FDR < 0.05) (`r tabRef("treatPLS")`). 
These SNPs form `r countBlocks(geno$bed, assoc$adj.p.val)` group of highly correlated genotypes,
likely due to local LD structure.

```{r treatPLStable, cache=FALSE}
sel <- which(assoc$adj.p.val < 0.05)
coef <- assoc$Results[sel, , 1]
coef[, "JointModel"] <- assoc$adj.p.val[sel]
colnames(coef)[ncol(coef)] <- "Joint model FDR"
set.caption(tabRef("treatPLS", "Standardised coefficients and adjusted p-values for the top SNPs."))
pander(coef)
```


```{r treatPLSpval, cache=FALSE, fig.cap=figRef("permPval", paste0("P-values observed after permutation of global response phenotype for SNP ", dimnames(assoc$Result)[[1]][1], " (left) and the minimum p-value observed in each iteration across all SNPs (right)."))}

fig1 <- ggplot(data.frame(p.value=pPerm[,bins[dimnames(assoc$Result)[[1]][1]]]), aes(x=-log10(p.value))) + 
  geom_histogram(fill="lightgrey", binwidth=0.1) +
  geom_vline(xintercept=-log10(assoc$Results[1, "JointModel",2]), colour="red") + 
  theme_bw() + xlab(expression(-log[10]("p-value")))

fig2 <- ggplot(data.frame(p.value=apply(pPerm,1, min)), aes(x=-log10(p.value))) + 
  geom_histogram(fill="lightgrey", binwidth=0.1) +
  geom_vline(xintercept=-log10(assoc$Results[1, "JointModel",2]), colour="red") + 
  theme_bw() + xlab(expression(-log[10]("p-value")))

multiplot(fig1, fig2, cols=2)
```

The SNP rs7789228 is located in an intron of RSBN1L and is annotated as falling within an
enhancer by Ensembl's regulatory build (Ensembl release 81). Although none of the cell lines
studied here are included in the regulatory build genome segmentation data are available for
GM12878, a HapMap cell line of Caucasian ancestry, providing a useful reference for the
samples in this study. In GM12878 the enhancer element is annotated as being active with
evidence of CTCF and RAD21 binding.

```{r plsGenoPlot, cache=FALSE, fig.cap=figRef("plsGeno", "Global response to heat shock by genotype.")}
genotypes <- data.frame(sample=colnames(geno$bed), stringsAsFactors=FALSE)
genotypes <- cbind(genotypes, t(geno$bed[dimnames(assoc$Results)[[1]][sel],]))
plsComp <- dplyr::left_join(plsComp, genotypes, by="sample")

plsPlot(plsComp, dimnames(assoc$Results)[[1]][1])
```

## Response QTL analysis

```{r reQTLGeneTable, cache=FALSE}
set.caption(tabRef("reQTLgenes", "List of genes with strongest evidence for response QTL."))
reQTLSigif <- dplyr::filter(reQTL, FDR < 0.05)
pander(dplyr::select(reQTLSigif, -NuID))
```

The association test designed to identify genetic modulators of gene expression response to 
heat shock only uncovered very limited evidence for the presence of such effects
(`r tabRef("reQTLgenes")`). A closer inspection of the potential cis associations for
the highest ranked gene (`r geneP$Symbol[1]`) suggests `r if(length(geneSNPcount)) geneSNPcount[1] else "no"`
SNP`r if(!length(geneSNPcount) || geneSNPcount[1] != 1) "s" else ""` 
that may be of interest (`r tabRef("topGeneSNPs")`).

```{r topGeneSNPs, cache=FALSE, eval=nrow(selGenes) > 0, include=nrow(selGenes) > 0}
set.caption(tabRef("topGeneSNPs", paste("Possible Response QTL for", geneP$Symbol[1])))
cut <- meGene[[1]]$cis$eqtls$FDR[5]
n <- max(which(meGene[[1]]$cis$eqtls$FDR == cut))
tab <- head(meGene[[1]]$cis$eqtls, max(geneSNPcount[1], n))
tab$gene <- geneP$Symbol[1]
pander(tab)
```

```{r topGeneSNPplot, cache=FALSE, eval=nrow(selGenes) > 0, include=nrow(selGenes) > 0, fig.cap=figRef("topGeneSNP", paste("Gene expression response to heat shock for", geneP$Symbol[1], "by genotype. Fold changes were corrected for the effect of covariates."))}
snp <- as.character(meGene[[1]]$cis$eqtls$snps[1])
gene <- geneP$Symbol[1]
probe <- geneP$NuID[1]

fit <- lm(selectedFC[probe, ] ~ geno$bed[snp,] + t(covar[[1]]))
offset <- colSums(fit$coefficients[-(1:2)]*covar[[1]])

plotData <- data.frame(genotype=geno$bed[snp,], fold_change=selectedFC[probe, ]-offset)
ggplot(plotData, aes(x=factor(genotype), y=fold_change)) + 
  geom_point(position=position_jitter(width=0.1), size=2.5) +
  geom_boxplot(alpha=0.4, outlier.colour="white", fill="grey") +
  theme_bw() + xlab(snp) + ylab("log fold change")
```


## Analysis of eQTL in basal and stimulated cells

```{r eQTLStimGeneTable, cache=FALSE}
set.caption(tabRef("eQTLstimGenes", "List of genes with strongest evidence for genetic modulation of gene expression after heat shock treatment. Significant associations only observed after heat shock treatment are highlight."))
highlight <- which(head(geneP_stim)$Symbol %in% stimGenes$Symbol)
rows <- max(6, max(which(geneP_stim$Gene.FDR < 0.05)))
pander(dplyr::select(head(geneP_stim, rows), -ArrayAddress, -NuID), 
       emphasize.strong.rows=if(length(highlight)) highlight else NULL)
```

```{r eQTLBasalGeneTable, cache=FALSE}
set.caption(tabRef("eQTLbasalGenes", "List of genes with strongest evidence for genetic modulation of gene expression in basal state. Significant associations specific to the basal state are highlighted."))
highlight <- which(head(geneP_basal)$Symbol %in% basalGenes$Symbol)
rows <- max(6, max(which(geneP_basal$Gene.FDR < 0.05)))
pander(dplyr::select(head(geneP_basal, rows), -ArrayAddress, -NuID),
       emphasize.strong.rows=if(length(highlight)) highlight else NULL)
```

Condition specific associations were identified for `r nrow(stimGenes)` 
gene`r if(nrow(stimGenes) != 1) "s" else ""` in the stimulated and for `r nrow(basalGenes)`
in the basal state (`r tabRef("eQTLstimGenes")` and `r tabRef("eQTLbasalGenes")`).

```{r topEgeneSNPplot, cache=FALSE, fig.cap=figRef("topEgeneSNP", paste("Gene expression by genotype for genes showing evidence of condition specific eQTL. Gene expression estimates were corrected for the effect of covariates.")), include=nrow(basalGenes) + nrow(stimGenes) > 0, eval=nrow(basalGenes) + nrow(stimGenes) > 0}

stimPlot <- NULL
if(nrow(stimGenes)){
  snp <- as.character(meGene_stim[[stimGenes$NuID[1]]]$cis$eqtls$snps[1])
  gene <- stimGenes$Symbol[1]
  probe <- stimGenes$NuID[1]
  
  stimFit <- lm(stimExpr[probe, ] ~ geno$bed[snp,] + t(covarStim[[1]]))
  stimOffset <- colSums(stimFit$coefficients[-(1:2)]*covarStim[[1]])
  
  basalFit <- lm(basalExpr[probe, ] ~ geno$bed[snp,] + t(covarBasal[[1]]))
  basalOffset <- colSums(basalFit$coefficients[-(1:2)]*covarBasal[[1]])
  
  
  plotData <- data.frame(genotype=rep(geno$bed[snp,], 2), 
                         expression=c(stimExpr[probe, ]-stimOffset, basalExpr[probe,]-basalOffset),
                         condition=c(rep("stimulated", ncol(stimExpr)), rep("basal", ncol(basalExpr))))
  plotData <- dplyr::filter(plotData, !is.na(genotype))
  stimPlot <- ggplot(plotData, aes(x=factor(genotype), y=expression)) + 
    geom_boxplot(aes(fill=condition), alpha=0.4, outlier.colour="white", outlier.size=0) +
    theme_bw() + xlab(snp) + ylab(paste(gene, "expression"))
}

basalPlot <- NULL
if(nrow(basalGenes)){
  snp <- as.character(meGene_basal[[basalGenes$NuID[1]]]$cis$eqtls$snps[1])
  gene <- basalGenes$Symbol[1]
  probe <- basalGenes$NuID[1]
  
  stimFit <- lm(stimExpr[probe, ] ~ geno$bed[snp,] + t(covarStim[[1]]))
  stimOffset <- colSums(stimFit$coefficients[-(1:2)]*covarStim[[1]])
  
  basalFit <- lm(basalExpr[probe, ] ~ geno$bed[snp,] + t(covarBasal[[1]]))
  basalOffset <- colSums(basalFit$coefficients[-(1:2)]*covarBasal[[1]])
  
  plotData <- data.frame(genotype=rep(geno$bed[snp,], 2), 
                         expression=c(stimExpr[probe, ]-stimOffset, basalExpr[probe,]-basalOffset),
                         condition=c(rep("stimulated", ncol(stimExpr)), rep("basal", ncol(basalExpr))))
  plotData <- dplyr::filter(plotData, !is.na(genotype))
  basalPlot <- ggplot(plotData, aes(x=factor(genotype), y=expression)) + 
    geom_boxplot(aes(fill=condition), alpha=0.4, outlier.colour="white", outlier.size=0) +
    theme_bw() + xlab(snp) + ylab(paste(gene, "expression"))
}

plots <- list(stimPlot, basalPlot)
plots <- plots[!sapply(plots, is.null)]
multiplot(plotlist=plots)
```
## Trans associations for SNPs upstream of HSF1 and HSF2

```{r transGeneTable, cache=FALSE}
set.caption(tabRef("transGenes", "List of genes with strongest evidence for trans associations between heat shock response and SNPs upstream of HSF1 and HSF2."))
if(nrow(hsfQTL)) pander(head(hsfQTL, 10))
```

```{r transStimGeneTable, cache=FALSE}
set.caption(tabRef("transStimGenes", "List of genes with strongest evidence for trans associations between gene expression following heat shock and SNPs upstream of HSF1 and HSF2."))
if(nrow(hsfStimQTL)) pander(head(hsfStimQTL, 10))
```


```{r transBasalGeneTable, cache=FALSE}
set.caption(tabRef("transBasalGenes", "List of genes with strongest evidence for trans associations between gene expression in the basal state and SNPs upstream of HSF1 and HSF2."))
if(nrow(hsfBasalQTL)) pander(head(hsfBasalQTL, 10))
```


# Appendix {-}

```{r bundlePlots, include=FALSE, cache=FALSE}
pngDir <- "heatshock_analysis_files/figure-html"
pdfDir <- "heatshock_analysis_files/figure-latex"
if(file.exists(pngDir) && length(dir(pngDir, "*.png")) &&
     !file.exists(file.path(pngDir, "figures_png.tar.gz"))){
  system("(cd heatshock_analysis_files/figure-html && tar -czf figures_png.tar.gz *.png)")
}
if(file.exists(pdfDir) && length(dir(pdfDir, "*.pdf")) &&
     !file.exists(file.path(pdfDir, "figures_pdf.tar.gz"))){
  system("(cd heatshock_analysis_files/figure-latex && tar -czf figures_pdf.tar.gz *.pdf)")
}
```

## Custom functions used
```{r, ref.label="functions", eval=FALSE, echo=TRUE}
```

## Session Info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References {-}
