<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

  <head>

    <meta charset="utf-8" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="generator" content="pandoc" />

    
    
    <title>Data analysis for Characterisation of the global transcriptional response to heat shock and the impact of individual genetic variation</title>

        <script src="heatshock_analysis_files/jquery-1.11.3/jquery.min.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link href="heatshock_analysis_files/bootstrap3-3.2.0/css/bootstrap.min.css" rel="stylesheet" />
    <link href="heatshock_analysis_files/bootstrap3-3.2.0/css/themes/default/bootstrap.min.css" rel="stylesheet" />
    <script src="heatshock_analysis_files/bootstrap3-3.2.0/js/bootstrap.min.js"></script>
    <link href="heatshock_analysis_files/highlightjs-8.2/highlight/github.css" rel="stylesheet" />
    <script src="heatshock_analysis_files/highlightjs-8.2/highlight.pack.js"></script>
    <link href="heatshock_analysis_files/MagnificPopup-0.9.9/magnific-popup.css" rel="stylesheet" />
    <script src="heatshock_analysis_files/MagnificPopup-0.9.9/magnific-popup.js"></script>
    <script src="heatshock_analysis_files/Navigation-1.0/tabsets.js"></script>
    <link href="heatshock_analysis_files/knitrBootstrap-0.0.1/css/knitrBootstrap.css" rel="stylesheet" />
    <script src="heatshock_analysis_files/knitrBootstrap-0.0.1/js/knitrBootstrap.js"></script>
        <style type=text/css>
    .downloads .row {
      margin-left: 0;
    }
    </style>
    
    
  </head>

  <body data-spy="scroll" data-target="#toc">
    <div id="wrapper" class="container">
      <div class="row">
        <div class="contents col-xs-12 col-md-10">

          
                    <div class="page-header">
            <h1 class="title">Data analysis for <em>Characterisation of the global transcriptional response to heat shock and the impact of individual genetic variation</em></h1>
                                                <h4 class="author"><em>Peter Humburg</em></h4>
                        <address class="author_afil">
              Wellcome Trust Centre for Human Genetics, University of Oxford, Roosevelt Dr., Oxford, OX3 7BN, UK<br>                            <a class="author_email" href="mailto:#"><a href="mailto:peter.humburg@well.ox.ac.uk">peter.humburg@well.ox.ac.uk</a></a>
            </address>
                                                            <h4 class="author"><em>Wanseon Lee</em></h4>
                        <address class="author_afil">
              Wellcome Trust Centre for Human Genetics, University of Oxford, Roosevelt Dr., Oxford, OX3 7BN, UK<br>                            <a class="author_email" href="mailto:#"><a href="mailto:wanseon.lee@well.ox.ac.uk">wanseon.lee@well.ox.ac.uk</a></a>
            </address>
                                                              <h4 class="date"><em>Sun 28 Aug 2016</em></h4>
                      </div>
                    
<div id="introduction" class="section level1"><h1>Introduction</h1><p>The heat shock response is a highly conserved mechanism responsible for the maintenance of cellular function under stress. Originally identified as part of the response to increased environmental temperatures <span class="citation">(Ritossa 1962)</span> heat shock proteins (HSPs) have since been implicated in susceptibility to infection <span class="citation">(Kee et al. 2008)</span>, cancer risk <span class="citation">(Sfar et al. 2010)</span> and drug response <span class="citation">(Martin et al. 2004)</span>. Despite its obvious significance many details of the heat shock response still remain uncertain.</p><p>Here we study the genome-wide changes to gene expression induced by heat shock in 60 HapMap cell lines from Yoruba individuals <span class="citation">(Gibbs 2010)</span> to identify genes and pathways involved in the human heat shock response. To further elucidate underlying mechanisms genetic variants modulating gene expression are mapped across all relevant genes.</p></div><div id="methods" class="section level1"><h1>Methods</h1><div id="pre-processing-and-quality-control" class="section level2"><h2>Pre-processing and Quality Control</h2><div id="raw-gene-expression-qc" class="section level3"><h3>Raw gene expression QC</h3><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">rawIntensity <- readr::read_tsv("data/TPS00590_FinalReport_12Oct09.txt", skip = 8, 
    progress = FALSE)

sig.col <- seq(3, 722, 6)
det.col <- seq(8, 722, 6)

sig.data.raw <- rawIntensity[, sig.col]
det.data.raw <- rawIntensity[, det.col]

names(sig.data.raw) <- gsub(":AVG_Signal", "", names(sig.data.raw))
names(det.data.raw) <- gsub(":Detection Pval", "", names(det.data.raw))

info <- readr::read_tsv("data/YRI_sample_list_file_03NOV09.txt", col_types = list(`Date of Harvest` = col_date("%d.%m.%y"), 
    `Date of RNA extraction` = col_date("%d.%m.%y")))

## remove empty lines at end of file
info <- info[1:120, ]
factors <- which(sapply(info, is.factor))
for (i in factors) {
    info[[i]] <- factor(as.character(info[[i]]))
}
info[["Cell Line"]] <- sub("^GM", "NA", info[["Cell Line"]])
info[["Cell Line"]] <- sub("_.$", "", info[["Cell Line"]])
info <- mutate(info, sample = paste(`Cell Line`, Treatment, sep = "_"))

newNames <- paste(info[["Cell Line"]], info$Treatment, sep = "_")
names(newNames) <- info[["ARRAY CODE"]]

## typo in sample ID
names(sig.data.raw)[13] <- "NM-75"
names(det.data.raw)[13] <- "NM-75"

names(sig.data.raw) <- gsub("-", "_", fixed = TRUE, names(sig.data.raw))
names(det.data.raw) <- gsub("-", "_", fixed = TRUE, names(det.data.raw))  #differing name format
names(sig.data.raw) <- newNames[names(sig.data.raw)]
names(det.data.raw) <- newNames[names(det.data.raw)]

sig.data.raw <- cbind(probe = rawIntensity$ProbeID, sig.data.raw)
det.data.raw <- cbind(probe = rawIntensity$ProbeID, det.data.raw)

sig.data.log <- cbind(probe = rawIntensity$ProbeID, log2(sig.data.raw[-1]))</code></pre></div><p>Probe intensities for resting and stimulated cells were imported into R for further processing together with associated meta data.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">exclude <- list()
exclude$sample <- character()
exclude$probe <- character()</code></pre></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sig.data.long <- reshapeData(sig.data.log, "intensity")
sig.data.raw.long <- reshapeData(sig.data.raw, "intensity")
det.data.long <- reshapeData(det.data.raw, "p.value")</code></pre></div><p>Annotations for all probes were obtained via the <em>illuminaHumanv3.db</em> Bioconductor package <span class="citation">(Barbosa-Morais et al. 2010)</span>. Only probes that are considered to be of perfect or good quality were taken forward for analysis. Additionally, all probes mapping to more than one genomic location or to a location that contains a known SNP were excluded.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">annot <- illuminaHumanv3fullReannotation()
exclude$probe <- union(exclude$probe, dplyr::filter(annot, !grepl("Good|Perfect", 
    ProbeQuality) | !is.na(OverlappingSNP) | !is.na(RepeatMask))$ArrayAddress)
annot <- dplyr::filter(annot, !ArrayAddress %in% exclude$probe)

txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
probePos <- strsplit(dplyr::filter(annot, !is.na(GenomicLocation))$GenomicLocation, 
    "[:, ]")
probePos <- data.frame(ArrayAddress = dplyr::filter(annot, !is.na(GenomicLocation))$ArrayAddress, 
    chrom = sapply(probePos, "[[", 1), start = as.numeric(sapply(probePos, "[[", 
        2)), end = as.numeric(sapply(probePos, "[[", 3)), strand = sapply(probePos, 
        "[[", 4))
probePos <- makeGRangesFromDataFrame(probePos, keep.extra.columns = TRUE)
tx <- transcriptsByOverlaps(txdb, probePos)
hits <- findOverlaps(probePos, tx)
probeTx <- data.frame(ArrayAddress = probePos$ArrayAddress[queryHits(hits)], 
    tx_id = tx$tx_id[subjectHits(hits)])
tx2gene <- transcripts(txdb, vals = list(tx_id = unique(probeTx$tx_id)), columns = c("tx_id", 
    "gene_id"))
tx2gene <- dplyr::select(as.data.frame(tx2gene), tx_id, gene_id)
probeGene <- dplyr::left_join(probeTx, tx2gene, by = "tx_id")
probeGene <- unique(dplyr::select(probeGene, ArrayAddress, gene_id))
probeGene$gene_id <- as.character(probeGene$gene_id)

excl <- table(probeGene$ArrayAddress)
exclude$probe <- union(exclude$probe, names(excl)[excl > 1])
annot <- dplyr::filter(annot, !ArrayAddress %in% exclude$probe)

annot <- dplyr::left_join(annot, probeGene, by = "ArrayAddress")</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in left_join_impl(x, y, by$x, by$y): joining factor and character
## vector, coercing into character vector
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">annot$EntrezReannoated <- ifelse(is.na(annot$gene_id), annot$EntrezReannotated, 
    annot$gene_id)
annot <- dplyr::select(annot, -gene_id)

eg <- mappedkeys(org.Hs.egSYMBOL)
symbols <- as.list(org.Hs.egSYMBOL[eg])
symbols <- data.frame(entrez = names(symbols), symbol = unlist(symbols), stringsAsFactors = FALSE)
annot <- dplyr::left_join(annot, symbols, by = c(EntrezReannotated = "entrez"))
annot$SymbolReannotated <- ifelse(is.na(annot$symbol), annot$SymbolReannotated, 
    annot$symbol)
annot <- dplyr::select(annot, -symbol)</code></pre></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">det.pval <- 0.01
det.n <- 10
sample.missing <- 0.7
sd.min <- 0.8</code></pre></div><p>Probes were required to exhibit significant signal (detection p-value &lt; 0.01) in at least 10 samples and samples with less than 30% of the remaining probes providing significant signal were excluded (together with the paired sample). Samples showing exceptionally low variation in probe intensities (standard deviation of the log intensities of all retained probes below 0.8) were also removed.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">detected <- tally(group_by(subset(det.data.long, p.value < det.pval), probe))
exclude$probe <- union(exclude$probe, filter(detected, n < det.n)$probe)</code></pre></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sig.data.long <- filter(sig.data.long, !probe %in% exclude$probe)
sig.data.raw.long <- filter(sig.data.raw.long, !probe %in% exclude$probe)
det.data.long <- filter(det.data.long, !probe %in% exclude$probe)</code></pre></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">missingSample <- tally(group_by(filter(det.data.long, p.value > det.pval), sample, 
    treatment))
missingSample$n <- missingSample$n/count(det.data.long, sample, treatment)$n
exclude$sample <- union(exclude$sample, filter(missingSample, n >= sample.missing)$sample)</code></pre></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sampleSD <- summarise(group_by(filter(sig.data.long, is.finite(intensity)), 
    sample, treatment), sd(intensity))
exclude$sample <- union(exclude$sample, filter(sampleSD, `sd(intensity)` < sd.min)$sample)</code></pre></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">dat <- filterSample(sig.data.long, sig.data.raw.long, det.data.long, exclude$sample)
sig.data.long <- dat$sig
sig.data.raw.long <- dat$sig.raw
det.data.long <- dat$det</code></pre></div><p>After filtering 12416 of 48803 probes (25.4%) and 56 samples remain. See Figure 1 for the distributions of intensities for the remaining probes in each sample.</p><div class="container-fluid"><div id="densityPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">ggplot(sig.data.long, aes(x = intensity, colour = sample, linetype = treatment)) + 
    geom_density() + theme_bw() + guides(colour = "none")</code></pre><div class="row"><div id="fig:densityPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/densityPlot-1.png" alt="**Figure 1:** Distribution of probe intensities."></img> </a></p><p class="caption"><strong>Figure 1:</strong> Distribution of probe intensities.</p></div></div></div></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">flt <- filter(sig.data.long, !is.finite(intensity))$probe
sig.data.filter <- full_join(spread(filter(sig.data.long, treatment == "Basal" & 
    !probe %in% flt), sample, intensity)[-1], spread(filter(sig.data.long, treatment == 
    "Stim" & !probe %in% flt), sample, intensity)[-1], by = "probe")
names(sig.data.filter) <- sub(".x", "_Basal", names(sig.data.filter), fixed = TRUE)
names(sig.data.filter) <- sub(".y", "_Stim", names(sig.data.filter), fixed = TRUE)

sig.pca <- prcomp(t(sig.data.filter[-1]), scale = TRUE)
sig.pc <- as.data.frame(sig.pca$x)
sig.pc$sample <- rownames(sig.pc)
sig.pc <- dplyr::select(as.tbl(sig.pc), sample, PC1, PC2)
sig.pc <- inner_join(sig.pc, info, by = "sample")

cluster <- dplyr::filter(sig.pc, `Date cRNA Amplified` != "CE_29Sep09")
clusterBound <- chull(cluster$PC1, cluster$PC2)</code></pre></div><div class="container-fluid"><div id="pcaPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">ggplot(sig.pc, aes(x = PC1, y = PC2)) + geom_polygon(data = cluster[clusterBound, 
    ], fill = "lightgrey", colour = "grey") + geom_point(aes(colour = as.character(Sentrix_ID), 
    shape = `Date cRNA Amplified`), size = 3) + theme_bw() + scale_colour_discrete("BeadChip") + 
    scale_shape_discrete("RNA Amplification")</code></pre><div class="row"><div id="fig:pcaPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/pcaPlot-1.png" alt="**Figure 2:** PCA plot of probe intensities by sample. Colours correspond to different BeadChips while shapes indicate different amplification dates. Samples from the two amplifications in October 2009 are clustered together (highlighted by the grey area)."></img> </a></p><p class="caption"><strong>Figure 2:</strong> PCA plot of probe intensities by sample. Colours correspond to different BeadChips while shapes indicate different amplification dates. Samples from the two amplifications in October 2009 are clustered together (highlighted by the grey area).</p></div></div></div></div><p>A principle component analysis of the gene expression for the remaining samples indicates a clustering of samples from two RNA amplification batches in October 2009 (Figure 2). This batch effect is noted for later correction during the analysis.</p><p>Further analysis of the gene expression response to heat shock revealed that four of the samples (NA18502, NA18508, NA18522, NA18853, NA18504) exhibit patterns of expression inconsistent with those observed in other samples, including reversal of fold change for strongly differentially expressed genes. This suggests a potential mislabelling of these samples, which were consequently excluded.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">exclude$sample <- union(exclude$sample, c("NA18502", "NA18508", "NA18522", "NA18853", 
    "NA18504"))
dat <- filterSample(sig.data.long, sig.data.raw.long, det.data.long, exclude$sample)
sig.data.long <- dat$sig
sig.data.raw.long <- dat$sig.raw
det.data.long <- dat$det</code></pre></div></div><div id="genotype-qc" class="section level3"><h3>Genotype QC</h3><div class="container-fluid"><div id="sampleList" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">keepSamples <- updateSampleList(sig.data.long$sample, "data/hapmap_YRI_r23a_filtered.fam")</code></pre></div></div><div class="container-fluid"><div id="plinkSNPs" class="row"><button class="source bash toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> bash source</button><pre style="display:none"><code class="source bash">plink --bfile data/hapmap_YRI_r23a_filtered --out tmp/hapmap_yri.snp_flt --keep tmp/sample.lst --maf 0.1 --geno 0.1 --hwe 1e-4 --autosome --make-bed > log/hapmap_yri.snp_flt.log</code></pre></div></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">totalSNPs <- R.utils::countLines("data/hapmap_YRI_r23a_filtered.bim")
remainSNPs <- R.utils::countLines("tmp/hapmap_yri.snp_flt.bim")</code></pre></div><p>Genotype data provided by the HapMap project <span class="citation">(Gibbs 2010)</span> was processed with Plink <span class="citation">(Chang et al. 2015)</span> to restrict the data to autosomes and remove SNPs with low genotyping rate and and those with a minor allele frequency of less than 10% in the sample. This results in the exclusion of 794511 of 2582999 SNPs (30.76%);</p><div class="container-fluid"><div id="plinkSample" class="row"><button class="source bash toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> bash source</button><pre style="display:none"><code class="source bash">plink --bfile tmp/hapmap_yri.snp_flt --out tmp/hapmap_yri.smpl_flt --mind 0.1 --neighbour 1 3 --genome gz nudge --make-bed > log/hapmap_yri.smpl_flt.log</code></pre></div></div><div class="container-fluid"><div id="nnIBS" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">nnIBS <- readr::read_table("tmp/hapmap_yri.smpl_flt.nearest")</code></pre></div></div><p>The IBS nearest neighbour calculation shows that the majority of samples are quite similar with regard to the extend of genetic differences between them while a few show markedly increased sharing of alleles (see Figure 3 and Table 1). This is consistent with a population of largely unrelated individuals including a few individuals that are more closely related than expected. One individual (NA18913) also shows evidence of reduced similarity with the second and third nearest neighbour.</p><div class="container-fluid"><div id="nnIBSplot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">ggplot(nnIBS, aes(x = Z, colour = as.factor(NN))) + geom_density(aes(group = NN)) + 
    theme_bw()</code></pre><div class="row"><div id="fig:nnIBSplot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/nnIBSplot-1.png" alt="**Figure 3:** Similarity between individuals. For each individual the similarity, in terms of IBS, to all other individuals is determined and individuals are ranked acording to this measure. From this a Z-score is computed for for each of the three closest neighbours (see the Plink [documentation](http://pngu.mgh.harvard.edu/~purcell/plink/strat.shtml#outlier) for details). Note the spikes corresponding to individuals with positive Z-scores indicating increased similarity of their genomes compared to the rest of the cohort."></img> </a></p><p class="caption"><strong>Figure 3:</strong> Similarity between individuals. For each individual the similarity, in terms of IBS, to all other individuals is determined and individuals are ranked acording to this measure. From this a Z-score is computed for for each of the three closest neighbours (see the Plink <a href="http://pngu.mgh.harvard.edu/~purcell/plink/strat.shtml#outlier">documentation</a> for details). Note the spikes corresponding to individuals with positive Z-scores indicating increased similarity of their genomes compared to the rest of the cohort.</p></div></div></div></div><div class="container-fluid"><div id="nnIBStable" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">offset <- mean(filter(nnIBS, Z < 0 & NN == 1)$Z)
nnSuspect <- filter(nnIBS, (abs(Z - offset) > 2 & NN == 1) | abs(Z) > 2)
nnSuspect <- nnSuspect[order(abs(nnSuspect$Z), decreasing = TRUE), ]
set.caption(tabRef("nnIBS", "Pairs of Yoruban HapMap samples with evidence of decreased or increased genetic similarity (absolute IBS-based nearest neighbour Z-score > 2)."))
pander(nnSuspect)</code></pre><table><caption><strong>Table 1:</strong> Pairs of Yoruban HapMap samples with evidence of decreased or increased genetic similarity (absolute IBS-based nearest neighbour Z-score &gt; 2).</caption><colgroup><col width="15%" /><col width="11%" /><col width="6%" /><col width="13%" /><col width="9%" /><col width="15%" /><col width="9%" /></colgroup><thead><tr class="header"><th align="center">FID</th><th align="center">IID</th><th align="center">NN</th><th align="center">MIN_DST</th><th align="center">Z</th><th align="center">FID2</th><th align="center">IID2</th></tr></thead><tbody><tr class="odd"><td align="center">Yoruba_028</td><td align="center">NA18913</td><td align="center">1</td><td align="center">0.8121</td><td align="center">4.343</td><td align="center">Yoruba_117</td><td align="center">NA19238</td></tr><tr class="even"><td align="center">Yoruba_117</td><td align="center">NA19238</td><td align="center">1</td><td align="center">0.8121</td><td align="center">4.343</td><td align="center">Yoruba_028</td><td align="center">NA18913</td></tr><tr class="odd"><td align="center">Yoruba_024</td><td align="center">NA18862</td><td align="center">3</td><td align="center">0.7007</td><td align="center">2.618</td><td align="center">Yoruba_048</td><td align="center">NA19203</td></tr><tr class="even"><td align="center">Yoruba_028</td><td align="center">NA18913</td><td align="center">2</td><td align="center">0.6985</td><td align="center">-2.340</td><td align="center">Yoruba_048</td><td align="center">NA19204</td></tr><tr class="odd"><td align="center">Yoruba_028</td><td align="center">NA18913</td><td align="center">3</td><td align="center">0.6984</td><td align="center">-2.325</td><td align="center">Yoruba_112</td><td align="center">NA19192</td></tr><tr class="even"><td align="center">Yoruba_024</td><td align="center">NA18862</td><td align="center">2</td><td align="center">0.7009</td><td align="center">2.183</td><td align="center">Yoruba_017</td><td align="center">NA18871</td></tr><tr class="odd"><td align="center">Yoruba_101</td><td align="center">NA19130</td><td align="center">1</td><td align="center">0.7535</td><td align="center">1.908</td><td align="center">Yoruba_112</td><td align="center">NA19192</td></tr><tr class="even"><td align="center">Yoruba_112</td><td align="center">NA19192</td><td align="center">1</td><td align="center">0.7535</td><td align="center">1.908</td><td align="center">Yoruba_101</td><td align="center">NA19130</td></tr></tbody></table></div></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">ibdReport <- read.table("tmp/hapmap_yri.smpl_flt.genome.gz", header = TRUE)
ibdSD <- sd(ibdReport$PI_HAT)
ibdHigh <- dplyr::filter(ibdReport, PI_HAT > 2 * ibdSD)</code></pre></div><p>A similar picture emerges if we consider the estimated proportion of IBD for all sample pairs with 3 pairs showing evidence of higher than expected relatedness (see Table 2 and Figure 4 for details).</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">pairs <- rbind(as.matrix(dplyr::select(ibdHigh, IID1, IID2)), as.matrix(dplyr::select(nnSuspect, 
    IID, IID2)))
exSample <- character()
while (length(pairs)) {
    ids <- sort(table(pairs), decreasing = TRUE)
    exSample <- c(exSample, names(ids[1]))
    pairs <- pairs[!apply(pairs, 1, `%in%`, x = names(ids)[1]), , drop = FALSE]
}
if ("NA19130" %in% exSample && !"NA19192" %in% exSample) {
    exSample[which(exSample == "NA19130")] <- "NA19192"
}

exclude$sample <- union(exclude$sample, exSample)</code></pre></div><p>To ensure that at most one sample of each suspect pair are included in the analysis the samples <em>NA18913</em>, <em>NA19192</em>, <em>NA18862</em> and <em>NA19092</em> are excluded.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">set.caption(tabRef("ibdHigh", paste("Sample pairs with estimated proportion of IBD exceeding", 
    format(2 * ibdSD, nsmall = 2, digits = 2))))
pander(dplyr::select(ibdHigh, IID1, IID2, PI_HAT, DST, RATIO))</code></pre><table><caption><strong>Table 2:</strong> Sample pairs with estimated proportion of IBD exceeding 0.032</caption><colgroup><col width="11%" /><col width="11%" /><col width="12%" /><col width="8%" /><col width="9%" /></colgroup><thead><tr class="header"><th align="center">IID1</th><th align="center">IID2</th><th align="center">PI_HAT</th><th align="center">DST</th><th align="center">RATIO</th></tr></thead><tbody><tr class="odd"><td align="center">NA18913</td><td align="center">NA19238</td><td align="center">0.5</td><td align="center">0.812</td><td align="center">76.8</td></tr><tr class="even"><td align="center">NA19092</td><td align="center">NA19101</td><td align="center">0.104</td><td align="center">0.723</td><td align="center">2.82</td></tr><tr class="odd"><td align="center">NA19130</td><td align="center">NA19192</td><td align="center">0.239</td><td align="center">0.753</td><td align="center">4.5</td></tr></tbody></table></div><div class="container-fluid"><div id="ibsDendroPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">plot(ibsClust(ibdReport), main = "", xlab = "IBS", sub = "")</code></pre><div class="row"><div id="fig:ibsDendroPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/ibsDendroPlot-1.png" alt="**Figure 4:** Dendrogram of individuals included in study. Distances are based on identity by state. Three pairs show clear indications of increased relatedness."></img> </a></p><p class="caption"><strong>Figure 4:</strong> Dendrogram of individuals included in study. Distances are based on identity by state. Three pairs show clear indications of increased relatedness.</p></div></div></div></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">dat <- filterSample(sig.data.long, sig.data.raw.long, det.data.long, exclude$sample)
sig.data.long <- dat$sig
sig.data.raw.long <- dat$sig.raw
det.data.long <- dat$det</code></pre></div><div class="container-fluid"><div id="sampleList2" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">keepSamples <- updateSampleList(sig.data.long$sample, "data/hapmap_YRI_r23a_filtered.fam")</code></pre></div></div><div class="container-fluid"><div id="plinkMdsPlot" class="row"><button class="source bash toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> bash source</button><pre style="display:none"><code class="source bash">plink --bfile tmp/hapmap_yri.smpl_flt --out tmp/hapmap_yri.smpl_flt2 --keep tmp/sample.lst --cluster --mds-plot 4 --make-bed > log/hapmap_yri.smpl_flt.log</code></pre></div></div><div class="container-fluid"><div id="mdsPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">mds <- read.table("tmp/hapmap_yri.smpl_flt2.mds", header = TRUE)
ggplot(mds, aes(x = C1, y = C2)) + geom_point(size = 3) + theme_bw()</code></pre><div class="row"><div id="fig:mdsPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/mdsPlot-1.png" alt="**Figure 5:** MDS plot of samples remaining after genotype QC."></img> </a></p><p class="caption"><strong>Figure 5:</strong> MDS plot of samples remaining after genotype QC.</p></div></div></div></div><p>Plotting the first two MDS components of the remaining 47 samples shows no further evidence for outliers (Figure 5).</p></div><div id="normalising-gene-expression-estimates" class="section level3"><h3>Normalising gene expression estimates</h3><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sig.data.norm <- justvsn(acast(sig.data.raw.long, formula = probe ~ sample + 
    treatment, value.var = "intensity"))</code></pre></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">probeMean <- rowMeans(sig.data.norm)
probeSD <- apply(sig.data.norm, 1, sd)
sig.select <- as.tbl(as.data.frame(sig.data.norm))
sig.select <- mutate(sig.select, mean = probeMean, sd = probeSD)
sig.select <- filter(sig.select, mean > 7)
sig.select <- arrange(sig.select, desc(sd), desc(mean))
sig.select <- head(sig.select, 2000)</code></pre></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sig.pca <- prcomp(t(sig.select), scale = TRUE)
sig.pc <- as.data.frame(sig.pca$x)
sig.pc$sample <- rownames(sig.pc)
sig.pc <- dplyr::select(as.tbl(sig.pc), sample, PC1, PC2, PC3, PC4)
sig.pc <- inner_join(sig.pc, info, by = "sample")</code></pre></div><div class="container-fluid"><div id="vsnPCAplot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">ggOutlier <- outlierPlot(sig.pc, level = 0.985, hjust = 0.2, vjust = -0.5)
sampleCount <- table(ggOutlier$data$Sentrix_ID)
excludeChip <- unique(dplyr::filter(ggOutlier$data, Sentrix_ID %in% names(sampleCount)[sampleCount <= 
    4])$Sentrix_ID)
outlierSample <- unique(dplyr::filter(ggOutlier$data, outlier)$Cell.Line)
chipExclusion <- unique(dplyr::filter(ggOutlier$data, Sentrix_ID %in% excludeChip)$Cell.Line)
exclude$sample <- Reduce(union, list(exclude$sample, outlierSample, chipExclusion))
ggOutlier</code></pre><button class="message R toggle btn btn-xs btn-info"><span class="glyphicon glyphicon-chevron-up"></span> R message</button><pre style="display:none"><code class="message r">## Too few points to calculate an ellipse
## Too few points to calculate an ellipse
## Too few points to calculate an ellipse
</code></pre><button class="message R toggle btn btn-xs btn-info"><span class="glyphicon glyphicon-chevron-up"></span> R message</button><pre style="display:none"><code class="message r">## geom_path: Each group consist of only one observation. Do you need to adjust the group aesthetic?
</code></pre><div class="row"><div id="fig:vsnPCAplot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/vsnPCAplot-1.png" alt="**Figure 6:** PCA plot of normalised gene expression estimates by sample. Each panal is comprised of the remaining samples for a singel BeadChip. For each BeadChip a 99% confidence ellipse is shown. While the majority of samples cluster well within chips some clear ouliers are visible. Variability between chips is relatively high with chips 563403730 and 563403752 appearing distinct from the bulk of the samples."></img> </a></p><p class="caption"><strong>Figure 6:</strong> PCA plot of normalised gene expression estimates by sample. Each panal is comprised of the remaining samples for a singel BeadChip. For each BeadChip a 99% confidence ellipse is shown. While the majority of samples cluster well within chips some clear ouliers are visible. Variability between chips is relatively high with chips 563403730 and 563403752 appearing distinct from the bulk of the samples.</p></div></div></div></div><p>Probe intensities were normalised with VSN <span class="citation">(Huber et al. 2002)</span>. A PCA plot of the initial VSN normalised data (Figure 6) shows clear evidence of samples clustering by BeadChip, as well as 2 obvious outliers, which were removed. In addition the few remaining samples from chip 4831371025 were removed as the entire chip appears to be suspect.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">dat <- filterSample(sig.data.long, sig.data.raw.long, det.data.long, exclude$sample)
sig.data.long <- dat$sig
sig.data.raw.long <- dat$sig.raw
det.data.long <- dat$det</code></pre></div><div class="container-fluid"><div id="exclOutlGeno" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">keepSamples <- updateSampleList(sig.data.long$sample, "tmp/hapmap_yri.smpl_flt2.fam")</code></pre></div></div><div class="container-fluid"><div id="plinkExclude" class="row"><button class="source bash toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> bash source</button><pre style="display:none"><code class="source bash">plink --bfile tmp/hapmap_yri.smpl_flt2 --out tmp/hapmap_yri.smpl_flt3 --keep tmp/sample.lst --make-bed > log/hapmap_yri.smpl_flt3.log</code></pre></div></div><p>The remaining samples were normalised separately for each BeadChip and differences between groups corrected with ComBat <span class="citation">(Johnson, Li, and Rabinovic 2007; Leek et al. 2015)</span>, preserving differences due to heat shock stimulation.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">batch <- with(subset(info, !`Cell Line` %in% exclude$sample), data.frame(sample = sample, 
    cluster = as.integer(factor(Sentrix_ID)), treatment = Treatment, stringsAsFactors = FALSE))
sig.raw <- acast(sig.data.raw.long, formula = probe ~ sample + treatment, value.var = "intensity")
colnames(sig.raw) <- sub("_._", "_", colnames(sig.raw))
sig.raw <- sig.raw[, batch$sample]

vsnFit <- vector(mode = "list", max(batch$cluster))
sig.data.norm <- vector(mode = "list", max(batch$cluster))
for (i in 1:max(batch$cluster)) {
    vsnFit[[i]] <- vsn2(sig.raw[, batch$cluster == i])
    sig.data.norm[[i]] <- predict(vsnFit[[i]], newdata = sig.raw[, batch$cluster == 
        i], useDataInFit = TRUE)
}
sig.data.norm <- do.call(cbind, sig.data.norm)</code></pre></div><div class="container-fluid"><div id="vsnBatchFit" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">par(mfrow = c(3, 3))
mapply(meanSdPlot, vsnFit, main = paste("Cluster", 1:max(batch$cluster)), MoreArgs = list(ylim = c(0, 
    1)))</code></pre><div class="row"><div id="fig:vsnBatchFit" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/vsnBatchFit-1.png" alt="**Figure 7:** VSN fits to grouped dataset."></img> </a></p><p class="caption"><strong>Figure 7:</strong> VSN fits to grouped dataset.</p></div></div><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">par(mfrow = c(1, 1))</code></pre></div></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">batch <- batch[order(batch$cluster), ]
sig.data.norm <- sig.data.norm[, batch$sample]
sig.data.norm <- ComBat(sig.data.norm, batch = batch$cluster, mod = model.matrix(~treatment, 
    data = batch))</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## Found 9 batches
## Adjusting for 1 covariate(s) or covariate level(s)
## Standardizing Data across genes
## Fitting L/S model and finding priors
## Finding parametric adjustments
## Adjusting the Data
</code></pre></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sig.pca <- prcomp(t(sig.data.norm), scale = TRUE)
sig.pc <- as.data.frame(sig.pca$x)
sig.pc$sample <- rownames(sig.pc)
sig.pc <- dplyr::select(as.tbl(sig.pc), sample, PC1, PC2, PC3, PC4)
sig.pc <- inner_join(sig.pc, info, by = "sample")</code></pre></div><div class="container-fluid"><div id="combatPCAplot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">ggplot(sig.pc, aes(x = PC1, y = PC2)) + geom_point(aes(colour = as.character(Sentrix_ID)), 
    size = 3) + theme_bw() + scale_colour_discrete("BeadChip")</code></pre><div class="row"><div id="fig:combatPCAplot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/combatPCAplot-1.png" alt="**Figure 8:** PCA plot of ComBat corrected gene expression."></img> </a></p><p class="caption"><strong>Figure 8:</strong> PCA plot of ComBat corrected gene expression.</p></div></div></div></div><div class="container-fluid"><div id="writeExpr" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">exprOut <- gzfile("/analysis/tmp/heatshock_expr_norm.tab.gz", open = "w")
write.table(sig.data.norm, file = exprOut, quote = FALSE)
close(exprOut)</code></pre></div></div></div></div><div id="differential-expression-analysis" class="section level2"><h2>Differential expression analysis</h2><p>All remaining samples were analysed for differences in gene expression between the basal stimulated states, pairing samples from the same individual, using the <em>limma</em> Bioconductor package <span class="citation">(M. E. Ritchie et al. 2015)</span>.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">labels <- strsplit(colnames(sig.data.norm), "_")
sample <- factor(sapply(labels, "[[", 1))
state <- factor(sapply(labels, "[[", 2), levels = c("Basal", "Stim"))
design <- model.matrix(~sample + state)
fit <- lmFit(sig.data.norm, design)
fit <- contrasts.fit(fit, coefficients = "stateStim")
fit <- eBayes(fit)
limmaTable <- topTable(fit, n = length(fit$coefficients))</code></pre></div><p>Individual probes were associated with corresponding genes using the updated annotations provided by the <em>illuminaHumanv3.db</em> Bioconductor package <span class="citation">(Barbosa-Morais et al. 2010)</span>.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">limmaTable <- cbind(ArrayAddress = rownames(limmaTable), limmaTable)
limmaTable <- inner_join(dplyr::select(annot, ArrayAddress, NuID, SymbolReannotated, 
    EntrezReannotated, GenomicLocation), limmaTable)</code></pre><button class="message R toggle btn btn-xs btn-info"><span class="glyphicon glyphicon-chevron-up"></span> R message</button><pre style="display:none"><code class="message r">## Joining by: "ArrayAddress"
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in inner_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">limmaTable <- limmaTable[order(limmaTable$P.Value), ]</code></pre></div><div class="container-fluid"><div id="writeDiffExpr" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">write.table(limmaTable, file = "/analysis/tmp/differential_expression.tab", 
    row.names = FALSE, quote = FALSE, sep = "\t")</code></pre></div></div><p>Gene N4BP2L2 is associated with two different probes that show evidence of differential expression. One appears to be significantly up regulated while the other is down regulated following heat shock (Table 3). We note that this is the only instance in which multiple probes for a differentially expressed gene provide conflicting evidence in this way. While it is possible that these probes measure two different transcripts that exhibit opposing responses to heat shock we chose to exclude these probes from further analysis to avoid confusion.</p><div class="container-fluid"><div id="N4BP2L2Table" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">set.caption(tabRef("N4BP2L2Table"), "Probes for N4BP2L2")
tab <- dplyr::filter(limmaTable, SymbolReannotated == "N4BP2L2")
tab <- dplyr::select(tab, NuID, SymbolReannotated, logFC, AveExpr, t, P.Value, 
    adj.P.Val, B)
pander(tab)</code></pre>Quitting from lines 1014-1018 (heatshock_analysis.Rmd)</div></div><div class="container-fluid"><div id="exludeInconsistent" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">limmaTable <- dplyr::filter(limmaTable, SymbolReannotated != "N4BP2L2")</code></pre></div></div></div><div id="go-enrichment-analysis" class="section level2"><h2>GO enrichment analysis</h2><p>To further investigate the patterns of differential expression a GO enrichment analysis was carried out using the Bioconductor package <em>topGO</em> <span class="citation">(Alexa and Rahnenfuhrer 2010)</span>. Fisher’s Exact Test was used to determine enrichment separately for significantly (FDR &lt; 0.01) up (Table 4) and down (Table 5) regulated genes.</p><div class="container-fluid"><div id="preTopGO" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">fdrCut <- 0.01
fcCut <- log2(1.2)

tg <- prep_topGO(limmaTable, fdrCut, fcCut)</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## 
## Building most specific GOs ..... ( 8253 GO terms found. )
## 
## Build GO DAG topology .......... ( 11983 GO terms and 28774 relations. )
## 
## Annotating nodes ............... ( 7963 genes annotated to the GO terms. )
## 
## Building most specific GOs ..... ( 8253 GO terms found. )
## 
## Build GO DAG topology .......... ( 11983 GO terms and 28774 relations. )
## 
## Annotating nodes ............... ( 7963 genes annotated to the GO terms. )
## 
## Building most specific GOs ..... ( 8253 GO terms found. )
## 
## Build GO DAG topology .......... ( 11983 GO terms and 28774 relations. )
## 
## Annotating nodes ............... ( 7963 genes annotated to the GO terms. )
</code></pre></div></div><div class="container-fluid"><div id="testTopGO" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">resultUp <- runTest(tg$up, algorithm = "classic", statistic = "fisher")</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## 
##           -- Classic Algorithm -- 
## 
##       the algorithm is scoring 2722 nontrivial nodes
##       parameters: 
##           test statistic:  fisher
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">resultDown <- runTest(tg$down, algorithm = "classic", statistic = "fisher")</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## 
##           -- Classic Algorithm -- 
## 
##       the algorithm is scoring 2752 nontrivial nodes
##       parameters: 
##           test statistic:  fisher
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">goTabUp <- GenTable(tg$up, up = resultUp, down = resultDown, topNodes = length(score(resultUp)), 
    orderBy = "up")
goTabUp[["up FDR"]] <- p.adjust(goTabUp$up)
goTabUp[["down FDR"]] <- p.adjust(goTabUp$down)
goTabDown <- GenTable(tg$down, down = resultDown, up = resultUp, topNodes = length(score(resultDown)), 
    orderBy = "down")
goTabDown[["up FDR"]] <- p.adjust(goTabDown$up)
goTabDown[["down FDR"]] <- p.adjust(goTabDown$down)</code></pre></div></div></div><div id="gene-annotions" class="section level2"><h2>Gene annotions</h2><p>In addition to identifying common themes amongst differentially expressed genes the question arises to what extend these genes have been previously associated with the heat shock or, more generally, stress response. We use the set of genes previously linked directly to heat shock <span class="citation">(Richter, Haslbeck, and Buchner 2010)</span> and from this create an extended set based on GO terms and PubMed articles linking differentially expressed genes to heat shock response and closely related processes.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">knownList <- list()
knownList$known <- readLines("data/heatshock_genes.lst")</code></pre></div><div id="gene-ontology-terms" class="section level3"><h3>Gene Ontology terms</h3><p>As a first step in highlighting genes not previously known to play a role in this context we identify all significantly up regulated genes that lack GO annotations of obvious relevance to heat shock response. In addition to terms related to stress response and protein folding we also include terms related to cell death and proliferation. To account for the presence of EBV in these cells we ignore all genes annotated with terms related to viral infections. Finally, any remaining genes related to regulation of gene expression are considered to be explained by the large scale changes in gene expression that are taking place in response to heat shock.</p><div class="container-fluid"><div id="goTerms" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sig.entrez <- dplyr::filter(limmaTable, !SymbolReannotated %in% knownList$known, 
    adj.P.Val < fdrCut, logFC > fcCut)$EntrezReannotated
mart <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host = "ensembl.org")
gene.data <- getBM(attributes = c("hgnc_symbol", "entrezgene", "go_id", "name_1006", 
    "namespace_1003"), filters = "entrezgene", values = sig.entrez, mart = mart)
gene.data <- dplyr::filter(gene.data, namespace_1003 == "biological_process")
termFilter_strict <- c("heat", "stress", "folding", "folded")
termFilter_viral <- c("virus", "viral", "immune")
termFilter_extended <- c("death", "apoptosis", "apoptotic", "cell cycle", "mitosis", 
    "mitotic", "proliferation", "transcription", "translation", "gene expression")
termFilter <- c(termFilter_strict, termFilter_extended)

strictGenes <- dplyr::filter(gene.data, str_detect(name_1006, paste(termFilter_strict, 
    collapse = "|")), !str_detect(name_1006, paste(termFilter_viral, collapse = "|")))
extendedGenes <- dplyr::filter(gene.data, str_detect(name_1006, paste(termFilter, 
    collapse = "|")), !str_detect(name_1006, paste(termFilter_viral, collapse = "|")))
knownList$strict <- strictGenes$hgnc_symbol
knownList$extended <- extendedGenes$hgnc_symbol
selGenes_strict <- dplyr::filter(limmaTable, adj.P.Val < fdrCut, logFC > fcCut, 
    !SymbolReannotated %in% knownList$strict, !SymbolReannotated %in% knownList$known)
selGenes <- dplyr::filter(limmaTable, adj.P.Val < fdrCut, logFC > fcCut, !SymbolReannotated %in% 
    knownList$extended, !SymbolReannotated %in% knownList$known)</code></pre></div></div><div class="container-fluid"><div id="goTermTotals" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">goKnownList <- knownList
gene.data_all <- getBM(attributes = c("hgnc_symbol", "entrezgene", "go_id", 
    "name_1006", "namespace_1003"), filters = "entrezgene", values = unique(limmaTable$EntrezReannotated), 
    mart = mart)
filterTotal_strict <- dplyr::filter(gene.data_all, str_detect(name_1006, paste(termFilter_strict, 
    collapse = "|")), !str_detect(name_1006, paste(termFilter_viral, collapse = "|")))
filterTotal_extended <- dplyr::filter(gene.data_all, str_detect(name_1006, paste(termFilter, 
    collapse = "|")), !str_detect(name_1006, paste(termFilter_viral, collapse = "|")))</code></pre></div></div><div class="container-fluid"><div id="goTermTest" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sig.genes <- with(limmaTable, adj.P.Val < fdrCut & logFC > fcCut)
genes_strict <- with(limmaTable, EntrezReannotated %in% filterTotal_strict$entrezgene)
genes_extended <- with(limmaTable, EntrezReannotated %in% filterTotal_extended$entrezgene)
go.test_strict <- fisher.test(sig.genes, genes_strict, alternative = "g")
go.test_extended <- fisher.test(sig.genes, genes_extended, alternative = "g")</code></pre></div></div></div><div id="pubmed" class="section level3"><h3>PubMed</h3><p>All genes not annotated with obvious GO terms were subjected to a PubMed search to find publications that link the gene to heat shock or stress response.</p><div class="container-fluid"><div id="pubmed" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">selGenes <- dplyr::filter(selGenes, SymbolReannotated != "")
selGenes_strict <- dplyr::filter(selGenes_strict, SymbolReannotated != "")
queryString <- paste(unique(selGenes_strict$SymbolReannotated), "AND (heat shock OR stress)")

query <- vector(mode = "list", length(queryString))
for (i in 1:length(query)) {
    if (i%%10 == 0) 
        Sys.sleep(1)
    query[[i]] <- tryCatch(EUtilsSummary(queryString[i]), error = function(e) {
        Sys.sleep(2)
        EUtilsSummary(queryString[i])
    })
}
names(query) <- unique(selGenes_strict$SymbolReannotated)
queryEmpty <- query[sapply(query, QueryCount) == 0]
queryFound <- query[sapply(query, QueryCount) != 0]


pmids <- vector(mode = "list", length(queryFound))
names(pmids) <- names(queryFound)
for (i in 1:length(pmids)) {
    if (i%%10 == 0) 
        Sys.sleep(1)
    pmids[[i]] <- tryCatch(PMID(EUtilsGet(queryFound[[i]])), error = function(e) {
        Sys.sleep(2)
        PMID(EUtilsGet(queryFound[[i]]))
    })
}
knownList$extended <- union(knownList$extended, names(queryFound))
knownList$strict <- union(knownList$strict, names(queryFound))
knownList$novel <- setdiff(names(queryEmpty), knownList$extended)

putativeGenes <- dplyr::filter(selGenes, SymbolReannotated %in% knownList$novel)
putativeGenes <- dplyr::left_join(putativeGenes, dplyr::select(gene.data, hgnc_symbol, 
    go_id, name_1006), by = c(SymbolReannotated = "hgnc_symbol"))
putativeGenes_strict <- dplyr::filter(selGenes_strict, SymbolReannotated %in% 
    names(queryEmpty))
putativeGenes_strict <- dplyr::left_join(putativeGenes_strict, dplyr::select(gene.data, 
    hgnc_symbol, go_id, name_1006), by = c(SymbolReannotated = "hgnc_symbol"))</code></pre></div></div><div class="container-fluid"><div id="geneRecords" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">write.table(data.frame(gene = names(queryFound), pmid = sapply(pmids, paste, 
    collapse = ",")), file = "tmp/heatshock_pmid.txt", row.names = FALSE, quote = FALSE, 
    sep = "\t")
write.table(putativeGenes, file = "tmp/heatshock_novel.tab", sep = "\t", quote = FALSE, 
    row.names = FALSE)
write.table(putativeGenes_strict, file = "tmp/heatshock_novel_strict.tab", sep = "\t", 
    quote = FALSE, row.names = FALSE)</code></pre></div></div></div><div id="heat-shock-factor-binding" class="section level3"><h3>Heat shock factor binding</h3><p>To further establish links between differentially expressed genes and heat shock response we consider the evidence for binding of HSF1 and HSF2 in the promoter regions of these genes. Using binding sites derived fro ChIP-seq data obtained from the K562 cell line <span class="citation">(Vihervaara et al. 2013)</span> we annotated our list of differentially expressed genes by cross-referencing it with the list HSF binding genes. Groups of genes corresponding to up or down regulated genes as well as those with existing heat shock related annotations and those without were tested for enrichment of HSF binding genes using Fisher’s exact test.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">peaks <- list.files(path = "/analysis/data", pattern = "HSF.*\\.txt", full.name = TRUE)

peaks2genes <- data.frame()
for (i in peaks) {
    nm <- gsub("/analysis/data/(.*).txt", "\\1", i)
    f <- read.table(i, sep = "\t", header = T)
    genes <- paste(f[, 11], collapse = ",")
    genes <- unique(strsplit(genes, ",|-")[[1]])
    genes[genes == ""] <- NA
    
    genes <- genes[!is.na(genes)]
    peaks2genes <- rbind(peaks2genes, data.frame(cond = nm, genes = genes))
}
peaks2genes$cond2 <- gsub("^M|^U", "comb", peaks2genes$cond)
peaks2genes$treatment <- str_extract(peaks2genes$cond2, "C|HS")
peaks2genes$tf <- str_extract(peaks2genes$cond2, "HSF.")

peakSplit <- by(peaks2genes, list(peaks2genes$treatment, peaks2genes$tf), function(x) x)
peakSplit[[1]] <- subset(peakSplit[[1]], genes %in% peakSplit[[3]]$genes)
if (nrow(peakSplit[[1]])) {
    peakSplit[[1]]$tf <- "HSF1+HSF2"
    peakSplit[[1]]$cond2 <- paste0(peakSplit[[1]]$cond2, "+HSF2")
}
peakSplit[[2]] <- subset(peakSplit[[2]], genes %in% peakSplit[[4]]$genes)
if (nrow(peakSplit[[2]])) {
    peakSplit[[2]]$tf <- "HSF1+HSF2"
    peakSplit[[2]]$cond2 <- paste0(peakSplit[[2]]$cond2, "+HSF2")
}

peaks2genes <- rbind(peaks2genes, peakSplit[[1]], peakSplit[[2]])</code></pre></div><div class="container-fluid"><div id="peakGroups" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">groupUp <- unique(subset(limmaTable, limmaTable$adj.P.Val < fdrCut & limmaTable$logFC > 
    fcCut)$SymbolReannotated)
groupDown <- unique(subset(limmaTable, limmaTable$adj.P.Val < fdrCut & limmaTable$logFC < 
    -fcCut)$SymbolReannotated)

totalNumberOfGenes <- length(unique(limmaTable$SymbolReannotated))
NofgroupUp <- length(groupUp)
NofgroupDown <- length(groupDown)

genes_in_both <- data.frame()
hsfPval <- data.frame()
for (g in unique(peaks2genes$cond2)) {
    anno_genes <- unique(subset(peaks2genes, cond2 == g)$genes)
    groupUp_in_anno_genes <- length(anno_genes[anno_genes %in% groupUp])
    groupUp_not_in_anno_genes <- length(anno_genes[!(anno_genes %in% groupUp)])
    
    groupDown_in_anno_genes <- length(anno_genes[anno_genes %in% groupDown])
    groupDown_not_in_anno_genes <- length(anno_genes[!(anno_genes %in% groupDown)])
    
    if (length(anno_genes[anno_genes %in% groupUp]) > 0) {
        genes_in_both <- rbind(genes_in_both, data.frame(cond = g, exp = "up", 
            genes = anno_genes[anno_genes %in% groupUp]))
    }
    
    if (length(anno_genes[anno_genes %in% groupDown]) > 0) {
        genes_in_both <- rbind(genes_in_both, data.frame(cond = g, exp = "down", 
            genes = anno_genes[anno_genes %in% groupDown]))
    }
    dat_up <- matrix(c(groupUp_in_anno_genes, groupUp_not_in_anno_genes, NofgroupUp, 
        totalNumberOfGenes - NofgroupUp), ncol = 2, byrow = TRUE)
    rst_up <- fisher.test(dat_up, alternative = "greater")
    
    dat_down <- matrix(c(groupDown_in_anno_genes, groupDown_not_in_anno_genes, 
        NofgroupDown, totalNumberOfGenes - NofgroupDown), ncol = 2)
    rst_down <- fisher.test(dat_down, alternative = "greater")
    
    
    hsfPval <- rbind(hsfPval, data.frame(cond = g, up = rst_up$p.value, down = rst_down$p.value, 
        up.or = rst_up$estimate, down.or = rst_down$estimate, sigGenes_up = groupUp_in_anno_genes, 
        sigGenes_down = groupDown_in_anno_genes))
}


genes_in_both$heatshock_novel_genes <- ifelse(genes_in_both$genes %in% knownList$novel, 
    2, ifelse(genes_in_both$genes %in% knownList$extended, 1, 0))</code></pre></div></div><p>In addition to the direct evidence from the ChIP-seq data we carried out a scan for the presence of HSF binding motifs in the promoter region (1200bp upstream - 300bp downstream of the TSS) of differentially expressed genes. The scan was based on the PWM defined by SwissRegulon <span class="citation">(Pachkov et al. 2007)</span> and carried out with the Bioconductor package PWMEnrich <span class="citation">(Stojnic and Diez 2014)</span>.</p><div class="container-fluid"><div id="prepPWMEnrich" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">txdb <- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene, columns = c("gene_id", 
    "tx_id", "tx_name"))
promoter.seqs <- getPromoterSeq(txdb, Hsapiens, upstream = 1200, downstream = 300)
names(promoter.seqs) <- as.character(txdb$gene_id)
promoter.seqs <- unique(promoter.seqs)
promoter.seqs_noNA <- promoter.seqs[!is.na(names(promoter.seqs))]

group_up <- dplyr::filter(limmaTable, adj.P.Val < fdrCut, logFC > fcCut)
group_dn <- dplyr::filter(limmaTable, adj.P.Val < fdrCut, logFC < -fcCut)
group_others <- dplyr::filter(limmaTable, !SymbolReannotated %in% group_up$SymbolReannotated, 
    !SymbolReannotated %in% group_dn$SymbolReannotated)

group_up_noNa <- subset(group_up, !is.na(EntrezReannotated))
group_dn_noNa <- subset(group_dn, !is.na(EntrezReannotated))
group_others_noNa <- subset(group_others, !is.na(EntrezReannotated))


promoter.seqs.group_up <- unique(promoter.seqs_noNA[names(promoter.seqs_noNA) %in% 
    group_up_noNa$EntrezReannotated])
promoter.seqs.group_dn <- unique(promoter.seqs_noNA[names(promoter.seqs_noNA) %in% 
    group_dn_noNa$EntrezReannotated])
promoter.seqs.group_others <- unique(promoter.seqs_noNA[names(promoter.seqs_noNA) %in% 
    group_others_noNa$EntrezReannotated])</code></pre></div></div><div class="container-fluid"><div id="pwmEnrich" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">registerCoresPWMEnrich(4)
motif.HSF1 <- readMotifs("data/HSF1_pwm.jaspar")
genomic.acgt = getBackgroundFrequencies("hg19")
pwms.denovo = toPWM(motif.HSF1, prior = genomic.acgt)
bg.denovo = makeBackground(pwms.denovo, organism = "hg19", type = "logn", quick = TRUE)</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## NOTE: Using the 'human' algorithm to infer background parameters,
##       appropriate only for human data.
## Parallel scanning with 4 cores
## Parameter estimation...
## Recording distribution properties for 250 bp tiles (this may take a while...)
## Estimating parameters for motif 1 / 1 
## Recording distribution properties for 500 bp tiles (this may take a while...)
## Estimating parameters for motif 1 / 1 
## Recording distribution properties for 1000 bp tiles (this may take a while...)
## Estimating parameters for motif 1 / 1 
## Recording distribution properties for 2000 bp tiles (this may take a while...)
## Estimating parameters for motif 1 / 1 
## Recording distribution properties for 4000 bp tiles (this may take a while...)
## Estimating parameters for motif 1 / 1
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">
r_promoter.seqs.group_up <- promoterScan(promoter.seqs.group_up, bg.denovo, 
    "up")</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## Parallel scanning with 4 cores
## Calculating motif enrichment scores ...
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">r_promoter.seqs.group_dn <- promoterScan(promoter.seqs.group_dn, bg.denovo, 
    "down")</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## Parallel scanning with 4 cores
## Calculating motif enrichment scores ...
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">r_promoter.seqs.group_others <- promoterScan(promoter.seqs.group_others, bg.denovo, 
    "others")</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## Parallel scanning with 4 cores
## Calculating motif enrichment scores ...
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">
r_promoter <- rbind(r_promoter.seqs.group_up, r_promoter.seqs.group_dn, r_promoter.seqs.group_others)
r_promoter <- dplyr::left_join(r_promoter, limmaTable, by = "EntrezReannotated")</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in left_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre></div></div><div class="container-fluid"><div id="pwmAnnotate" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">r_promoter$HSFpeak <- ifelse(r_promoter$SymbolReannotated %in% peaks2genes$genes, 
    1, 0)
r_promoter$novel <- ifelse(r_promoter$SymbolReannotated %in% knownList$novel, 
    3, ifelse(r_promoter$SymbolReannotated %in% knownList$extended & !r_promoter$SymbolReannotated %in% 
        knownList$strict, 2, ifelse(r_promoter$SymbolReannotated %in% knownList$strict, 
        1, 0)))

trx_numbers <- data.frame(table(r_promoter$SymbolReannotated))
colnames(trx_numbers) <- c("SymbolReannotated", "trx_numbers")
totalGeneNumber <- nrow(trx_numbers)

genePval <- do.call(rbind, by(r_promoter, r_promoter$SymbolReannotated, function(x) {
    x[which.min(x$p.value), , drop = FALSE]
}))
genePval$pwm.FDR <- p.adjust(genePval$p.value, "BH")
r_promoter <- dplyr::left_join(r_promoter, dplyr::select(genePval, EntrezReannotated, 
    pwm.FDR), by = "EntrezReannotated")
r_promoter <- dplyr::select(r_promoter, -rank, -target, -id)
r_promoter <- dplyr::select(r_promoter, ArrayAddress, NuID, EntrezReannotated, 
    SymbolReannotated, GenomicLocation:B, expr, novel, HSFpeak, raw.score, pwm.p.value = p.value, 
    pwm.FDR)
r_gene <- by(r_promoter, r_promoter$EntrezReannotated, function(x) {
    ans <- x[1, , drop = FALSE]
    if (nrow(x) > 1) {
        exprCols <- c("ArrayAddress", "NuID", "GenomicLocation", "logFC", "AveExpr", 
            "t", "P.Value", "adj.P.Val", "B", "expr")
        pwmCols <- c("raw.score", "pwm.p.value", "pwm.FDR")
        expr <- which.min(x$adj.P.Val)
        ans[1, exprCols] <- x[expr, exprCols]
        pwm <- which.min(x[, "pwm.p.value"])
        ans[1, pwmCols] <- x[pwm, pwmCols]
    }
    ans
})
r_gene <- do.call(rbind, r_gene)</code></pre></div></div><div class="container-fluid"><div id="peakOutput" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">write.table(hsfPval, "tmp/hsf_genes_pvalue.txt", sep = "\t", quote = FALSE, 
    row.names = FALSE)
write.table(genes_in_both, "tmp/hsf_genes.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(r_promoter, "tmp/hsf_motif_genes.txt", sep = "\t", quote = FALSE, 
    row.names = FALSE)
system("tar -czf tmp/gene_annotation.tar.gz -C tmp heatshock_pmid.txt heatshock_novel.tab heatshock_novel_strict.tab hsf_genes.txt hsf_genes_pvalue.txt hsf_motif_genes.txt")</code></pre></div></div></div></div><div id="global-heat-shock-response" class="section level2"><h2>Global heat shock response</h2><p>To assess the global difference in gene expression induced by heat shock we carried out partial least squares regression (PLS), using the treatment state (basal or stimulated) as a binary response variable and all gene expression probes that passed QC as explanatory variables.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">names <- strsplit(colnames(sig.data.norm), "_")
treat <- factor(sapply(names, "[[", 2), levels = c("Basal", "Stim"))
plsFit <- plsreg1(t(sig.data.norm), data.frame(treatment = as.integer(treat) - 
    1))
plsComp <- as.data.frame(plsFit$x.scores)
plsComp$sample <- sapply(names, "[[", 1)
plsComp$treatment <- treat</code></pre></div><p>To explore to what extend this response is modulated by genetic variation we used the length and direction of the response vector, i.e. the vector between the basal and stimulated sample for each individual in the space spanned by the first two PLS components, together with the location of the basal sample in the same space as a multivariate phenotype. This was then tested for association with genotypes for SNPs close to differentially expressed genes and their upstream regulators (identified via Ingenuity Pathway Analysis) using the <em>MultiPhen</em> R package <span class="citation">(O’Reilly 2012)</span>. The GRCh37 coordinates for SNPs were obtained via the <em>SNPlocs.Hsapiens.dbSNP142.GRCh37</em> Bioconductor package <span class="citation">(Pagès 2014)</span> and gene coordinates via the <em>TxDb.Hsapiens.UCSC.hg19.knownGene</em> package <span class="citation">(Carlson 2015)</span>.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">geno <- read_plink("tmp/hapmap_yri.smpl_flt3")
geno$bim <- dplyr::select(geno$bim, 2, 1, 4)
names(geno$bim) <- c("snp", "chrom", "pos")

geno$bim <- by(geno$bim, geno$bim$chrom, function(x) {
    locs <- getSNPlocs(paste0("ch", x$chrom[1]), caching = FALSE)
    locs$RefSNP_id <- paste0("rs", locs$RefSNP_id)
    snps <- inner_join(x, locs, by = c(snp = "RefSNP_id"))
    if (nrow(snps) < nrow(x)) 
        warning("Dropped ", nrow(x) - nrow(snps), " SNP(s) from chromosome ", 
            x$chrom[1])
    dplyr::select(snps, snp, chrom, loc)
})</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 243 SNP(s) from
## chromosome 1
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 251 SNP(s) from
## chromosome 2
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 159 SNP(s) from
## chromosome 3
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 154 SNP(s) from
## chromosome 4
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 141 SNP(s) from
## chromosome 5
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 1286 SNP(s) from
## chromosome 6
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 265 SNP(s) from
## chromosome 7
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 134 SNP(s) from
## chromosome 8
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 101 SNP(s) from
## chromosome 9
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 106 SNP(s) from
## chromosome 10
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 61 SNP(s) from
## chromosome 11
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 160 SNP(s) from
## chromosome 12
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 475 SNP(s) from
## chromosome 13
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 51 SNP(s) from
## chromosome 14
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 45 SNP(s) from
## chromosome 15
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 33 SNP(s) from
## chromosome 16
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 112 SNP(s) from
## chromosome 17
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 35 SNP(s) from
## chromosome 18
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 49 SNP(s) from
## chromosome 19
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 19 SNP(s) from
## chromosome 20
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 57 SNP(s) from
## chromosome 21
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in FUN(data[x, , drop = FALSE], ...): Dropped 121 SNP(s) from
## chromosome 22
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">
geno$bim <- do.call(rbind, geno$bim)
geno$bim$chrom <- paste0("chr", geno$bim$chrom)
geno$bim <- geno$bim[order(geno$bim$chrom, geno$bim$loc), ]</code></pre></div><div class="container-fluid"><div id="cisSNPs" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">cisWindow <- 10000

exclChrom <- c("chrX", "chrY", "chrM")
regs <- readr::read_tsv("data/upstream_regulators.tab", progres = FALSE)
regs[[1]] <- stringr::str_extract(regs[[1]], "\\S+")
regs[["FDR"]] <- p.adjust(regs[["P-value"]], method = "BH")
selectedGenes <- dplyr::filter(limmaTable, adj.P.Val < fdrCut & abs(logFC) > 
    fcCut & !str_detect(GenomicLocation, paste(exclChrom, collapse = "|")))
regs <- dplyr::filter(regs, FDR <= 0.01, !`Upstream Regulator` %in% selectedGenes$SymbolReannotated, 
    `Upstream Regulator` %in% limmaTable$SymbolReannotated)
selectedGenes <- rbind(selectedGenes, dplyr::filter(limmaTable, SymbolReannotated %in% 
    regs[["Upstream Regulator"]]))
selectedGenes <- dplyr::select(selectedGenes, ArrayAddress, NuID, GenomicLocation, 
    EntrezReannotated, SymbolReannotated)

selectedGenes$GenomicLocation <- str_extract(selectedGenes$GenomicLocation, 
    "chr[[:digit:]]{1,2}:[[:digit:]]+:[[:digit:]]+")
selectedGenes <- dplyr::filter(selectedGenes, !is.na(GenomicLocation) & !is.na(EntrezReannotated))

txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
txTable <- select(txdb, keys = selectedGenes$EntrezReannotated, columns = c("TXID", 
    "TXCHROM", "TXSTART", "TXEND"), keytype = "GENEID")
txTable <- dplyr::filter(txTable, !is.na(TXCHROM) & !str_detect(TXCHROM, "_"))
txTable <- by(txTable, txTable$GENEID, function(x) data.frame(entrez = x[["GENEID"]][1], 
    chrom = x[["TXCHROM"]][1], start = min(x[["TXSTART"]]), end = max(x[["TXEND"]])))
txTable <- do.call(rbind, txTable)
selectedGenes <- dplyr::right_join(selectedGenes, txTable, by = c(EntrezReannotated = "entrez"))</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in right_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">
probePos <- dplyr::select(selectedGenes, NuID, chrom, start, end)
probePos <- dplyr::arrange(probePos, chrom, start)

probePosGR <- makeGRangesFromDataFrame(probePos, ignore.strand = TRUE, keep.extra.columns = TRUE)
snpPosGR <- makeGRangesFromDataFrame(geno$bim, ignore.strand = TRUE, keep.extra.columns = TRUE, 
    start.field = "loc", end.field = "loc")
geneSNPs <- do.call(GRangesList, mapply(function(x, chr, s, e) subset(x, seqnames == 
    chr & start >= s - cisWindow & end <= e + cisWindow), as.character(seqnames(probePosGR)), 
    start(probePosGR), end(probePosGR), MoreArgs = list(x = snpPosGR), SIMPLIFY = FALSE))
names(geneSNPs) <- probePos$NuID
noSNP <- probePos$NuID[sapply(geneSNPs, function(x) length(x) == 0)]
snpRange <- unique(unlist(lapply(geneSNPs, function(x) x$snp)))</code></pre></div></div><div class="container-fluid"><div id="multiPheno" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">stimComp <- dplyr::filter(plsComp, treatment == "Stim")
basalComp <- dplyr::filter(plsComp, treatment == "Basal")
globalPheno <- data.frame(sample = stimComp$sample, basal.x = basalComp$t1, 
    basal.y = basalComp$t2, distance = sqrt((stimComp$t1 - basalComp$t1)^2 + 
        (stimComp$t2 - basalComp$t2)^2), direction = atan((stimComp$t2 - basalComp$t2)/(stimComp$t1 - 
        basalComp$t1)))
globalPheno <- dplyr::left_join(globalPheno, dplyr::select(geno$fam, V2, V5), 
    by = c(sample = "V2"))</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in left_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">globalPheno <- dplyr::rename(globalPheno, sex = V5)

globalPheno <- cbind(scale(dplyr::select(globalPheno, -sex, -sample)), sex = globalPheno$sex - 
    1)
rownames(globalPheno) <- stimComp$sample

rownames(geno$bed) <- str_extract(rownames(geno$bed), "NA[[:digit:]]+$")
cisGeno <- geno$bed[rownames(globalPheno), snpRange]
assoc <- mPhen(cisGeno, globalPheno, covariates = "sex")</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## [1] "excluding 0 samples based on exclusion criteria"
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">idx <- order(assoc$Results[1, , "JointModel", 2])
assoc$Results <- assoc$Results[, idx, , ]
assoc$maf <- assoc$maf[idx]
assoc$adj.p.val <- p.adjust(assoc$Results[, "JointModel", "pvalue"], "BH")</code></pre></div></div><p>Significance of the observed associations was assessed through a permutation test to account for the structure inherent to the data.</p><div class="container-fluid"><div id="multiPhenoPerm" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">set.seed(54864548L)
nPerm <- 1000
bins <- sapply(apply(cisGeno, 2, table), paste, collapse = "_")
names(bins) <- colnames(cisGeno)
pPerm <- matrix(NA, nrow = nPerm, ncol = length(unique(bins)))
colnames(pPerm) <- unique(bins)
binSNPs <- tapply(colnames(cisGeno), bins, function(x) x)
permGeno <- cisGeno[, sapply(binSNPs, "[", 1)]

for (i in 1:nPerm) {
    permPheno <- dplyr::sample_n(as.data.frame(globalPheno), nrow(globalPheno), 
        replace = FALSE)
    rownames(permPheno) <- rownames(globalPheno)
    assocPerm <- tryCatch(mPhen(permGeno, as.matrix(permPheno), covariates = "sex"), 
        error = function(e) {
            permPheno <- dplyr::sample_n(as.data.frame(globalPheno), nrow(globalPheno), 
                replace = FALSE)
            mPhen(permGeno, as.matrix(permPheno), covariates = "sex")
        })
    pPerm[i, ] <- assocPerm$Results[1, , "JointModel", 2]
}</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred

## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: algorithm did not converge
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
</code></pre></div></div><p>To this end the observed global response phenotype for each individual and covariates used in the model were randomly assigned to one of the observed set of genotypes 1000 times and p-values for the joint model were computed for each permutation. From these we computed false discovery rates by comparing observed p-values to the empirical distribution of minimum p-values from each permutation.</p><div class="container-fluid"><div id="permFDR" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">minp <- apply(pPerm, 1, min)
assoc$adj.p.val <- sapply(assoc$Results[, "JointModel", 2], function(x) sum(minp < 
    x)/length(minp))
names(assoc$adj.p.val) <- dimnames(assoc$Results)[[1]]
assoc$perm.p.val <- numeric(length(assoc$maf))
names(assoc$perm.p.val) <- dimnames(assoc$Results)[[1]]
for (i in 1:ncol(pPerm)) {
    assoc$perm.p.val[binSNPs[[i]]] <- sapply(assoc$Results[binSNPs[[i]], "JointModel", 
        2], function(x, p) sum(p < x)/length(p), pPerm[, i])
}</code></pre></div></div><p>Following up on the observed associations between SNPs and the global heat shock response we investigated possible associations between these SNPs and differentially expreressed genes implicated in the heat shock response. Here we define the gene set of interest as those genes that were significantly up regulated (FDR &lt; 0.01 and fold change &gt; 1.2) following heat shock and that were found to have HSF1 or HSF2 ChIP-seq peaks or binding motifs (P &lt; 0.05) in their promoter region.</p><div class="container-fluid"><div id="selectHeatShock" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">selSymbol <- dplyr::filter(r_gene, expr == "up", HSFpeak == 1 | pwm.p.value < 
    0.05)$SymbolReannotated
selSymbol <- union(selSymbol, c("CDKN1A", regs[["Upstream Regulator"]]))
selectedGenes <- dplyr::filter(limmaTable, SymbolReannotated %in% selSymbol)
probePos <- t(sapply(selectedGenes$GenomicLocation, str_extract, c("^chr([[:digit:]]+|X)(_[[:alnum:]]+_hap[[:digit:]]+)*", 
    ":[[:digit:]]+", "[[:digit:]]+:[+-]")))
rownames(probePos) <- selectedGenes$NuID
probePos[, 2] <- str_replace(probePos[, 2], ":", "")
probePos[, 3] <- str_replace_all(probePos[, 3], "[:+-]", "")
probePos <- cbind(rownames(probePos), probePos)</code></pre></div></div><div class="container-fluid"><div id="probeFC" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">selectedExpr <- sig.data.norm[selectedGenes$ArrayAddress, ]
basalIdx <- grep("_Basal", colnames(selectedExpr))
stimIdx <- grep("_Stim", colnames(selectedExpr))
selectedFC <- selectedExpr[, stimIdx] - selectedExpr[, basalIdx]
colnames(selectedFC) <- str_extract(colnames(selectedFC), "NA[[:digit:]]+")
rownames(selectedFC) <- selectedGenes$NuID</code></pre></div></div><div class="container-fluid"><div id="writeFcFile" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">write.table(selectedFC, file = "tmp/selected_probes_fc.tab", sep = "\t", quote = FALSE)</code></pre></div></div><div class="container-fluid"><div id="prepGeno" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">rownames(geno$bed) <- str_extract(rownames(geno$bed), "NA[[:digit:]]+")
geno$bed <- geno$bed[colnames(selectedFC), geno$bim$snp]
geno$bed <- t(geno$bed)
flip <- rowSums(geno$bed, na.rm = TRUE) > ncol(geno$bed)
if (any(flip)) {
    geno$bed[flip, ] <- -geno$bed[flip] + 2
}</code></pre></div></div><div class="container-fluid"><div id="selectSNPs" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">selGlobal <- dimnames(assoc$Result)[[1]][which(assoc$adj.p.val < 0.1)]
globalSNPs <- geno$bed[selGlobal[c(1, 3)], , drop = FALSE]</code></pre></div></div><div class="container-fluid"><div id="writeGlobalSNPs" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">write.table(globalSNPs, file = "tmp/global_geno.tab", sep = "\t", quote = FALSE)</code></pre></div></div><div class="container-fluid"><div id="selectedExpr" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">stimExpr <- selectedExpr[, stimIdx]
basalExpr <- selectedExpr[, basalIdx]
colnames(stimExpr) <- str_extract(colnames(stimExpr), "NA[[:digit:]]+")
colnames(basalExpr) <- str_extract(colnames(basalExpr), "NA[[:digit:]]+")
rownames(stimExpr) <- rownames(basalExpr) <- selectedGenes$NuID</code></pre></div></div><div class="container-fluid"><div id="writeSelExpr" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">write.table(stimExpr, file = "tmp/selected_probes_stim_expr.tab", sep = "\t", 
    quote = FALSE)
write.table(basalExpr, file = "tmp/selected_probes_basal_expr.tab", sep = "\t", 
    quote = FALSE)</code></pre></div></div><div class="container-fluid"><div id="fc_pc" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">nPC <- 1
if (nPC > 0) {
    pca <- prcomp(t(selectedFC), center = TRUE, scale = TRUE)
    pc <- pca$x
}</code></pre></div></div><div class="container-fluid"><div id="covariates" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">geno$fam <- dplyr::filter(dplyr::select(geno$fam, 2, 5), V2 %in% colnames(geno$bed))
geno$fam <- t(geno$fam)
colnames(geno$fam) <- geno$fam[1, ]
geno$fam <- geno$fam[, colnames(selectedFC)]
covarFC <- geno$fam[-1, , drop = FALSE]
if (nPC > 0) covarFC <- rbind(covarFC, t(pc[, 1:nPC]))</code></pre></div></div><div class="container-fluid"><div id="expr_pc" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">nPC_expr <- 2
if (nPC_expr > 0) {
    pca_stim <- prcomp(t(stimExpr), center = TRUE, scale = TRUE)
    pc_stim <- pca_stim$x
}
covarStim <- geno$fam[-1, , drop = FALSE]
if (nPC_expr > 0) covarStim <- rbind(covarStim, t(pc_stim[, 1:nPC_expr]))

if (nPC_expr > 0) {
    pca_basal <- prcomp(t(basalExpr), center = TRUE, scale = TRUE)
    pc_basal <- pca_basal$x
}
covarBasal <- geno$fam[-1, , drop = FALSE]
if (nPC_expr > 0) covarBasal <- rbind(covarBasal, t(pc_basal[, 1:nPC_expr]))</code></pre></div></div><div class="container-fluid"><div id="writeExprCovar" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">write.table(covarStim, file = "tmp/covar_expr_stim.tab", sep = "\t", quote = FALSE)
write.table(covarBasal, file = "tmp/covar_expr_basal.tab", sep = "\t", quote = FALSE)</code></pre></div></div><div class="container-fluid"><div id="prepGlobalQTL" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">snpsGlobal <- SlicedData$new()
snpsGlobal$fileDelimiter <- "\t"
snpsGlobal$fileOmitCharacters <- "NA"
snpsGlobal$fileSkipRows <- 1
snpsGlobal$fileSkipColumns <- 1
snpsGlobal$fileSliceSize <- 2000
snpsGlobal$LoadFile("tmp/global_geno.tab")

pheno <- SlicedData$new()
pheno$fileDelimiter <- "\t"
pheno$fileOmitCharacters <- "NA"
pheno$fileSkipRows <- 1
pheno$fileSkipColumns <- 1
pheno$fileSliceSize <- 2000
pheno$LoadFile("tmp/selected_probes_fc.tab")

covar <- SlicedData$new()
covar$fileDelimiter <- "\t"
covar$fileOmitCharacters <- "NA"
covar$fileSkipRows <- 1
covar$fileSkipColumns <- 1
covar$fileSliceSize <- 2000
covar$LoadFile("tmp/covariates.tab")</code></pre></div></div><div class="container-fluid"><div id="runGlobalMatrixEQTL" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">meGlobal <- Matrix_eQTL_main(snps = snpsGlobal, gene = pheno, cvrt = covar, 
    output_file_name = "tmp/reQTL_global_analysis", pvOutputThreshold = 1, useModel = modelLINEAR, 
    errorCovariance = numeric(), verbose = FALSE, pvalue.hist = FALSE, noFDRsaveMemory = FALSE)</code></pre></div></div><div class="container-fluid"><div id="prepGlobalEQTL" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">phenoStim <- SlicedData$new()
phenoStim$fileDelimiter <- "\t"
phenoStim$fileOmitCharacters <- "NA"
phenoStim$fileSkipRows <- 1
phenoStim$fileSkipColumns <- 1
phenoStim$fileSliceSize <- 2000
phenoStim$LoadFile("tmp/selected_probes_stim_expr.tab")

covarStim <- SlicedData$new()
covarStim$fileDelimiter <- "\t"
covarStim$fileOmitCharacters <- "NA"
covarStim$fileSkipRows <- 1
covarStim$fileSkipColumns <- 1
covarStim$fileSliceSize <- 2000
covarStim$LoadFile("tmp/covar_expr_stim.tab")

phenoBasal <- SlicedData$new()
phenoBasal$fileDelimiter <- "\t"
phenoBasal$fileOmitCharacters <- "NA"
phenoBasal$fileSkipRows <- 1
phenoBasal$fileSkipColumns <- 1
phenoBasal$fileSliceSize <- 2000
phenoBasal$LoadFile("tmp/selected_probes_basal_expr.tab")

covarBasal <- SlicedData$new()
covarBasal$fileDelimiter <- "\t"
covarBasal$fileOmitCharacters <- "NA"
covarBasal$fileSkipRows <- 1
covarBasal$fileSkipColumns <- 1
covarBasal$fileSliceSize <- 2000
covarBasal$LoadFile("tmp/covar_expr_basal.tab")</code></pre></div></div><div class="container-fluid"><div id="runGlobalEQTL" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">meGlobalStim <- Matrix_eQTL_main(snps = snpsGlobal, gene = phenoStim, cvrt = covarStim, 
    output_file_name = "tmp/eQTL_stim_global_analysis", pvOutputThreshold = 1, 
    useModel = modelLINEAR, errorCovariance = numeric(), verbose = FALSE, pvalue.hist = FALSE, 
    noFDRsaveMemory = FALSE)

meGlobalBasal <- Matrix_eQTL_main(snps = snpsGlobal, gene = phenoBasal, cvrt = covarBasal, 
    output_file_name = "tmp/eQTL_basal_global_analysis", pvOutputThreshold = 1, 
    useModel = modelLINEAR, errorCovariance = numeric(), verbose = FALSE, pvalue.hist = FALSE, 
    noFDRsaveMemory = FALSE)</code></pre></div></div><div id="interpreting-the-pls-components" class="section level3"><h3>Interpreting the PLS components</h3><p>Further considering the two PLS components it is apparent that the first component has a strong correspondence with the heat shock response, i.e. it separates basal from stimulated samples, whereas the interpretation of the second component is less obvious. We note that treated samples exhibit increased variance in the second component compared to basal samples. We note that the second component does reveal some structure, presenting three distinct clusters, within the stimulated samples but it is not immediately obvious what is underlying this grouping. In an attempted ti elucidate the drivers behind this structure we first assigned each stimulated sample to one of three clusters and then considered patterns of differential expression between them.</p><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">t2Stim <- plsComp$t2[plsComp$treatment == "Stim"]
names(t2Stim) <- plsComp$sample[plsComp$treatment == "Stim"]
plsClust <- hclust(dist(t2Stim))
cutHeight <- 40
plsClusters <- cut(as.dendrogram(plsClust), cutHeight)
clustID <- integer(length(t2Stim))
names(clustID) <- names(t2Stim)
for (i in 1:length(plsClusters$lower)) {
    clustID[labels(plsClusters$lower[[i]])] <- i
}
plsComp$cluster <- as.factor(rep(clustID, each = 2))</code></pre></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">labels <- strsplit(colnames(sig.data.norm), "_")
sample <- factor(sapply(labels, "[[", 1))
state <- factor(sapply(labels, "[[", 2), levels = c("Basal", "Stim"))
cluster <- factor(clustID[as.character(sample)])
stimIdx <- which(state == "Stim")

plsLimma_stim <- plsLimma(sig.data.norm[, stimIdx], cluster[stimIdx])</code></pre><button class="message R toggle btn btn-xs btn-info"><span class="glyphicon glyphicon-chevron-up"></span> R message</button><pre style="display:none"><code class="message r">## Joining by: "ArrayAddress"
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in inner_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="message R toggle btn btn-xs btn-info"><span class="glyphicon glyphicon-chevron-up"></span> R message</button><pre style="display:none"><code class="message r">## Joining by: "ArrayAddress"
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in inner_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="message R toggle btn btn-xs btn-info"><span class="glyphicon glyphicon-chevron-up"></span> R message</button><pre style="display:none"><code class="message r">## Joining by: "ArrayAddress"
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in inner_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">clust1_2Table <- plsLimma_stim$clust1_2Table

plsLimma_basal <- plsLimma(sig.data.norm[, -stimIdx], cluster[-stimIdx])</code></pre><button class="message R toggle btn btn-xs btn-info"><span class="glyphicon glyphicon-chevron-up"></span> R message</button><pre style="display:none"><code class="message r">## Joining by: "ArrayAddress"
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in inner_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="message R toggle btn btn-xs btn-info"><span class="glyphicon glyphicon-chevron-up"></span> R message</button><pre style="display:none"><code class="message r">## Joining by: "ArrayAddress"
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in inner_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="message R toggle btn btn-xs btn-info"><span class="glyphicon glyphicon-chevron-up"></span> R message</button><pre style="display:none"><code class="message r">## Joining by: "ArrayAddress"
</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in inner_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre></div><div class="container-fluid"><div id="writeClusterDiffExpr" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">write.table(clust1_2Table, file = "/analysis/tmp/cluster_1_2_diff_expression.tab", 
    row.names = FALSE, quote = FALSE, sep = "\t")</code></pre></div></div><div class="container-fluid"><div id="preTopGO_pls" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">tg_pls <- prep_topGO(clust1_2Table, fdrCut, fcCut)</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## 
## Building most specific GOs ..... ( 8254 GO terms found. )
## 
## Build GO DAG topology .......... ( 11984 GO terms and 28778 relations. )
## 
## Annotating nodes ............... ( 7964 genes annotated to the GO terms. )
## 
## Building most specific GOs ..... ( 8254 GO terms found. )
## 
## Build GO DAG topology .......... ( 11984 GO terms and 28778 relations. )
## 
## Annotating nodes ............... ( 7964 genes annotated to the GO terms. )
## 
## Building most specific GOs ..... ( 8254 GO terms found. )
## 
## Build GO DAG topology .......... ( 11984 GO terms and 28778 relations. )
## 
## Annotating nodes ............... ( 7964 genes annotated to the GO terms. )
</code></pre></div></div><div class="container-fluid"><div id="testTopGO_pls" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">resultPlsUp <- runTest(tg_pls$up, algorithm = "classic", statistic = "fisher")</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## 
##           -- Classic Algorithm -- 
## 
##       the algorithm is scoring 3373 nontrivial nodes
##       parameters: 
##           test statistic:  fisher
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">resultPlsDown <- runTest(tg_pls$down, algorithm = "classic", statistic = "fisher")</code></pre><button class="output R toggle btn btn-xs btn-success"><span class="glyphicon glyphicon-chevron-up"></span> R output</button><pre style="display:none"><code class="output r">## 
##           -- Classic Algorithm -- 
## 
##       the algorithm is scoring 3258 nontrivial nodes
##       parameters: 
##           test statistic:  fisher
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">goTabPlsUp <- GenTable(tg_pls$up, cluster1 = resultPlsUp, cluster2 = resultPlsDown, 
    topNodes = length(score(resultPlsUp)), orderBy = "cluster1")
goTabPlsDown <- GenTable(tg_pls$down, cluster2 = resultPlsDown, cluster1 = resultPlsUp, 
    topNodes = length(score(resultPlsDown)), orderBy = "cluster2")</code></pre></div></div></div></div></div><div id="results-1" class="section level1"><h1>Results</h1><div id="differential-expression-analysis-1" class="section level2"><h2>Differential expression analysis</h2><p>In total 1549 probes are identified as differentially expressed at an FDR threshold of 0.01 (see Table 6 for the top 20 results). Of these 701 are up and 848 down regulated following heat shock treatment. Further restricting this list by excluding all probes with a fold change of less than 1.2 leaves 500 probes with 249 and 251 up and down regulated respectively. Figure 9 shows a smoothed volcano plot highlighting probes of particular interest.</p><div class="container-fluid"><div id="limmaResult" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">set.caption(tabRef("topTable", "Top 20 differentially expressed probes."))
rownames(limmaTable) <- NULL
pander(head(dplyr::select(limmaTable, -ArrayAddress, -EntrezReannotated, -GenomicLocation), 
    20))</code></pre><table><caption><strong>Table 6:</strong> Top 20 differentially expressed probes.</caption><colgroup><col width="21%" /><col width="22%" /><col width="8%" /><col width="11%" /><col width="6%" /><col width="11%" /><col width="13%" /><col width="4%" /></colgroup><thead><tr class="header"><th align="center">NuID</th><th align="center">SymbolReannotated</th><th align="center">logFC</th><th align="center">AveExpr</th><th align="center">t</th><th align="center">P.Value</th><th align="center">adj.P.Val</th><th align="center">B</th></tr></thead><tbody><tr class="odd"><td align="center">Tiuh76h0KH_ee.1ztM</td><td align="center">HSPA1B</td><td align="center">4.47</td><td align="center">11.9</td><td align="center">53.9</td><td align="center">1.12e-52</td><td align="center">1.39e-48</td><td align="center">106</td></tr><tr class="even"><td align="center">xT97CkN_kRuqfFLpUo</td><td align="center">DNAJB1</td><td align="center">3.66</td><td align="center">10.9</td><td align="center">42.2</td><td align="center">1.82e-46</td><td align="center">1.13e-42</td><td align="center">93.4</td></tr><tr class="odd"><td align="center">i8lcq61QU7jt6JMo5M</td><td align="center">CLK1</td><td align="center">1.81</td><td align="center">10.8</td><td align="center">41.1</td><td align="center">8.42e-46</td><td align="center">3.48e-42</td><td align="center">92</td></tr><tr class="even"><td align="center">Wivj4OlfbjC0nuHtKk</td><td align="center">HSPA6</td><td align="center">4.2</td><td align="center">9.35</td><td align="center">29.9</td><td align="center">7.83e-38</td><td align="center">2.43e-34</td><td align="center">75.2</td></tr><tr class="odd"><td align="center">EXlCsv9_T.ndT0k7u4</td><td align="center">HSPH1</td><td align="center">1.23</td><td align="center">12.7</td><td align="center">23.3</td><td align="center">8.16e-32</td><td align="center">2.03e-28</td><td align="center">61.9</td></tr><tr class="even"><td align="center">rlU_JFNUR6eeN_.vio</td><td align="center">ZFAND2A</td><td align="center">1.92</td><td align="center">9.45</td><td align="center">23.2</td><td align="center">9.79e-32</td><td align="center">2.03e-28</td><td align="center">61.8</td></tr><tr class="odd"><td align="center">fq3hG6IHSKuXSqAgB0</td><td align="center">TNFSF14</td><td align="center">1.42</td><td align="center">9.37</td><td align="center">23</td><td align="center">1.42e-31</td><td align="center">2.52e-28</td><td align="center">61.4</td></tr><tr class="even"><td align="center">ElrogpXmo5QnoQpUj0</td><td align="center">HSPA6</td><td align="center">2.76</td><td align="center">7.78</td><td align="center">21.9</td><td align="center">2.33e-30</td><td align="center">3.61e-27</td><td align="center">58.7</td></tr><tr class="odd"><td align="center">r3kqEepjH6S4e9xfik</td><td align="center">FXR1</td><td align="center">0.933</td><td align="center">9.88</td><td align="center">20.8</td><td align="center">3.28e-29</td><td align="center">4.53e-26</td><td align="center">56.1</td></tr><tr class="even"><td align="center">Bnl1Unc1QXdUHMAcrk</td><td align="center">SERPINH1</td><td align="center">1.9</td><td align="center">8.73</td><td align="center">20.4</td><td align="center">8.6e-29</td><td align="center">1.07e-25</td><td align="center">55.2</td></tr><tr class="odd"><td align="center">B.7mrn_C5fk0qPKJDw</td><td align="center">TDG</td><td align="center">-0.876</td><td align="center">10.7</td><td align="center">-20.4</td><td align="center">1.06e-28</td><td align="center">1.2e-25</td><td align="center">55</td></tr><tr class="even"><td align="center">WXeRTp3eQUvd5NH554</td><td align="center">CCL3</td><td align="center">-1.29</td><td align="center">11.4</td><td align="center">-19.8</td><td align="center">5.07e-28</td><td align="center">5.25e-25</td><td align="center">53.4</td></tr><tr class="odd"><td align="center">63dUgevknAEiI3h56c</td><td align="center">KIAA0907</td><td align="center">0.94</td><td align="center">9.49</td><td align="center">19.5</td><td align="center">1.04e-27</td><td align="center">9.92e-25</td><td align="center">52.7</td></tr><tr class="even"><td align="center">rsFGq1N.xN6xEQSGdI</td><td align="center">HSPA4L</td><td align="center">0.917</td><td align="center">9.31</td><td align="center">19.2</td><td align="center">2.22e-27</td><td align="center">1.97e-24</td><td align="center">52</td></tr><tr class="odd"><td align="center">WqK.uImKeJcSOB396U</td><td align="center">JUN</td><td align="center">1.22</td><td align="center">9.22</td><td align="center">19</td><td align="center">3.51e-27</td><td align="center">2.9e-24</td><td align="center">51.5</td></tr><tr class="even"><td align="center">raspt5na.ip2mar9e8</td><td align="center">CACYBP</td><td align="center">0.911</td><td align="center">11.4</td><td align="center">18.9</td><td align="center">5.66e-27</td><td align="center">4.25e-24</td><td align="center">51.1</td></tr><tr class="odd"><td align="center">iRfeA55zKnrN7AIM14</td><td align="center">DNAJB4</td><td align="center">1.28</td><td align="center">7.16</td><td align="center">18.9</td><td align="center">5.82e-27</td><td align="center">4.25e-24</td><td align="center">51</td></tr><tr class="even"><td align="center">uUnkaFK9WQBPlJy4pM</td><td align="center">DNAJA4</td><td align="center">1.88</td><td align="center">8.83</td><td align="center">18.6</td><td align="center">1.18e-26</td><td align="center">8.17e-24</td><td align="center">50.3</td></tr><tr class="odd"><td align="center">6qPyUIkSH6PX3tXVRc</td><td align="center">IER5</td><td align="center">1.39</td><td align="center">10.5</td><td align="center">18.1</td><td align="center">4.29e-26</td><td align="center">2.81e-23</td><td align="center">49.1</td></tr><tr class="even"><td align="center">Qt4iQSFdNN7l6CL0e0</td><td align="center">LMAN2L</td><td align="center">0.769</td><td align="center">8.37</td><td align="center">18.1</td><td align="center">4.94e-26</td><td align="center">3.07e-23</td><td align="center">48.9</td></tr></tbody></table></div></div><div class="container-fluid"><div id="limmaPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">volcanoPlot(limmaTable, fcCut = fcCut, fdrCut = fdrCut)</code></pre><div class="row"><div id="fig:limmaPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/limmaPlot-1.png" alt="**Figure 9:** Volcano plot of differential expression results. Probes with an adjusted p-value below 0.01 and a log fold cange of at least 0.5 are shown as yellow and red dots. Probes showing particularly strong evidence of changes in gene expression through a combination of p-value and fold change are labelled with the corresponding gene symbol."></img> </a></p><p class="caption"><strong>Figure 9:</strong> Volcano plot of differential expression results. Probes with an adjusted p-value below 0.01 and a log fold cange of at least 0.5 are shown as yellow and red dots. Probes showing particularly strong evidence of changes in gene expression through a combination of p-value and fold change are labelled with the corresponding gene symbol.</p></div></div></div></div><div class="container-fluid"><div id="heatmap" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sortedTable <- dplyr::arrange(dplyr::filter(limmaTable, adj.P.Val < fdrCut, 
    abs(logFC) > fcCut), desc(logFC))
sideCols <- c(Basal = "purple", Stim = "orange")
treat <- str_extract(colnames(sig.data.norm), "(Basal|Stim)$")
heatmap.2(sig.data.norm[sortedTable$ArrayAddress, ], trace = "none", Rowv = FALSE, 
    Colv = TRUE, scale = "row", dendrogram = "column", col = rev(brewer.pal(11, 
        "RdBu")), ColSideColors = sideCols[treat], key = FALSE, labRow = "", 
    labCol = "", xlab = "Samples", ylab = "Genes")</code></pre><div class="row"><div id="fig:heatmap" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/heatmap-1.png" alt="**Figure 10:** Heatmap comparing gene expression for differentially expressed genes between basal and stimulated samples. Samples were clustered by gene  with stimulated (orange) and basal (purple) samples forming two distinct groups. Expression estimates for each gene were scaled and centred across samples. Blue cells correspond to lower and red cells to higher than average expression."></img> </a></p><p class="caption"><strong>Figure 10:</strong> Heatmap comparing gene expression for differentially expressed genes between basal and stimulated samples. Samples were clustered by gene with stimulated (orange) and basal (purple) samples forming two distinct groups. Expression estimates for each gene were scaled and centred across samples. Blue cells correspond to lower and red cells to higher than average expression.</p></div></div></div></div></div><div id="go-enrichment-analysis-1" class="section level2"><h2>GO enrichment analysis</h2><p>The GO analysis highlights several categories that are significantly enriched among the up and down regulated genes (7 and 0 categories with an FDR &lt; 0.05 respectively). Considering the top categories (Table 4 and Table 5) it appears that genes up regulated in response to heat shock are predominantly related to stress response (including <em>response to heat</em> [<a href="GO:0009408"><a href="http://amigo.geneontology.org/amigo/term/GO:0009408" target="_blank">GO:0009408</a></a>]) and protein folding (including <em>response to unfolded protein</em> [<a href="GO:0006986"><a href="http://amigo.geneontology.org/amigo/term/GO:0006986" target="_blank">GO:0006986</a></a>] and <em>protein refolding</em> [<a href="GO:0042026"><a href="http://amigo.geneontology.org/amigo/term/GO:0042026" target="_blank">GO:0042026</a></a>]) whereas the leading GO categories for down regulated genes broadly relate to cell proliferation and normal lymphocyte function and include <em>positive regulation of cell cycle</em> [<a href="GO:0045787"><a href="http://amigo.geneontology.org/amigo/term/GO:0045787" target="_blank">GO:0045787</a></a>], <em>spindle assembly</em> [<a href="GO:0051225"><a href="http://amigo.geneontology.org/amigo/term/GO:0051225" target="_blank">GO:0051225</a></a>] and <em>chemokine production</em> [<a href="GO:0032602"><a href="http://amigo.geneontology.org/amigo/term/GO:0032602" target="_blank">GO:0032602</a></a>] but are not sufficiently enriched to be considered significant.</p><div class="container-fluid"><div id="tableTopGOUp" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">set.caption(tabRef("topGOUp", "Top 20 GO categories enriched for up regulated genes."))
pander(head(dplyr::select(goTabUp, -up, -down), 20))</code></pre><table><caption><strong>Table 5:</strong> Top 20 GO categories enriched for up regulated genes.</caption><colgroup><col width="9%" /><col width="26%" /><col width="10%" /><col width="12%" /><col width="9%" /><col width="13%" /><col width="8%" /><col width="8%" /></colgroup><thead><tr class="header"><th align="center">GO.ID</th><th align="center">Term</th><th align="center">Annotated</th><th align="center">Significant</th><th align="center">Expected</th><th align="center">Rank in down</th><th align="center">up FDR</th><th align="center">down FDR</th></tr></thead><tbody><tr class="odd"><td align="center"><a href="GO:0009408"><a href="http://amigo.geneontology.org/amigo/term/GO:0009408" target="_blank">GO:0009408</a></a></td><td align="center">response to heat</td><td align="center">70</td><td align="center">13</td><td align="center">1.5</td><td align="center">2483</td><td align="center">0.000226</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0006986"><a href="http://amigo.geneontology.org/amigo/term/GO:0006986" target="_blank">GO:0006986</a></a></td><td align="center">response to unfolded protein</td><td align="center">97</td><td align="center">15</td><td align="center">2.09</td><td align="center">2753</td><td align="center">0.000273</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0006457"><a href="http://amigo.geneontology.org/amigo/term/GO:0006457" target="_blank">GO:0006457</a></a></td><td align="center">protein folding</td><td align="center">133</td><td align="center">17</td><td align="center">2.86</td><td align="center">1593</td><td align="center">0.00064</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0035966"><a href="http://amigo.geneontology.org/amigo/term/GO:0035966" target="_blank">GO:0035966</a></a></td><td align="center">response to topologically incorrect prot…</td><td align="center">106</td><td align="center">15</td><td align="center">2.28</td><td align="center">2362</td><td align="center">0.000939</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0009266"><a href="http://amigo.geneontology.org/amigo/term/GO:0009266" target="_blank">GO:0009266</a></a></td><td align="center">response to temperature stimulus</td><td align="center">96</td><td align="center">13</td><td align="center">2.06</td><td align="center">2257</td><td align="center">0.00981</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0042026"><a href="http://amigo.geneontology.org/amigo/term/GO:0042026" target="_blank">GO:0042026</a></a></td><td align="center">protein refolding</td><td align="center">11</td><td align="center">5</td><td align="center">0.24</td><td align="center">2754</td><td align="center">0.0277</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0034605"><a href="http://amigo.geneontology.org/amigo/term/GO:0034605" target="_blank">GO:0034605</a></a></td><td align="center">cellular response to heat</td><td align="center">50</td><td align="center">9</td><td align="center">1.07</td><td align="center">2755</td><td align="center">0.035</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0043618"><a href="http://amigo.geneontology.org/amigo/term/GO:0043618" target="_blank">GO:0043618</a></a></td><td align="center">regulation of transcription from RNA pol…</td><td align="center">35</td><td align="center">7</td><td align="center">0.75</td><td align="center">1904</td><td align="center">0.179</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:1900034"><a href="http://amigo.geneontology.org/amigo/term/GO:1900034" target="_blank">GO:1900034</a></a></td><td align="center">regulation of cellular response to heat</td><td align="center">26</td><td align="center">6</td><td align="center">0.56</td><td align="center">2756</td><td align="center">0.273</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0043620"><a href="http://amigo.geneontology.org/amigo/term/GO:0043620" target="_blank">GO:0043620</a></a></td><td align="center">regulation of DNA-templated transcriptio…</td><td align="center">39</td><td align="center">7</td><td align="center">0.84</td><td align="center">2008</td><td align="center">0.367</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0090083"><a href="http://amigo.geneontology.org/amigo/term/GO:0090083" target="_blank">GO:0090083</a></a></td><td align="center">regulation of inclusion body assembly</td><td align="center">10</td><td align="center">4</td><td align="center">0.21</td><td align="center">2757</td><td align="center">0.469</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0034976"><a href="http://amigo.geneontology.org/amigo/term/GO:0034976" target="_blank">GO:0034976</a></a></td><td align="center">response to endoplasmic reticulum stress</td><td align="center">124</td><td align="center">12</td><td align="center">2.67</td><td align="center">2687</td><td align="center">0.724</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0042542"><a href="http://amigo.geneontology.org/amigo/term/GO:0042542" target="_blank">GO:0042542</a></a></td><td align="center">response to hydrogen peroxide</td><td align="center">58</td><td align="center">8</td><td align="center">1.25</td><td align="center">2758</td><td align="center">0.809</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0080135"><a href="http://amigo.geneontology.org/amigo/term/GO:0080135" target="_blank">GO:0080135</a></a></td><td align="center">regulation of cellular response to stres…</td><td align="center">303</td><td align="center">20</td><td align="center">6.51</td><td align="center">1467</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0043066"><a href="http://amigo.geneontology.org/amigo/term/GO:0043066" target="_blank">GO:0043066</a></a></td><td align="center">negative regulation of apoptotic process</td><td align="center">425</td><td align="center">25</td><td align="center">9.14</td><td align="center">1267</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0070841"><a href="http://amigo.geneontology.org/amigo/term/GO:0070841" target="_blank">GO:0070841</a></a></td><td align="center">inclusion body assembly</td><td align="center">13</td><td align="center">4</td><td align="center">0.28</td><td align="center">2759</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0043069"><a href="http://amigo.geneontology.org/amigo/term/GO:0043069" target="_blank">GO:0043069</a></a></td><td align="center">negative regulation of programmed cell d…</td><td align="center">428</td><td align="center">25</td><td align="center">9.2</td><td align="center">1299</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0060548"><a href="http://amigo.geneontology.org/amigo/term/GO:0060548" target="_blank">GO:0060548</a></a></td><td align="center">negative regulation of cell death</td><td align="center">459</td><td align="center">26</td><td align="center">9.87</td><td align="center">563</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0036003"><a href="http://amigo.geneontology.org/amigo/term/GO:0036003" target="_blank">GO:0036003</a></a></td><td align="center">positive regulation of transcription fro…</td><td align="center">14</td><td align="center">4</td><td align="center">0.3</td><td align="center">2760</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:1902175"><a href="http://amigo.geneontology.org/amigo/term/GO:1902175" target="_blank">GO:1902175</a></a></td><td align="center">regulation of oxidative stress-induced i…</td><td align="center">15</td><td align="center">4</td><td align="center">0.32</td><td align="center">2761</td><td align="center">1</td><td align="center">1</td></tr></tbody></table></div></div><div class="container-fluid"><div id="tableTopGODown" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">set.caption(tabRef("topGODown", "Top 20 GO categories enriched for down regulated genes"))
pander(head(dplyr::select(goTabDown, -up, -down), 20))</code></pre><table><caption><strong>Table 6:</strong> Top 20 GO categories enriched for down regulated genes</caption><colgroup><col width="9%" /><col width="27%" /><col width="10%" /><col width="12%" /><col width="9%" /><col width="11%" /><col width="8%" /><col width="9%" /></colgroup><thead><tr class="header"><th align="center">GO.ID</th><th align="center">Term</th><th align="center">Annotated</th><th align="center">Significant</th><th align="center">Expected</th><th align="center">Rank in up</th><th align="center">up FDR</th><th align="center">down FDR</th></tr></thead><tbody><tr class="odd"><td align="center"><a href="GO:0051225"><a href="http://amigo.geneontology.org/amigo/term/GO:0051225" target="_blank">GO:0051225</a></a></td><td align="center">spindle assembly</td><td align="center">37</td><td align="center">6</td><td align="center">0.82</td><td align="center">2723</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0043207"><a href="http://amigo.geneontology.org/amigo/term/GO:0043207" target="_blank">GO:0043207</a></a></td><td align="center">response to external biotic stimulus</td><td align="center">342</td><td align="center">20</td><td align="center">7.56</td><td align="center">1917</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0051707"><a href="http://amigo.geneontology.org/amigo/term/GO:0051707" target="_blank">GO:0051707</a></a></td><td align="center">response to other organism</td><td align="center">342</td><td align="center">20</td><td align="center">7.56</td><td align="center">1918</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0007049"><a href="http://amigo.geneontology.org/amigo/term/GO:0007049" target="_blank">GO:0007049</a></a></td><td align="center">cell cycle</td><td align="center">1037</td><td align="center">45</td><td align="center">22.9</td><td align="center">2046</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0045931"><a href="http://amigo.geneontology.org/amigo/term/GO:0045931" target="_blank">GO:0045931</a></a></td><td align="center">positive regulation of mitotic cell cycl…</td><td align="center">64</td><td align="center">7</td><td align="center">1.41</td><td align="center">2724</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0007143"><a href="http://amigo.geneontology.org/amigo/term/GO:0007143" target="_blank">GO:0007143</a></a></td><td align="center">female meiotic division</td><td align="center">10</td><td align="center">3</td><td align="center">0.22</td><td align="center">2725</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0009607"><a href="http://amigo.geneontology.org/amigo/term/GO:0009607" target="_blank">GO:0009607</a></a></td><td align="center">response to biotic stimulus</td><td align="center">355</td><td align="center">20</td><td align="center">7.85</td><td align="center">1711</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0032496"><a href="http://amigo.geneontology.org/amigo/term/GO:0032496" target="_blank">GO:0032496</a></a></td><td align="center">response to lipopolysaccharide</td><td align="center">128</td><td align="center">10</td><td align="center">2.83</td><td align="center">1582</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:1903047"><a href="http://amigo.geneontology.org/amigo/term/GO:1903047" target="_blank">GO:1903047</a></a></td><td align="center">mitotic cell cycle process</td><td align="center">552</td><td align="center">27</td><td align="center">12.2</td><td align="center">2251</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0008219"><a href="http://amigo.geneontology.org/amigo/term/GO:0008219" target="_blank">GO:0008219</a></a></td><td align="center">cell death</td><td align="center">1022</td><td align="center">43</td><td align="center">22.6</td><td align="center">44</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0002237"><a href="http://amigo.geneontology.org/amigo/term/GO:0002237" target="_blank">GO:0002237</a></a></td><td align="center">response to molecule of bacterial origin</td><td align="center">134</td><td align="center">10</td><td align="center">2.96</td><td align="center">1662</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0016265"><a href="http://amigo.geneontology.org/amigo/term/GO:0016265" target="_blank">GO:0016265</a></a></td><td align="center">death</td><td align="center">1025</td><td align="center">43</td><td align="center">22.7</td><td align="center">36</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0001819"><a href="http://amigo.geneontology.org/amigo/term/GO:0001819" target="_blank">GO:0001819</a></a></td><td align="center">positive regulation of cytokine producti…</td><td align="center">180</td><td align="center">12</td><td align="center">3.98</td><td align="center">57</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0006353"><a href="http://amigo.geneontology.org/amigo/term/GO:0006353" target="_blank">GO:0006353</a></a></td><td align="center">DNA-templated transcription, termination</td><td align="center">57</td><td align="center">6</td><td align="center">1.26</td><td align="center">1559</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0000278"><a href="http://amigo.geneontology.org/amigo/term/GO:0000278" target="_blank">GO:0000278</a></a></td><td align="center">mitotic cell cycle</td><td align="center">627</td><td align="center">29</td><td align="center">13.9</td><td align="center">2236</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:1901654"><a href="http://amigo.geneontology.org/amigo/term/GO:1901654" target="_blank">GO:1901654</a></a></td><td align="center">response to ketone</td><td align="center">60</td><td align="center">6</td><td align="center">1.33</td><td align="center">688</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0022402"><a href="http://amigo.geneontology.org/amigo/term/GO:0022402" target="_blank">GO:0022402</a></a></td><td align="center">cell cycle process</td><td align="center">783</td><td align="center">34</td><td align="center">17.3</td><td align="center">2087</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0000226"><a href="http://amigo.geneontology.org/amigo/term/GO:0000226" target="_blank">GO:0000226</a></a></td><td align="center">microtubule cytoskeleton organization</td><td align="center">213</td><td align="center">13</td><td align="center">4.71</td><td align="center">2593</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0000070"><a href="http://amigo.geneontology.org/amigo/term/GO:0000070" target="_blank">GO:0000070</a></a></td><td align="center">mitotic sister chromatid segregation</td><td align="center">80</td><td align="center">7</td><td align="center">1.77</td><td align="center">2004</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0051726"><a href="http://amigo.geneontology.org/amigo/term/GO:0051726" target="_blank">GO:0051726</a></a></td><td align="center">regulation of cell cycle</td><td align="center">557</td><td align="center">26</td><td align="center">12.3</td><td align="center">1865</td><td align="center">1</td><td align="center">1</td></tr></tbody></table></div></div><div class="container-fluid"><div id="goPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">tblUp <- GenTable(tg$up, cluster1 = resultUp, topNodes = length(score(resultUp)))
tblUp$cluster1 <- as.numeric(tblUp$cluster1)
tblDown <- GenTable(tg$down, cluster2 = resultDown, topNodes = length(score(resultDown)))
tblDown$cluster2 <- as.numeric(tblDown$cluster2)
tbl <- bind_rows(gather(tblUp, "direction", "p.value", cluster1), gather(tblDown, 
    "direction", "p.value", cluster2))</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in rbind_all(list(x, ...)): Unequal factor levels: coercing to
## character
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">ggplot(tbl, aes(x = Expected, y = Significant)) + geom_point(data = dplyr::filter(tbl, 
    p.value >= 1e-04), colour = "grey") + geom_point(data = dplyr::filter(tbl, 
    p.value < 1e-04), aes(fill = -log10(p.value), shape = direction), size = 2.5) + 
    geom_abline(intercept = 0, slope = 1) + scale_shape_manual(values = c(up = 24, 
    down = 25)) + theme_bw()</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: Removed 10 rows containing missing values (geom_point).
</code></pre><div class="row"><div id="fig:goPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/goPlot-1.png" alt="**Figure 11:** GO category enrichment for differentially expressed genes."></img> </a></p><p class="caption"><strong>Figure 11:</strong> GO category enrichment for differentially expressed genes.</p></div></div></div></div></div><div id="gene-annotations" class="section level2"><h2>Gene annotations</h2><p>Of the 226 significantly up regulated genes 24 have been directly linked to the heat shock response and 98 are associated with GO terms that clearly relate to heat shock response (Fisher’s exact test P-value: <em>2.36e-10</em>) and a further 21 are otherwise linked to the heat shock response in the literature. This leaves 83 genes with no obvious association to heat shock. Using a more inclusive definition of heat shock related processes, also including GO terms related to cell death and proliferation, accounts for an additional 30 genes, leaving 53 genes without any known heatshock association.</p><div class="container-fluid"><div id="peakEnrichTab" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">condition <- str_replace(hsfPval$cond, "comb-C_", "basal ")
condition <- str_replace(condition, "comb-HS_", "heat shock ")
tab <- data.frame(Condition = condition, up = paste0(hsfPval$sigGenes_up, " (P=", 
    format(hsfPval$up, digits = 2), ", OR=", format(hsfPval$up.or, digits = 2), 
    ")"), down = paste0(hsfPval$sigGenes_down, " (P=", format(hsfPval$down, 
    digits = 2), ", OR=", format(hsfPval$down.or, digits = 2), ")"))
tab$up <- str_replace(tab$up, fixed("e+00"), "")
tab$down <- str_replace(tab$down, fixed("e+00"), "")
tab$up <- sub("\\(([[:digit:]]\\.[[:digit:]])e(-[[:digit:]]+)\\)", "\\$(\\1 \\\\times 10^{\\2})\\$", 
    tab$up)

set.caption(tabRef("peakEnrich", "Number (and corresponding p-values) of ChIP-seq peaks for HSF1 and HSF2 associated with significantly up and down regulated genes."))
highlightStrong <- NULL
if (any(hsfPval$up < 0.001)) highlightStrong <- cbind(which(hsfPval$up < 0.001), 
    2)
if (any(hsfPval$down < 0.001)) highlightStrong <- rbind(highlightStrong, cbind(which(hsfPval$down < 
    0.001), 3))

highlight <- NULL
if (any(hsfPval$up < 0.01 & hsfPval$up > 0.001)) highlight <- cbind(which(hsfPval$up < 
    0.01 & hsfPval$up > 0.001), 2)
if (any(hsfPval$down < 0.01 & hsfPval$down > 0.001)) highlight <- rbind(highlight, 
    cbind(which(hsfPval$down < 0.01 & hsfPval$down > 0.001), 3))

pander(tab, emphasize.cells = highlight, emphasize.strong.cells = highlightStrong)</code></pre><table><caption><strong>Table 7:</strong> Number (and corresponding p-values) of ChIP-seq peaks for HSF1 and HSF2 associated with significantly up and down regulated genes.</caption><colgroup><col width="29%" /><col width="38%" /><col width="27%" /></colgroup><thead><tr class="header"><th align="center">Condition</th><th align="center">up</th><th align="center">down</th></tr></thead><tbody><tr class="odd"><td align="center">basal HSF1</td><td align="center">1 (P=4.3e-01, OR= 1.8)</td><td align="center">0 (P=1.00, OR=0.00)</td></tr><tr class="even"><td align="center">basal HSF2</td><td align="center"><em>7 (P=9.6e-03, OR= 3.2)</em></td><td align="center">2 (P=0.71, OR=0.81)</td></tr><tr class="odd"><td align="center">heat shock HSF1</td><td align="center"><strong>51 (P=4.7e-10, OR= 3.0)</strong></td><td align="center">6 (P=1.00, OR=0.31)</td></tr><tr class="even"><td align="center">heat shock HSF2</td><td align="center"><strong>55 (P=9.4e-09, OR= 2.6)</strong></td><td align="center">10 (P=1.00, OR=0.43)</td></tr><tr class="odd"><td align="center">basal HSF1+HSF2</td><td align="center">1 (P=8.6e-02, OR=14.7)</td><td align="center">0 (P=1.00, OR=0.00)</td></tr><tr class="even"><td align="center">heat shock HSF1+HSF2</td><td align="center"><strong>46 (P=9.1e-15, OR= 4.5)</strong></td><td align="center">4 (P=1.00, OR=0.34)</td></tr></tbody></table></div></div><p>There are 7 up-regulated genes without any established role, 8 genes with limited evidence and 29 with substantial evidence for their involvement in the heat shock response that show heat shock factor binding in their promoter region. Further evidence for the binding of HSF1 and HSF2 in these regions is provided by scanning for the presence of the HSF1 binding motif. While this provides less direct evidence than the presence of a ChIP-seq peak and even a good match for the motif is no guarantee for the presence of an active binding site, it enables the detection of potential binding sites that may be inactive in the cell type or state studied through ChIP-seq experiments. Although overall the evidence for the presence of the HSF1 motif is relatively weak, there is a clear increase in the average score for genes that show evidence of HSF binding compared to those that don’t (Figure 12).</p><div class="container-fluid"><div id="pwmPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">ggplot(dplyr::filter(r_gene, expr == "up"), aes(x = factor(HSFpeak, levels = c("0", 
    "1"), labels = c("absent", "present")), y = -log10(pwm.p.value), fill = factor(novel, 
    levels = c("0", "1", "2", "3"), labels = c("established", "strong", "limited", 
        "novel")))) + geom_boxplot() + theme_bw() + xlab("HSF peak") + ylab(expression(-log[10]("p-value"))) + 
    scale_fill_discrete(name = "heat shock association")</code></pre><div class="row"><div id="fig:pwmPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/pwmPlot-1.png" alt="**Figure 12:** Distribution of p-values for HSF1 motif score."></img> </a></p><p class="caption"><strong>Figure 12:</strong> Distribution of p-values for HSF1 motif score.</p></div></div></div></div><div class="container-fluid"><div id="HSFtable" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">hsfTab <- data.frame(expression = r_gene$expr, ChIPseq = ifelse(r_gene$HSFpeak == 
    1, "yes", "no"), motif = ifelse(r_gene$pwm.p.value < 0.05, "present", "absent"))
tab <- tidyr::spread(as.data.frame(table(hsfTab)), motif, Freq)
tab[["frequency"]] <- paste0(sprintf("%0.2f", tab$present/(tab$present + tab$absent) * 
    100), "%")

upTest <- fisher.test(tab[tab$expression == "up", c("absent", "present")], alternative = "greater")
downTest <- fisher.test(tab[tab$expression == "down", c("absent", "present")], 
    alternative = "greater")

set.caption(tabRef("HSFtable", "Summary of HSF1 motif occurance in promoters of differentially expressed genes. with and without HSF ChIP-seq peaks."))
pander(tab)</code></pre><table><caption><strong>Table 8:</strong> Summary of HSF1 motif occurance in promoters of differentially expressed genes. with and without HSF ChIP-seq peaks.</caption><colgroup><col width="18%" /><col width="13%" /><col width="12%" /><col width="13%" /><col width="15%" /></colgroup><thead><tr class="header"><th align="center">expression</th><th align="center">ChIPseq</th><th align="center">absent</th><th align="center">present</th><th align="center">frequency</th></tr></thead><tbody><tr class="odd"><td align="center">up</td><td align="center">no</td><td align="center">155</td><td align="center">9</td><td align="center">5.49%</td></tr><tr class="even"><td align="center">up</td><td align="center">yes</td><td align="center">43</td><td align="center">17</td><td align="center">28.33%</td></tr><tr class="odd"><td align="center">down</td><td align="center">no</td><td align="center">206</td><td align="center">17</td><td align="center">7.62%</td></tr><tr class="even"><td align="center">down</td><td align="center">yes</td><td align="center">9</td><td align="center">3</td><td align="center">25.00%</td></tr><tr class="odd"><td align="center">others</td><td align="center">no</td><td align="center">7466</td><td align="center">776</td><td align="center">9.42%</td></tr><tr class="even"><td align="center">others</td><td align="center">yes</td><td align="center">407</td><td align="center">100</td><td align="center">19.72%</td></tr></tbody></table></div></div><p>Considering the frequency with wich HSF1 motifs occur (using a p-value threshold of 0.05) in the promoter regions of genes with and without ChIP-seq binding evidence (Table 8) it is apparent that the motif is enriched amongst up-regulated genes with corresponding protein binding evidence (Fisher’s exact test p-value: 1.2e-05). This enrichment is less pronounced for down regulated genes (0.071).</p><p>Further examining up-regulated genes for evidence of HSF binding reveals that 8 genes not previously characterised as heat shock response genes have ChIP-seq peaks and of these 1 also carry the HSF1 binding motif in their promoter (Table 9 and Table 10).</p><div class="container-fluid"><div id="HSFnovel" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">hsfTab$type <- ifelse(r_gene$novel == 2, "novel", ifelse(r_gene$novel == 1, 
    "limited", "established"))
hsfTab <- dplyr::filter(hsfTab, expression == "up") %>% dplyr::select(-expression)
tab <- tidyr::spread(as.data.frame(table(hsfTab)), type, Freq)

set.caption(tabRef("HSFnovel", "Summary of HSF binding evidence for the promoters of novel and established heat shock response genes."))
pander(tab, emphasize.strong.cells = matrix(c(3, 4, 5, 5), ncol = 2))</code></pre><table><caption><strong>Table 9:</strong> Summary of HSF binding evidence for the promoters of novel and established heat shock response genes.</caption><colgroup><col width="13%" /><col width="11%" /><col width="19%" /><col width="13%" /><col width="9%" /></colgroup><thead><tr class="header"><th align="center">ChIPseq</th><th align="center">motif</th><th align="center">established</th><th align="center">limited</th><th align="center">novel</th></tr></thead><tbody><tr class="odd"><td align="center">no</td><td align="center">absent</td><td align="center">49</td><td align="center">85</td><td align="center">21</td></tr><tr class="even"><td align="center">no</td><td align="center">present</td><td align="center">3</td><td align="center">5</td><td align="center">1</td></tr><tr class="odd"><td align="center">yes</td><td align="center">absent</td><td align="center">13</td><td align="center">23</td><td align="center"><strong>7</strong></td></tr><tr class="even"><td align="center">yes</td><td align="center">present</td><td align="center">10</td><td align="center">6</td><td align="center"><strong>1</strong></td></tr></tbody></table></div></div><div class="container-fluid"><div id="novelTable" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">set.caption(tabRef("novelTable", "Genes with newly established links to heat shock response"))
tab <- dplyr::filter(r_gene, novel == 2, HSFpeak == 1) %>% dplyr::select(`Gene Symbol` = SymbolReannotated, 
    `Fold Change` = logFC, `expression FDR` = adj.P.Val, `Motif p-value` = pwm.p.value) %>% 
    dplyr::arrange(desc(`Fold Change`), `Motif p-value`)
tab[["Fold Change"]] <- 2^tab[["Fold Change"]]
pander(tab)</code></pre><table><caption><strong>Table 10:</strong> Genes with newly established links to heat shock response</caption><colgroup><col width="19%" /><col width="19%" /><col width="23%" /><col width="23%" /></colgroup><thead><tr class="header"><th align="center">Gene Symbol</th><th align="center">Fold Change</th><th align="center">expression FDR</th><th align="center">Motif p-value</th></tr></thead><tbody><tr class="odd"><td align="center">NAA16</td><td align="center">1.57</td><td align="center">4.77e-17</td><td align="center">0.683</td></tr><tr class="even"><td align="center">USPL1</td><td align="center">1.54</td><td align="center">1.44e-14</td><td align="center">0.00444</td></tr><tr class="odd"><td align="center">FAM46A</td><td align="center">1.5</td><td align="center">3.92e-11</td><td align="center">0.32</td></tr><tr class="even"><td align="center">GPBP1</td><td align="center">1.36</td><td align="center">1.15e-08</td><td align="center">0.108</td></tr><tr class="odd"><td align="center">TYW3</td><td align="center">1.33</td><td align="center">1.82e-10</td><td align="center">0.0937</td></tr><tr class="even"><td align="center">EARS2</td><td align="center">1.31</td><td align="center">1.11e-12</td><td align="center">0.0716</td></tr><tr class="odd"><td align="center">MTCH1</td><td align="center">1.22</td><td align="center">5.38e-10</td><td align="center">0.127</td></tr><tr class="even"><td align="center">TRMT112</td><td align="center">1.22</td><td align="center">4.38e-07</td><td align="center">0.543</td></tr></tbody></table></div></div></div><div id="global-heat-shock-response-1" class="section level2"><h2>Global heat shock response</h2><p>Considering the outcome of the partial least squares regression it is apparent that there is a considerable change in overall gene expression in response to heat shock. Together the first two PLS components account for 96.1% in the variation observed due to heat shock and provide clear separation of the two treatment groups (Figure 13).</p><div class="container-fluid"><div id="treatPLSplot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">plsComp <- cbind(plsComp, sex = factor(globalPheno[plsComp$sample, "sex"]))
fig1 <- plsPlot(plsComp, group = c("treatment", "cluster"))
fig2 <- ggdendrogram(plsClust) + geom_hline(yint = cutHeight, col = "red")
multiplot(fig1, fig2, cols = 2)</code></pre><div class="row"><div id="fig:treatPLSplot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/treatPLSplot-1.png" alt="**Figure 13:** Samples separate clearly by treatment, showing a consistent global effect on gene expression from heat shock. Stimulated samples show evidence of three distinct groups."></img> </a></p><p class="caption"><strong>Figure 13:</strong> Samples separate clearly by treatment, showing a consistent global effect on gene expression from heat shock. Stimulated samples show evidence of three distinct groups.</p></div></div></div></div><p>Of the 19732 SNPs tested for association with the global response phenotype 2 are considered significant (FDR &lt; 0.05) (Table 11). These SNPs form 2 group of highly correlated genotypes, likely due to local LD structure.</p><div class="container-fluid"><div id="treatPLStable" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sel <- which(assoc$adj.p.val < 0.1)
coef <- assoc$Results[sel, , 1]
coef[, "JointModel"] <- assoc$adj.p.val[sel]
colnames(coef)[ncol(coef)] <- "Joint model FDR"
set.caption(tabRef("treatPLS", "Standardised coefficients and adjusted p-values for the top SNPs."))
pander(coef)</code></pre><table><caption><strong>Table 11:</strong> Standardised coefficients and adjusted p-values for the top SNPs.</caption><colgroup><col width="22%" /><col width="12%" /><col width="12%" /><col width="15%" /><col width="14%" /><col width="22%" /></colgroup><thead><tr class="header"><th align="center"> </th><th align="center">basal.x</th><th align="center">basal.y</th><th align="center">direction</th><th align="center">distance</th><th align="center">Joint model FDR</th></tr></thead><tbody><tr class="odd"><td align="center"><strong>rs10509407</strong></td><td align="center">-4.78</td><td align="center">-4.52</td><td align="center">-0.332</td><td align="center">-4.71</td><td align="center">0.021</td></tr><tr class="even"><td align="center"><strong>rs12259569</strong></td><td align="center">-4.78</td><td align="center">-4.85</td><td align="center">-0.576</td><td align="center">-4.51</td><td align="center">0.025</td></tr><tr class="odd"><td align="center"><strong>rs12207548</strong></td><td align="center">1.26</td><td align="center">0.452</td><td align="center">1.87</td><td align="center">0.925</td><td align="center">0.064</td></tr><tr class="even"><td align="center"><strong>rs7012294</strong></td><td align="center">-0.543</td><td align="center">-2.26</td><td align="center">-2.51</td><td align="center">-3.01</td><td align="center">0.066</td></tr></tbody></table></div></div><div class="container-fluid"><div id="treatPLSpval" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">
fig1 <- ggplot(data.frame(p.value = pPerm[, bins[dimnames(assoc$Result)[[1]][1]]]), 
    aes(x = -log10(p.value))) + geom_histogram(fill = "lightgrey", binwidth = 0.1) + 
    geom_vline(xintercept = -log10(assoc$Results[1, "JointModel", 2]), colour = "red") + 
    theme_bw() + xlab(expression(-log[10]("p-value")))

fig2 <- ggplot(data.frame(p.value = apply(pPerm, 1, min)), aes(x = -log10(p.value))) + 
    geom_histogram(fill = "lightgrey", binwidth = 0.1) + geom_vline(xintercept = -log10(assoc$Results[1, 
    "JointModel", 2]), colour = "red") + theme_bw() + xlab(expression(-log[10]("p-value")))

fig3 <- ggplot(data.frame(p.value = pPerm[, bins[dimnames(assoc$Result)[[1]][3]]]), 
    aes(x = -log10(p.value))) + geom_histogram(fill = "lightgrey", binwidth = 0.1) + 
    geom_vline(xintercept = -log10(assoc$Results[3, "JointModel", 2]), colour = "red") + 
    theme_bw() + xlab(expression(-log[10]("p-value")))

fig4 <- ggplot(data.frame(p.value = apply(pPerm, 1, min)), aes(x = -log10(p.value))) + 
    geom_histogram(fill = "lightgrey", binwidth = 0.1) + geom_vline(xintercept = -log10(assoc$Results[3, 
    "JointModel", 2]), colour = "red") + theme_bw() + xlab(expression(-log[10]("p-value")))


multiplot(fig1, fig2, fig3, fig4, layout = matrix(1:4, ncol = 2, byrow = TRUE))</code></pre><div class="row"><div id="fig:treatPLSpval" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/treatPLSpval-1.png" alt="**Figure 14:** P-values observed after permutation of global response phenotype for SNPs rs10509407 (top) and rs12207548 (bottom). The distribution of p-values for each snp after permutation is shown on the left and the minimum p-value observed in each iteration across all SNPs on the right."></img> </a></p><p class="caption"><strong>Figure 14:</strong> P-values observed after permutation of global response phenotype for SNPs rs10509407 (top) and rs12207548 (bottom). The distribution of p-values for each snp after permutation is shown on the left and the minimum p-value observed in each iteration across all SNPs on the right.</p></div></div></div></div><p>Figure 15 shows two examples of genotypes associating with the global heat shock response. Note that the second one (rs12207548) is located within a CTCF binding site immediately downstream of CDKN1A, an important regulator of cell cycle progression. This SNP is also found to be associated with the expession level of HSF1 in resting lymphocytes (Table 12) It may be interesting to note that allele frequencies vary markedly across populations. While the frequency of the ancestral T allele is 79% in the Yoruba 1000 genomes samples it is only observed with a frequency of 24% in European samples.</p><div class="container-fluid"><div id="plsGenoPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">genotypes <- data.frame(sample = colnames(geno$bed), stringsAsFactors = FALSE)
genotypes <- cbind(genotypes, t(geno$bed[dimnames(assoc$Results)[[1]][sel], 
    ]))
plsComp <- dplyr::left_join(plsComp, genotypes, by = "sample")

fig1 <- plsPlot(plsComp, dimnames(assoc$Results)[[1]][1]) + guides(colour = "none")
fig2 <- plsPlot(plsComp, dimnames(assoc$Results)[[1]][3]) + guides(colour = "none")
multiplot(fig1, fig2, cols = 2)</code></pre><div class="row"><div id="fig:plsGenoPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/plsGenoPlot-1.png" alt="**Figure 15:** Global response to heat shock by genotype for rs10509407 (left) and rs12207548 (right). Each individual is represented by two points corresponding to basal and stimulated state with arrows connecting paired samples. Genotypes are indicated by colour with blue corresponding to homozyguous carriers of the major allele and red indicating the presence of at least one minor allele."></img> </a></p><p class="caption"><strong>Figure 15:</strong> Global response to heat shock by genotype for rs10509407 (left) and rs12207548 (right). Each individual is represented by two points corresponding to basal and stimulated state with arrows connecting paired samples. Genotypes are indicated by colour with blue corresponding to homozyguous carriers of the major allele and red indicating the presence of at least one minor allele.</p></div></div></div></div><div class="container-fluid"><div id="globalSNPGenes_fc" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">tab <- meGlobal$all$eqtls
n <- max(sum(tab$FDR < 0.05), 5)
cut <- tab$FDR[n]
n <- min(max(which(tab$FDR <= cut)), 20)

tab <- dplyr::left_join(tab, dplyr::select(limmaTable, NuID, SymbolReannotated), 
    by = c(gene = "NuID"))</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in left_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">tab <- dplyr::rename(tab, Gene = SymbolReannotated)
tab <- dplyr::select(tab, SNP = snps, Gene, NuID = gene, Statistic = statistic, 
    `p-value` = pvalue, FDR, beta)
set.caption(tabRef("globalSNPGenes_fc", "Associations between global response SNPs and gene level heat shock response."))
pander(dplyr::select(head(tab, n), -NuID))</code></pre><table><caption><strong>Table 13:</strong> Associations between global response SNPs and gene level heat shock response.</caption><colgroup><col width="15%" /><col width="11%" /><col width="16%" /><col width="13%" /><col width="12%" /><col width="8%" /></colgroup><thead><tr class="header"><th align="center">SNP</th><th align="center">Gene</th><th align="center">Statistic</th><th align="center">p-value</th><th align="center">FDR</th><th align="center">beta</th></tr></thead><tbody><tr class="odd"><td align="center">rs12207548</td><td align="center">TNFRSF8</td><td align="center">-5.41</td><td align="center">3.21e-06</td><td align="center">0.000751</td><td align="center">-0.477</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">HSF1</td><td align="center">-5.35</td><td align="center">3.84e-06</td><td align="center">0.000751</td><td align="center">-0.643</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">EPHB1</td><td align="center">-5.25</td><td align="center">5.34e-06</td><td align="center">0.000751</td><td align="center">-0.532</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">SHC1</td><td align="center">-4.72</td><td align="center">2.92e-05</td><td align="center">0.00308</td><td align="center">-0.456</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">ZC3HAV1</td><td align="center">-4.6</td><td align="center">4.25e-05</td><td align="center">0.00358</td><td align="center">-0.399</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">ACBD3</td><td align="center">-4.17</td><td align="center">0.000157</td><td align="center">0.0102</td><td align="center">-0.279</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">UBQLN1</td><td align="center">-4.15</td><td align="center">0.000169</td><td align="center">0.0102</td><td align="center">-0.238</td></tr><tr class="even"><td align="center">rs10509407</td><td align="center">UBQLN1</td><td align="center">4.07</td><td align="center">0.000216</td><td align="center">0.0114</td><td align="center">0.232</td></tr></tbody></table></div></div><div class="container-fluid"><div id="globalSNPGenes_fcPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">n <- sum(tab$FDR < 0.05)
if (n < 2) n <- 2
fig <- vector(mode = "list", n)

for (i in 1:n) {
    snp <- as.character(tab$SNP[i])
    gene <- as.character(tab$Gene[i])
    probe <- as.character(tab$NuID[i])
    
    fit <- lm(selectedFC[probe, ] ~ geno$bed[snp, ] + t(covar[[1]]))
    offset <- colSums(fit$coefficients[-(1:2)] * covar[[1]])
    
    plotData <- data.frame(genotype = geno$bed[snp, ], fold_change = selectedFC[probe, 
        ] - offset)
    fig[[i]] <- ggplot(plotData, aes(x = factor(genotype), y = fold_change)) + 
        geom_point(position = position_jitter(width = 0.1), size = 2.5) + geom_boxplot(alpha = 0.4, 
        outlier.colour = "white", fill = "grey") + theme_bw() + xlab(snp) + 
        ylab(paste(gene, "(log fold change)"))
}
cols <- 1
if (n > 1) cols <- 2
if (n > 4) cols <- 3
multiplot(plotlist = fig, cols = cols)</code></pre><div class="row"><div id="fig:globalSNPGenes_fcPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/globalSNPGenes_fcPlot-1.png" alt="**Figure 16:** Heat shock response by genotype."></img> </a></p><p class="caption"><strong>Figure 16:</strong> Heat shock response by genotype.</p></div></div></div></div><div class="container-fluid"><div id="globalSNPGenes_stim" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">tab <- meGlobalStim$all$eqtls
cut <- tab$FDR[5]
n <- min(max(which(tab$FDR <= cut)), 20)
tab <- head(tab, n)
tab <- dplyr::left_join(tab, dplyr::select(limmaTable, NuID, SymbolReannotated), 
    by = c(gene = "NuID"))</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in left_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">tab <- dplyr::select(tab, -gene) %>% dplyr::rename(Gene = SymbolReannotated)
tab <- dplyr::select(tab, SNP = snps, Gene, Statistic = statistic, `p-value` = pvalue, 
    FDR, beta)
set.caption(tabRef("globalSNPGenes_stim", "Associations between global response SNPs and gene expression following heat shock."))
pander(tab)</code></pre><table><caption><strong>Table 14:</strong> Associations between global response SNPs and gene expression following heat shock.</caption><colgroup><col width="15%" /><col width="11%" /><col width="16%" /><col width="13%" /><col width="8%" /><col width="9%" /></colgroup><thead><tr class="header"><th align="center">SNP</th><th align="center">Gene</th><th align="center">Statistic</th><th align="center">p-value</th><th align="center">FDR</th><th align="center">beta</th></tr></thead><tbody><tr class="odd"><td align="center">rs12207548</td><td align="center">HSPA1B</td><td align="center">3.52</td><td align="center">0.00111</td><td align="center">0.467</td><td align="center">0.304</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">PPP5C</td><td align="center">2.8</td><td align="center">0.00787</td><td align="center">0.926</td><td align="center">0.204</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">RSRC2</td><td align="center">-2.65</td><td align="center">0.0116</td><td align="center">0.926</td><td align="center">-0.166</td></tr><tr class="even"><td align="center">rs10509407</td><td align="center">EGR1</td><td align="center">2.48</td><td align="center">0.0176</td><td align="center">0.926</td><td align="center">0.178</td></tr><tr class="odd"><td align="center">rs10509407</td><td align="center">HSPA4</td><td align="center">2.46</td><td align="center">0.0183</td><td align="center">0.926</td><td align="center">0.126</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">HSPA4</td><td align="center">2.39</td><td align="center">0.0218</td><td align="center">0.926</td><td align="center">0.153</td></tr><tr class="odd"><td align="center">rs10509407</td><td align="center">FAS</td><td align="center">-2.3</td><td align="center">0.0271</td><td align="center">0.926</td><td align="center">-0.147</td></tr><tr class="even"><td align="center">rs10509407</td><td align="center">E2F3</td><td align="center">2.19</td><td align="center">0.0342</td><td align="center">0.926</td><td align="center">0.114</td></tr><tr class="odd"><td align="center">rs10509407</td><td align="center">UBQLN1</td><td align="center">2.19</td><td align="center">0.0345</td><td align="center">0.926</td><td align="center">0.0915</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">HSF1</td><td align="center">-2.17</td><td align="center">0.0359</td><td align="center">0.926</td><td align="center">-0.171</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">PRR7</td><td align="center">-2.17</td><td align="center">0.0362</td><td align="center">0.926</td><td align="center">-0.177</td></tr><tr class="even"><td align="center">rs10509407</td><td align="center">MAPK14</td><td align="center">2.12</td><td align="center">0.04</td><td align="center">0.926</td><td align="center">0.106</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">PRKCE</td><td align="center">-2.08</td><td align="center">0.0445</td><td align="center">0.926</td><td align="center">-0.158</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">ZFAND2A</td><td align="center">2.06</td><td align="center">0.046</td><td align="center">0.926</td><td align="center">0.199</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">NEDD9</td><td align="center">2.05</td><td align="center">0.0472</td><td align="center">0.926</td><td align="center">0.141</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">SPR</td><td align="center">2.03</td><td align="center">0.0492</td><td align="center">0.926</td><td align="center">0.276</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">FAS</td><td align="center">2.02</td><td align="center">0.0504</td><td align="center">0.926</td><td align="center">0.163</td></tr><tr class="even"><td align="center">rs10509407</td><td align="center">TNFSF14</td><td align="center">-1.95</td><td align="center">0.0583</td><td align="center">0.926</td><td align="center">-0.171</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">SHC1</td><td align="center">-1.91</td><td align="center">0.0629</td><td align="center">0.926</td><td align="center">-0.164</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">GPR89A</td><td align="center">-1.91</td><td align="center">0.063</td><td align="center">0.926</td><td align="center">-0.0985</td></tr></tbody></table></div></div><div class="container-fluid"><div id="globalSNPGenes_basal" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">tab <- meGlobalBasal$all$eqtls
cut <- tab$FDR[5]
n <- min(max(which(tab$FDR <= cut)), 20)
tab <- head(tab, n)
tab <- dplyr::left_join(tab, dplyr::select(limmaTable, NuID, SymbolReannotated), 
    by = c(gene = "NuID"))</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in left_join_impl(x, y, by$x, by$y): joining character vector and
## factor, coercing into character vector
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">tab <- dplyr::select(tab, -gene) %>% dplyr::rename(Gene = SymbolReannotated)
tab <- dplyr::select(tab, SNP = snps, Gene, Statistic = statistic, `p-value` = pvalue, 
    FDR, beta)
set.caption(tabRef("globalSNPGenes_basal", "Associations between global response SNPs and gene expression in basal state."))
pander(tab)</code></pre><table><caption><strong>Table 12:</strong> Associations between global response SNPs and gene expression in basal state.</caption><colgroup><col width="15%" /><col width="11%" /><col width="16%" /><col width="13%" /><col width="9%" /><col width="9%" /></colgroup><thead><tr class="header"><th align="center">SNP</th><th align="center">Gene</th><th align="center">Statistic</th><th align="center">p-value</th><th align="center">FDR</th><th align="center">beta</th></tr></thead><tbody><tr class="odd"><td align="center">rs12207548</td><td align="center">HSF1</td><td align="center">4.23</td><td align="center">0.000138</td><td align="center">0.0582</td><td align="center">0.341</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">FOSL1</td><td align="center">3.58</td><td align="center">0.00094</td><td align="center">0.198</td><td align="center">0.159</td></tr><tr class="odd"><td align="center">rs10509407</td><td align="center">NR3C1</td><td align="center">-3.37</td><td align="center">0.0017</td><td align="center">0.239</td><td align="center">-0.154</td></tr><tr class="even"><td align="center">rs10509407</td><td align="center">UBQLN1</td><td align="center">-3.14</td><td align="center">0.00323</td><td align="center">0.311</td><td align="center">-0.14</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">ZC3HAV1</td><td align="center">3.01</td><td align="center">0.00461</td><td align="center">0.311</td><td align="center">0.203</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">SHC1</td><td align="center">2.96</td><td align="center">0.00522</td><td align="center">0.311</td><td align="center">0.182</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">TCP1</td><td align="center">2.92</td><td align="center">0.00581</td><td align="center">0.311</td><td align="center">0.186</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">BANP</td><td align="center">-2.86</td><td align="center">0.00677</td><td align="center">0.311</td><td align="center">-0.154</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">TRAF2</td><td align="center">2.85</td><td align="center">0.00686</td><td align="center">0.311</td><td align="center">0.162</td></tr><tr class="even"><td align="center">rs10509407</td><td align="center">MTCH1</td><td align="center">-2.75</td><td align="center">0.00897</td><td align="center">0.311</td><td align="center">-0.105</td></tr><tr class="odd"><td align="center">rs10509407</td><td align="center">USPL1</td><td align="center">2.72</td><td align="center">0.0096</td><td align="center">0.311</td><td align="center">0.119</td></tr><tr class="even"><td align="center">rs12207548</td><td align="center">GNB1</td><td align="center">-2.68</td><td align="center">0.0106</td><td align="center">0.311</td><td align="center">-0.132</td></tr><tr class="odd"><td align="center">rs12207548</td><td align="center">JUN</td><td align="center">-2.68</td><td align="center">0.0107</td><td align="center">0.311</td><td align="center">-0.266</td></tr><tr class="even"><td align="center">rs10509407</td><td align="center">PADI4</td><td align="center">2.68</td><td align="center">0.0108</td><td align="center">0.311</td><td align="center">0.121</td></tr><tr class="odd"><td align="center">rs10509407</td><td align="center">UBQLN1</td><td align="center">2.67</td><td align="center">0.0111</td><td align="center">0.311</td><td align="center">0.101</td></tr></tbody></table></div></div><p>Noting that the minor allele frequency of rs12207548 varies considerably between populations we thought to investigate wether there is evidence of selection at this locus. To this end we calculated F<sub>ST</sub> between the Yoruban and Central European populations for this SNP and compare it to all SNPs with similar MAF in the Yoruban samples.</p><div class="container-fluid"><div id="Fst" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">yri <- getSS("hmyriB36", paste0("", 1:22))
yri <- yri[, phenoData(yri)$isFounder == 1]
yri <- MAFfilter(yri, 0.17, 0.21)
geno.yri <- smList(yri)

ceu <- getSS("ceuhm3", paste0("chr", 1:22))
ceu <- ceu[, phenoData(ceu)$isFounder == 1]
geno.ceu <- smList(ceu)
geno.ceu <- mapply(function(x, y) {
    snps <- colnames(y)
    snps <- snps[snps %in% colnames(x)]
    x[, snps]
}, geno.ceu, geno.yri, SIMPLIFY = FALSE)
geno.yri <- mapply(function(x, y) {
    snps <- colnames(y)
    snps <- snps[snps %in% colnames(x)]
    x[, snps]
}, geno.yri, geno.ceu, SIMPLIFY = FALSE)
fst <- mapply(function(x, y) Fst(rbind(x, y), rep(c("ceu", "yri"), c(nrow(x), 
    nrow(y))))$Fst, geno.ceu, geno.yri, SIMPLIFY = FALSE)
fst <- unlist(fst)
names(fst) <- unlist(lapply(geno.yri, colnames))</code></pre></div></div><p>This estimates the F<sub>ST</sub> for rs12207548 to be 0.31. Of the 1175115 SNPs tested 10.55% have a more extreme value.</p><p>Investigating the differences in gene expression between cluster 1 and 2 of the second PLS component we identified a total of 1096 probes as differentially expressed at an FDR threshold of 0.01 (see Table 15 for the top 20 results). Of these 681 are up and 415 down regulated in cluster 2 compared to expression levels in cluster 1. Almost all of these (1094) have a fold change of at least 1.2. Figure 17 shows a smoothed volcano plot highlighting probes of particular interest.</p><div class="container-fluid"><div id="clusterTopTable" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">set.caption(tabRef("clusterTopTable", "Top 20 differentially expressed probes."))
rownames(clust1_2Table) <- NULL
pander(head(dplyr::select(clust1_2Table, -ArrayAddress, -EntrezReannotated, 
    -GenomicLocation), 20))</code></pre><table><caption><strong>Table 15:</strong> Top 20 differentially expressed probes.</caption><colgroup><col width="21%" /><col width="22%" /><col width="8%" /><col width="11%" /><col width="6%" /><col width="11%" /><col width="13%" /><col width="4%" /></colgroup><thead><tr class="header"><th align="center">NuID</th><th align="center">SymbolReannotated</th><th align="center">logFC</th><th align="center">AveExpr</th><th align="center">t</th><th align="center">P.Value</th><th align="center">adj.P.Val</th><th align="center">B</th></tr></thead><tbody><tr class="odd"><td align="center">Z5UZO6CcrqHT1w95b4</td><td align="center">MED22</td><td align="center">0.855</td><td align="center">9.28</td><td align="center">10.9</td><td align="center">1.69e-14</td><td align="center">2.1e-10</td><td align="center">22.6</td></tr><tr class="even"><td align="center">ltLUV5VtR.VN12swNM</td><td align="center">TOMM40</td><td align="center">0.93</td><td align="center">11.9</td><td align="center">10.2</td><td align="center">1.42e-13</td><td align="center">8.84e-10</td><td align="center">20.5</td></tr><tr class="odd"><td align="center">WV5XUSD6vUKnvUhB5Q</td><td align="center">GNAI2</td><td align="center">-1.09</td><td align="center">10.5</td><td align="center">-9.63</td><td align="center">8.95e-13</td><td align="center">3.7e-09</td><td align="center">18.8</td></tr><tr class="even"><td align="center">rlHpSl1Skrv870kMCs</td><td align="center">ZC3H3</td><td align="center">0.726</td><td align="center">8.77</td><td align="center">8.99</td><td align="center">7.42e-12</td><td align="center">1.62e-08</td><td align="center">16.8</td></tr><tr class="odd"><td align="center">HS7hLlJQ7kiujit_4A</td><td align="center">AIMP2</td><td align="center">0.779</td><td align="center">11.2</td><td align="center">8.98</td><td align="center">7.69e-12</td><td align="center">1.62e-08</td><td align="center">16.8</td></tr><tr class="even"><td align="center">B_gkHi0cdQCrXnUl6c</td><td align="center">C1orf112</td><td align="center">0.903</td><td align="center">8.81</td><td align="center">8.98</td><td align="center">7.81e-12</td><td align="center">1.62e-08</td><td align="center">16.7</td></tr><tr class="odd"><td align="center">x7leKemno5ulDACSFw</td><td align="center">INF2</td><td align="center">1.27</td><td align="center">9.56</td><td align="center">8.74</td><td align="center">1.75e-11</td><td align="center">3.11e-08</td><td align="center">16</td></tr><tr class="even"><td align="center">or3hBLFTVUSxV9J0v0</td><td align="center">BCL7B</td><td align="center">0.977</td><td align="center">8.13</td><td align="center">8.63</td><td align="center">2.59e-11</td><td align="center">4.02e-08</td><td align="center">15.6</td></tr><tr class="odd"><td align="center">WebhLrV3VaH5H71_.U</td><td align="center">PUF60</td><td align="center">0.967</td><td align="center">11.2</td><td align="center">8.42</td><td align="center">5.25e-11</td><td align="center">7.24e-08</td><td align="center">14.9</td></tr><tr class="even"><td align="center">rOezhdXiRW4eVTn9V8</td><td align="center">RANGAP1</td><td align="center">1.04</td><td align="center">10.9</td><td align="center">8.28</td><td align="center">8.53e-11</td><td align="center">9.84e-08</td><td align="center">14.5</td></tr><tr class="odd"><td align="center">Zhq6dHlPtFK_gd_Xe4</td><td align="center">NOP2</td><td align="center">0.776</td><td align="center">11.1</td><td align="center">8.27</td><td align="center">8.72e-11</td><td align="center">9.84e-08</td><td align="center">14.4</td></tr><tr class="even"><td align="center">9Xop_fqHop_foS9d7s</td><td align="center">HPS6</td><td align="center">0.937</td><td align="center">9.66</td><td align="center">8.22</td><td align="center">1.06e-10</td><td align="center">1.1e-07</td><td align="center">14.2</td></tr><tr class="odd"><td align="center">xKpnfhHuX0.1JUSXq0</td><td align="center">TYSND1</td><td align="center">0.88</td><td align="center">9.6</td><td align="center">8.18</td><td align="center">1.19e-10</td><td align="center">1.13e-07</td><td align="center">14.1</td></tr><tr class="even"><td align="center">lEdet6Oqeefi5Iqeaw</td><td align="center">SLC25A19</td><td align="center">1.23</td><td align="center">8.89</td><td align="center">8.06</td><td align="center">1.84e-10</td><td align="center">1.51e-07</td><td align="center">13.7</td></tr><tr class="odd"><td align="center">01OVCvVK5KEiu6XPsU</td><td align="center">C17orf53</td><td align="center">0.982</td><td align="center">8.75</td><td align="center">8.06</td><td align="center">1.84e-10</td><td align="center">1.51e-07</td><td align="center">13.7</td></tr><tr class="even"><td align="center">i15CjPunohOJCAJd6g</td><td align="center">LSP1</td><td align="center">-1.44</td><td align="center">8.31</td><td align="center">-8.04</td><td align="center">1.95e-10</td><td align="center">1.51e-07</td><td align="center">13.7</td></tr><tr class="odd"><td align="center">HzqJNpx567t9VBpGis</td><td align="center">CD74</td><td align="center">-1.49</td><td align="center">8.28</td><td align="center">-7.97</td><td align="center">2.52e-10</td><td align="center">1.84e-07</td><td align="center">13.4</td></tr><tr class="even"><td align="center">BesiCOkSjoplUIaEuw</td><td align="center">E4F1</td><td align="center">1.08</td><td align="center">9.03</td><td align="center">7.91</td><td align="center">3.11e-10</td><td align="center">2.15e-07</td><td align="center">13.2</td></tr><tr class="odd"><td align="center">NnJ1ITq_NFyKJ3pwpE</td><td align="center">NPRL3</td><td align="center">0.949</td><td align="center">10.1</td><td align="center">7.81</td><td align="center">4.35e-10</td><td align="center">2.75e-07</td><td align="center">12.9</td></tr><tr class="even"><td align="center">upeWHh7uk6JOnvVQuQ</td><td align="center">SNAPC4</td><td align="center">0.729</td><td align="center">8.9</td><td align="center">7.8</td><td align="center">4.5e-10</td><td align="center">2.75e-07</td><td align="center">12.9</td></tr></tbody></table></div></div><div class="container-fluid"><div id="clusterLimmaPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">volcanoPlot(clust1_2Table, fcCut = fcCut, fdrCut = fdrCut)</code></pre><div class="row"><div id="fig:clusterLimmaPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/clusterLimmaPlot-1.png" alt="**Figure 17:** Volcano plot of differential expression results between clusters 1 and 2. Probes with an adjusted p-value below 0.01 and a log fold cange of at least 0.5 are shown as yellow and red dots."></img> </a></p><p class="caption"><strong>Figure 17:</strong> Volcano plot of differential expression results between clusters 1 and 2. Probes with an adjusted p-value below 0.01 and a log fold cange of at least 0.5 are shown as yellow and red dots.</p></div></div></div></div><div class="container-fluid"><div id="clusterHeatmap" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">sortedTable <- dplyr::arrange(dplyr::filter(clust1_2Table, adj.P.Val < fdrCut, 
    abs(logFC) > fcCut), desc(logFC))
sideCols <- c(cluster1 = "purple", cluster2 = "orange", cluster3 = "white")
heatmap.2(sig.data.norm[sortedTable$ArrayAddress, plsComp$treatment == "Stim"], 
    trace = "none", Rowv = FALSE, Colv = as.dendrogram(plsClust), dendrogram = "column", 
    scale = "row", col = rev(brewer.pal(11, "RdBu")), ColSideColors = sideCols[plsComp$cluster][plsComp$treatment == 
        "Stim"], key = FALSE, labRow = "", labCol = "", xlab = "Samples", ylab = "Genes")</code></pre><div class="row"><div id="fig:clusterHeatmap" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/clusterHeatmap-1.png" alt="**Figure 18:** Heatmap comparing gene expression for genes differentially expressed between clusters 1 and 2. Expression estimates for each gene were scaled and centred across samples. Blue cells correspond to lower and red cells to higher than average expression."></img> </a></p><p class="caption"><strong>Figure 18:</strong> Heatmap comparing gene expression for genes differentially expressed between clusters 1 and 2. Expression estimates for each gene were scaled and centred across samples. Blue cells correspond to lower and red cells to higher than average expression.</p></div></div></div></div><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">goTabPlsUp[["cluster1 FDR"]] <- p.adjust(goTabPlsUp$cluster1)
goTabPlsUp[["cluster2 FDR"]] <- p.adjust(goTabPlsUp$cluster2)
goTabPlsDown[["cluster1 FDR"]] <- p.adjust(goTabPlsDown$cluster1)
goTabPlsDown[["cluster2 FDR"]] <- p.adjust(goTabPlsDown$cluster2)</code></pre></div><p>To further characterise the differnces in gene expression we carried out a GO enrichment analysis for genes differentially expressed between clusters 1 and 2. This reveals strong enrichment for several GO categories (21 and 6 enriched with FDR &lt; 0.05 in cluster 1 and cluster 2 respectively). Genes with increased expression in cluster 1 (Table 16) are enriched for biological processes related to DNA replication whereas those with increased expression in cluster 2 (Table 17) are annotated with processes related to the heat shock response.</p><div class="container-fluid"><div id="tablePlsTopGOUp" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">set.caption(tabRef("plsTopGOUp", "Top 20 GO categories enriched for genes up regulated in cluster 1"))
pander(dplyr::select(head(goTabPlsUp, 20), -cluster1, -cluster2))</code></pre><table><caption><strong>Table 18:</strong> Top 20 GO categories enriched for genes up regulated in cluster 1</caption><colgroup><col width="8%" /><col width="24%" /><col width="9%" /><col width="10%" /><col width="8%" /><col width="14%" /><col width="11%" /><col width="11%" /></colgroup><thead><tr class="header"><th align="center">GO.ID</th><th align="center">Term</th><th align="center">Annotated</th><th align="center">Significant</th><th align="center">Expected</th><th align="center">Rank in cluster2</th><th align="center">cluster1 FDR</th><th align="center">cluster2 FDR</th></tr></thead><tbody><tr class="odd"><td align="center"><a href="GO:0034660"><a href="http://amigo.geneontology.org/amigo/term/GO:0034660" target="_blank">GO:0034660</a></a></td><td align="center">ncRNA metabolic process</td><td align="center">257</td><td align="center">51</td><td align="center">15.8</td><td align="center">3245</td><td align="center">1.84e-06</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0034470"><a href="http://amigo.geneontology.org/amigo/term/GO:0034470" target="_blank">GO:0034470</a></a></td><td align="center">ncRNA processing</td><td align="center">170</td><td align="center">38</td><td align="center">10.5</td><td align="center">3252</td><td align="center">1.07e-05</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0022613"><a href="http://amigo.geneontology.org/amigo/term/GO:0022613" target="_blank">GO:0022613</a></a></td><td align="center">ribonucleoprotein complex biogenesis</td><td align="center">198</td><td align="center">41</td><td align="center">12.2</td><td align="center">3249</td><td align="center">2.69e-05</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0042254"><a href="http://amigo.geneontology.org/amigo/term/GO:0042254" target="_blank">GO:0042254</a></a></td><td align="center">ribosome biogenesis</td><td align="center">117</td><td align="center">29</td><td align="center">7.21</td><td align="center">3235</td><td align="center">7.26e-05</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0006310"><a href="http://amigo.geneontology.org/amigo/term/GO:0006310" target="_blank">GO:0006310</a></a></td><td align="center">DNA recombination</td><td align="center">146</td><td align="center">33</td><td align="center">9</td><td align="center">3159</td><td align="center">8.96e-05</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0006399"><a href="http://amigo.geneontology.org/amigo/term/GO:0006399" target="_blank">GO:0006399</a></a></td><td align="center">tRNA metabolic process</td><td align="center">96</td><td align="center">25</td><td align="center">5.92</td><td align="center">3259</td><td align="center">0.000247</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0006261"><a href="http://amigo.geneontology.org/amigo/term/GO:0006261" target="_blank">GO:0006261</a></a></td><td align="center">DNA-dependent DNA replication</td><td align="center">81</td><td align="center">21</td><td align="center">4.99</td><td align="center">2991</td><td align="center">0.00307</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0006271"><a href="http://amigo.geneontology.org/amigo/term/GO:0006271" target="_blank">GO:0006271</a></a></td><td align="center">DNA strand elongation involved in DNA re…</td><td align="center">25</td><td align="center">11</td><td align="center">1.54</td><td align="center">3260</td><td align="center">0.00469</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0022616"><a href="http://amigo.geneontology.org/amigo/term/GO:0022616" target="_blank">GO:0022616</a></a></td><td align="center">DNA strand elongation</td><td align="center">26</td><td align="center">11</td><td align="center">1.6</td><td align="center">3261</td><td align="center">0.00725</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0006281"><a href="http://amigo.geneontology.org/amigo/term/GO:0006281" target="_blank">GO:0006281</a></a></td><td align="center">DNA repair</td><td align="center">276</td><td align="center">45</td><td align="center">17</td><td align="center">3183</td><td align="center">0.00767</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0006260"><a href="http://amigo.geneontology.org/amigo/term/GO:0006260" target="_blank">GO:0006260</a></a></td><td align="center">DNA replication</td><td align="center">192</td><td align="center">35</td><td align="center">11.8</td><td align="center">2975</td><td align="center">0.00852</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0006364"><a href="http://amigo.geneontology.org/amigo/term/GO:0006364" target="_blank">GO:0006364</a></a></td><td align="center">rRNA processing</td><td align="center">86</td><td align="center">21</td><td align="center">5.3</td><td align="center">3205</td><td align="center">0.00895</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0000278"><a href="http://amigo.geneontology.org/amigo/term/GO:0000278" target="_blank">GO:0000278</a></a></td><td align="center">mitotic cell cycle</td><td align="center">627</td><td align="center">82</td><td align="center">38.7</td><td align="center">3222</td><td align="center">0.00937</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0016072"><a href="http://amigo.geneontology.org/amigo/term/GO:0016072" target="_blank">GO:0016072</a></a></td><td align="center">rRNA metabolic process</td><td align="center">88</td><td align="center">21</td><td align="center">5.43</td><td align="center">3210</td><td align="center">0.0132</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0006312"><a href="http://amigo.geneontology.org/amigo/term/GO:0006312" target="_blank">GO:0006312</a></a></td><td align="center">mitotic recombination</td><td align="center">28</td><td align="center">11</td><td align="center">1.73</td><td align="center">3262</td><td align="center">0.0175</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0031396"><a href="http://amigo.geneontology.org/amigo/term/GO:0031396" target="_blank">GO:0031396</a></a></td><td align="center">regulation of protein ubiquitination</td><td align="center">134</td><td align="center">27</td><td align="center">8.26</td><td align="center">2639</td><td align="center">0.0183</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0051340"><a href="http://amigo.geneontology.org/amigo/term/GO:0051340" target="_blank">GO:0051340</a></a></td><td align="center">regulation of ligase activity</td><td align="center">70</td><td align="center">18</td><td align="center">4.32</td><td align="center">3139</td><td align="center">0.0213</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0006259"><a href="http://amigo.geneontology.org/amigo/term/GO:0006259" target="_blank">GO:0006259</a></a></td><td align="center">DNA metabolic process</td><td align="center">532</td><td align="center">71</td><td align="center">32.8</td><td align="center">2681</td><td align="center">0.023</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:1903047"><a href="http://amigo.geneontology.org/amigo/term/GO:1903047" target="_blank">GO:1903047</a></a></td><td align="center">mitotic cell cycle process</td><td align="center">552</td><td align="center">72</td><td align="center">34</td><td align="center">3157</td><td align="center">0.0425</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0051438"><a href="http://amigo.geneontology.org/amigo/term/GO:0051438" target="_blank">GO:0051438</a></a></td><td align="center">regulation of ubiquitin-protein transfer…</td><td align="center">67</td><td align="center">17</td><td align="center">4.13</td><td align="center">3132</td><td align="center">0.0468</td><td align="center">1</td></tr></tbody></table></div></div><div class="container-fluid"><div id="tablePlsTopGODown" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">set.caption(tabRef("plsTopGODown", "Top 20 GO categories enriched for genes up regulated in cluster 2"))
pander(dplyr::select(head(goTabPlsDown, 20), -cluster1, -cluster2))</code></pre><table><caption><strong>Table 19:</strong> Top 20 GO categories enriched for genes up regulated in cluster 2</caption><colgroup><col width="8%" /><col width="24%" /><col width="9%" /><col width="10%" /><col width="8%" /><col width="14%" /><col width="11%" /><col width="11%" /></colgroup><thead><tr class="header"><th align="center">GO.ID</th><th align="center">Term</th><th align="center">Annotated</th><th align="center">Significant</th><th align="center">Expected</th><th align="center">Rank in cluster1</th><th align="center">cluster1 FDR</th><th align="center">cluster2 FDR</th></tr></thead><tbody><tr class="odd"><td align="center"><a href="GO:0034976"><a href="http://amigo.geneontology.org/amigo/term/GO:0034976" target="_blank">GO:0034976</a></a></td><td align="center">response to endoplasmic reticulum stress</td><td align="center">124</td><td align="center">20</td><td align="center">4.45</td><td align="center">2591</td><td align="center">1</td><td align="center">0.00683</td></tr><tr class="even"><td align="center"><a href="GO:0030968"><a href="http://amigo.geneontology.org/amigo/term/GO:0030968" target="_blank">GO:0030968</a></a></td><td align="center">endoplasmic reticulum unfolded protein r…</td><td align="center">74</td><td align="center">15</td><td align="center">2.66</td><td align="center">2180</td><td align="center">1</td><td align="center">0.00726</td></tr><tr class="odd"><td align="center"><a href="GO:0034620"><a href="http://amigo.geneontology.org/amigo/term/GO:0034620" target="_blank">GO:0034620</a></a></td><td align="center">cellular response to unfolded protein</td><td align="center">74</td><td align="center">15</td><td align="center">2.66</td><td align="center">2181</td><td align="center">1</td><td align="center">0.00726</td></tr><tr class="even"><td align="center"><a href="GO:0006986"><a href="http://amigo.geneontology.org/amigo/term/GO:0006986" target="_blank">GO:0006986</a></a></td><td align="center">response to unfolded protein</td><td align="center">97</td><td align="center">17</td><td align="center">3.48</td><td align="center">2381</td><td align="center">1</td><td align="center">0.0128</td></tr><tr class="odd"><td align="center"><a href="GO:0035967"><a href="http://amigo.geneontology.org/amigo/term/GO:0035967" target="_blank">GO:0035967</a></a></td><td align="center">cellular response to topologically incor…</td><td align="center">81</td><td align="center">15</td><td align="center">2.91</td><td align="center">1932</td><td align="center">1</td><td align="center">0.0243</td></tr><tr class="even"><td align="center"><a href="GO:0035966"><a href="http://amigo.geneontology.org/amigo/term/GO:0035966" target="_blank">GO:0035966</a></a></td><td align="center">response to topologically incorrect prot…</td><td align="center">106</td><td align="center">17</td><td align="center">3.81</td><td align="center">1780</td><td align="center">1</td><td align="center">0.0427</td></tr><tr class="odd"><td align="center"><a href="GO:0032075"><a href="http://amigo.geneontology.org/amigo/term/GO:0032075" target="_blank">GO:0032075</a></a></td><td align="center">positive regulation of nuclease activity</td><td align="center">49</td><td align="center">11</td><td align="center">1.76</td><td align="center">1838</td><td align="center">1</td><td align="center">0.064</td></tr><tr class="even"><td align="center"><a href="GO:0032069"><a href="http://amigo.geneontology.org/amigo/term/GO:0032069" target="_blank">GO:0032069</a></a></td><td align="center">regulation of nuclease activity</td><td align="center">53</td><td align="center">11</td><td align="center">1.9</td><td align="center">1992</td><td align="center">1</td><td align="center">0.141</td></tr><tr class="odd"><td align="center"><a href="GO:0006987"><a href="http://amigo.geneontology.org/amigo/term/GO:0006987" target="_blank">GO:0006987</a></a></td><td align="center">activation of signaling protein activity…</td><td align="center">46</td><td align="center">10</td><td align="center">1.65</td><td align="center">1692</td><td align="center">1</td><td align="center">0.213</td></tr><tr class="even"><td align="center"><a href="GO:0080135"><a href="http://amigo.geneontology.org/amigo/term/GO:0080135" target="_blank">GO:0080135</a></a></td><td align="center">regulation of cellular response to stres…</td><td align="center">303</td><td align="center">30</td><td align="center">10.9</td><td align="center">1060</td><td align="center">1</td><td align="center">0.554</td></tr><tr class="odd"><td align="center"><a href="GO:1903573"><a href="http://amigo.geneontology.org/amigo/term/GO:1903573" target="_blank">GO:1903573</a></a></td><td align="center">negative regulation of response to endop…</td><td align="center">18</td><td align="center">6</td><td align="center">0.65</td><td align="center">3374</td><td align="center">1</td><td align="center">0.554</td></tr><tr class="even"><td align="center"><a href="GO:0009101"><a href="http://amigo.geneontology.org/amigo/term/GO:0009101" target="_blank">GO:0009101</a></a></td><td align="center">glycoprotein biosynthetic process</td><td align="center">206</td><td align="center">22</td><td align="center">7.4</td><td align="center">3105</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0018196"><a href="http://amigo.geneontology.org/amigo/term/GO:0018196" target="_blank">GO:0018196</a></a></td><td align="center">peptidyl-asparagine modification</td><td align="center">91</td><td align="center">13</td><td align="center">3.27</td><td align="center">3279</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0018279"><a href="http://amigo.geneontology.org/amigo/term/GO:0018279" target="_blank">GO:0018279</a></a></td><td align="center">protein N-linked glycosylation via aspar…</td><td align="center">91</td><td align="center">13</td><td align="center">3.27</td><td align="center">3280</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:1900101"><a href="http://amigo.geneontology.org/amigo/term/GO:1900101" target="_blank">GO:1900101</a></a></td><td align="center">regulation of endoplasmic reticulum unfo…</td><td align="center">15</td><td align="center">5</td><td align="center">0.54</td><td align="center">3375</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0031344"><a href="http://amigo.geneontology.org/amigo/term/GO:0031344" target="_blank">GO:0031344</a></a></td><td align="center">regulation of cell projection organizati…</td><td align="center">185</td><td align="center">20</td><td align="center">6.65</td><td align="center">2556</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:1902275"><a href="http://amigo.geneontology.org/amigo/term/GO:1902275" target="_blank">GO:1902275</a></a></td><td align="center">regulation of chromatin organization</td><td align="center">83</td><td align="center">12</td><td align="center">2.98</td><td align="center">538</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0006487"><a href="http://amigo.geneontology.org/amigo/term/GO:0006487" target="_blank">GO:0006487</a></a></td><td align="center">protein N-linked glycosylation</td><td align="center">95</td><td align="center">13</td><td align="center">3.41</td><td align="center">3288</td><td align="center">1</td><td align="center">1</td></tr><tr class="odd"><td align="center"><a href="GO:0045664"><a href="http://amigo.geneontology.org/amigo/term/GO:0045664" target="_blank">GO:0045664</a></a></td><td align="center">regulation of neuron differentiation</td><td align="center">201</td><td align="center">21</td><td align="center">7.22</td><td align="center">2316</td><td align="center">1</td><td align="center">1</td></tr><tr class="even"><td align="center"><a href="GO:0045860"><a href="http://amigo.geneontology.org/amigo/term/GO:0045860" target="_blank">GO:0045860</a></a></td><td align="center">positive regulation of protein kinase ac…</td><td align="center">273</td><td align="center">26</td><td align="center">9.81</td><td align="center">3170</td><td align="center">1</td><td align="center">1</td></tr></tbody></table></div></div><div class="container-fluid"><div id="goPlsPlot" class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">tblUp <- GenTable(tg_pls$up, cluster1 = resultPlsUp, topNodes = length(score(resultPlsUp)))
tblUp$cluster1 <- as.numeric(tblUp$cluster1)
tblDown <- GenTable(tg_pls$down, cluster2 = resultPlsDown, topNodes = length(score(resultPlsDown)))
tblDown$cluster2 <- as.numeric(tblDown$cluster2)
tbl <- bind_rows(gather(tblUp, "direction", "p.value", cluster1), gather(tblDown, 
    "direction", "p.value", cluster2))</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning in rbind_all(list(x, ...)): Unequal factor levels: coercing to
## character
</code></pre><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">ggplot(tbl, aes(x = Expected, y = Significant)) + geom_point(data = dplyr::filter(tbl, 
    p.value >= 1e-04), colour = "grey") + geom_point(data = dplyr::filter(tbl, 
    p.value < 1e-04), aes(fill = -log10(p.value), shape = direction), size = 2.5) + 
    geom_abline(intercept = 0, slope = 1) + scale_shape_manual(values = c(up = 24, 
    down = 25)) + theme_bw()</code></pre><button class="warning R toggle btn btn-xs btn-warning"><span class="glyphicon glyphicon-chevron-up"></span> R warning</button><pre style="display:none"><code class="warning r">## Warning: Removed 51 rows containing missing values (geom_point).
</code></pre><div class="row"><div id="fig:goPlsPlot" class="figure col-md-offset-3 col-md-6"><p><a href="#" class="thumbnail"><img src="heatshock_analysis_files/figure-html/goPlsPlot-1.png" alt="**Figure 19:** GO category enrichment for genes differentially expressed between PLS clusters."></img> </a></p><p class="caption"><strong>Figure 19:</strong> GO category enrichment for genes differentially expressed between PLS clusters.</p></div></div></div></div></div></div><div id="appendix" class="section level1 unnumbered"><h1>Appendix</h1><div id="custom-functions-used" class="section level2"><h2>Custom functions used</h2><div class="row"><button class="source R toggle btn btn-xs btn-primary"><span class="glyphicon glyphicon-chevron-up"></span> R source</button><pre style="display:none"><code class="source r">reshapeData <- function(data, value) {
    ans <- reshape2::melt(data, id.vars = "probe", value.name = value, variable.name = "sample")
    splitLabel <- strsplit(as.character(ans$sample), "_", fixed = TRUE)
    ans$sample <- sapply(splitLabel, "[", 1)
    ans$treatment <- sapply(splitLabel, function(x) x[length(x)])
    as.tbl(ans[c("sample", "treatment", "probe", value)])
}

ibsClust <- function(genome, ...) {
    dst <- spread(dplyr::select(genome, IID1, IID2, DST), IID2, DST)
    size <- nrow(dst) + 1
    labels <- c(as.character(dst$IID1[1]), names(dst)[-1])
    dst <- t(1 - as.matrix(dst[-1]))
    dst <- dst[!is.na(dst)]
    attr(dst, "Size") <- size
    attr(dst, "Labels") <- labels
    attr(dst, "Diag") <- FALSE
    attr(dst, "Upper") <- FALSE
    class(dst) <- "dist"
    
    hclust(dst, ...)
}

pcaClusterPlot <- function(data, i, j, fillColour, legend = TRUE) {
    data <- data[, c("sample", paste0("PC", i), paste0("PC", j), "Sentrix_ID", 
        "cluster")]
    names(data) <- c("sample", "x", "y", "Sentrix_ID", "cluster")
    clusters <- by(data, data$cluster, function(x) x)
    fig <- ggplot(data, aes(x = x, y = y))
    for (cl in clusters) {
        ch <- chull(cl$x, cl$y)
        fig <- fig + geom_polygon(data = cl[ch, ], fill = fillColour[cl$cluster[1]], 
            alpha = 0.7)
    }
    fig <- fig + geom_point(aes(colour = as.character(Sentrix_ID)), size = 3) + 
        theme_bw() + scale_colour_discrete("BeadChip") + xlab(paste0("PC", i)) + 
        ylab(paste0("PC", j))
    if (!legend) 
        fig <- fig + theme(legend.position = "none")
    fig
}

adjustLabel <- function(x, y, centre) {
    just <- c(h = 0, v = 0)
    if (x < centre[1]) {
        just["h"] <- just["h"] + 0.9
    }
    if (y < centre[2]) {
        just["v"] <- just["v"] + 1
    }
    just
}

countBlocks <- function(geno, pval, pcut = 0.05, rcut = 0.8) {
    count <- 0
    selected <- which(pval < pcut)
    if (length(selected)) {
        if (length(selected) == 1) {
            count <- 1
        } else {
            if (names(pval)[1] %in% colnames(geno)) 
                geno <- geno[, names(pval)[selected]] else geno <- t(geno[names(pval)[selected], ])
            r2 <- cor(geno, use = "pairwise", method = "spearman")^2
            remain <- 1:ncol(r2)
            while (length(remain)) {
                i <- remain[1]
                count <- count + 1
                remain <- intersect(which(r2[i, ] < rcut), remain)
            }
        }
    }
    count
}

## Create a scatter plot with one panel per group (by default BeadChip) Data
## points (by default samples in PC space) are assumed to be from a bivariate
## t distribution. Confidence ellipses are drawn based on this (at a 95%
## level by default) and points outside these ellipses are marked as
## outliers.
outlierPlot <- function(data, x = "PC1", y = "PC2", group = "Sentrix_ID", shape = "Treatment", 
    id = "Cell Line", level = 0.95, ...) {
    ellipses <- by(data[, c(x, y)], data[[group]], function(d) tryCatch(car::dataEllipse(as.matrix(d), 
        levels = level, draw = FALSE, robust = TRUE), error = function(e) cbind(NA, 
        NA)))
    data[[group]] <- as.character(data[[group]])
    
    names(data)[which(names(data) == x)] <- make.names(x)
    x <- make.names(x)
    names(data)[which(names(data) == y)] <- make.names(y)
    y <- make.names(y)
    names(data)[which(names(data) == group)] <- make.names(group)
    group <- make.names(group)
    names(data)[which(names(data) == shape)] <- make.names(shape)
    shape <- make.names(shape)
    names(data)[which(names(data) == id)] <- make.names(id)
    id <- make.names(id)
    
    data$outlier <- logical(nrow(data))
    for (i in 1:nrow(data)) {
        data$outlier[i] <- !sp::point.in.polygon(data[i, x], data[i, y], ellipses[[as.character(data[i, 
            group])]][, 1], ellipses[[as.character(data[i, group])]][, 2])
    }
    data$colour <- as.character(data[[group]])
    data$colour[data$outlier] <- NA
    centre <- by(data[, c(x, y)], data[[group]], colMeans)
    data$h <- if (!is.null(list(...)$hjust)) 
        list(...)$hjust else 0
    data$v <- if (!is.null(list(...)$vjust)) 
        list(...)$vjust else 0
    labelJust <- mapply(function(x, y, g, c) adjustLabel(x, y, c[[g]]), data[[x]], 
        data[[y]], data[[group]], MoreArgs = list(centre))
    data$h <- ifelse(labelJust["h", ] > 0, -data$h, data$h)
    data$h <- data$h + labelJust["h", ]
    data$v <- ifelse(labelJust["v", ] > 0, -data$v, data$v)
    data$v <- data$v + labelJust["v", ]
    ggplot(data, aes_string(x = x, y = y, colour = "colour")) + geom_point(aes_string(shape = shape), 
        size = 3) + stat_ellipse(level = level) + facet_wrap(eval(parse(text = paste("~", 
        group))), ncol = 3) + geom_text(data = dplyr::filter(data, outlier), 
        aes_string(label = id, hjust = "h", vjust = "v"), size = 4) + theme_bw() + 
        scale_colour_discrete(guide = "none")
}

plsPlot <- function(plsComp, genotype, groups = "treatment", type = c("present", 
    "genotype")) {
    type <- match.arg(type)
    treatGrp <- by(dplyr::select(plsComp, t1, t2), plsComp[groups], colMeans)
    grps <- do.call(expand.grid, dimnames(treatGrp))
    grps <- dplyr::filter(grps, !sapply(treatGrp, is.null))
    names(grps) <- groups
    treatAvg <- cbind(do.call(rbind, treatGrp), grps)
    if (length(groups) > 1) 
        treatAvg$sample <- treatAvg[[groups[2]]] else treatAvg$sample <- "avg"
    
    if (!missing(genotype)) {
        if (type == "present") {
            plsComp[[genotype]] <- pmin(plsComp[[genotype]], 1)
        }
        plsComp[[genotype]] <- as.factor(plsComp[[genotype]])
        grpMean <- by(dplyr::select(plsComp, t1, t2), c(plsComp[genotype], plsComp[groups]), 
            colMeans)
        grps <- do.call(expand.grid, dimnames(grpMean))
        grps <- dplyr::filter(grps, !sapply(grpMean, is.null))
        names(grps) <- c(genotype, groups)
        genoAvg <- cbind(do.call(rbind, grpMean), grps)
        if (length(groups) > 1) {
            genoAvg$sample <- paste(genoAvg[[genotype]], genoAvg[[groups[2]]], 
                sep = "_")
        } else {
            genoAvg$sample <- genoAvg[[genotype]]
        }
    }
    
    if (length(groups) > 1) {
        fig <- ggplot(plsComp, aes_string(x = "t1", y = "t2", group = "sample", 
            shape = groups[2]))
    } else {
        fig <- ggplot(plsComp, aes_string(x = "t1", y = "t2", group = "sample"))
    }
    if (missing(genotype)) {
        fig <- fig + geom_point(data = treatAvg, aes_string(colour = groups[1]), 
            size = 5)
    } else {
        fig <- fig + geom_point(data = treatAvg, size = 5, colour = "black")
        fig <- fig + geom_point(data = genoAvg, aes_string(colour = genotype), 
            size = 5)
        fig <- fig + geom_line(data = genoAvg, aes_string(colour = genotype), 
            line.width = 2, arrow = grid::arrow(type = "closed", length = grid::unit(0.15, 
                "inches"), angle = 20))
    }
    fig <- fig + geom_line(data = treatAvg, colour = "black", line.width = 2, 
        arrow = grid::arrow(type = "closed", length = grid::unit(0.15, "inches"), 
            angle = 20))
    if (missing(genotype)) {
        fig <- fig + geom_point(aes_string(colour = groups[1]), size = 3, alpha = 0.6)
        fig <- fig + scale_colour_manual(values = c(Basal = "blue", Stim = "red"))
    } else {
        fig <- fig + geom_point(aes_string(colour = genotype), size = 3, alpha = 0.6)
        fig <- fig + scale_colour_manual(values = c(`0` = "#3E79F7", `1` = "#FA2AA4", 
            `2` = "#FFB72B"))
    }
    fig <- fig + geom_line(colour = "grey", alpha = 0.6, arrow = grid::arrow(type = "closed", 
        length = grid::unit(0.15, "inches"), angle = 20))
    
    fig <- fig + theme_bw() + xlab("Component 1") + ylab("Component 2")
    fig
}

volcanoPlot <- function(limmaTable, fcCut, fdrCut, fcCut2 = log2(2)) {
    fcScale <- max(-log10(limmaTable$adj.P.Val))/max(abs(limmaTable$logFC))
    limmaExtreme <- dplyr::filter(limmaTable, abs(logFC) > fcCut & adj.P.Val < 
        fdrCut)
    limmaExtreme <- mutate(limmaExtreme, distance = (fcScale * logFC)^2 + log10(adj.P.Val)^2)
    limmaExtreme <- limmaExtreme[order(limmaExtreme$distance, decreasing = TRUE), 
        ]
    ggplot(limmaTable, aes(x = logFC, y = -log10(adj.P.Val))) + stat_density2d(data = dplyr::filter(limmaTable, 
        abs(logFC) <= fcCut | adj.P.Val >= fdrCut), geom = "tile", aes(fill = ..density..^0.1), 
        contour = FALSE) + geom_point(data = limmaExtreme, size = 1.5, aes(colour = distance)) + 
        geom_text(data = dplyr::filter(limmaExtreme, distance > (2 * fcScale)^2 & 
            logFC >= fcCut2), aes(label = SymbolReannotated), vjust = 0, hjust = 0, 
            angle = 30) + geom_text(data = dplyr::filter(limmaExtreme, distance > 
        (2 * fcScale)^2 & logFC <= -fcCut2), aes(label = SymbolReannotated), 
        vjust = 0, hjust = 1, angle = 330) + geom_line(data = data.frame(x = c(-Inf, 
        -fcCut), y = c(-log10(fdrCut), -log10(fdrCut))), aes(x = x, y = y), 
        linetype = "dotted") + geom_line(data = data.frame(x = c(fcCut, Inf), 
        y = c(-log10(fdrCut), -log10(fdrCut))), aes(x = x, y = y), linetype = "dotted") + 
        geom_line(data = data.frame(x = c(fcCut, fcCut), y = c(-log10(fdrCut), 
            Inf)), aes(x = x, y = y), linetype = "dotted") + geom_line(data = data.frame(x = c(-fcCut, 
        -fcCut), y = c(-log10(fdrCut), Inf)), aes(x = x, y = y), linetype = "dotted") + 
        scale_fill_gradientn(colours = colorRampPalette(c("white", blues9))(256), 
            breaks = NULL) + scale_colour_gradientn(colours = colorRampPalette(c("yellow", 
        "red"))(256), guide = "none") + theme_bw() + xlab("log2(fold change)") + 
        ylab("-log10(FDR)") + expand_limits(x = c(min(limmaTable$logFC) - 0.25, 
        max(limmaTable$logFC) + 0.25), y = max(-log10(limmaTable$adj.P.Val)) + 
        2)
}
# Multiple plot function ggplot objects can be passed in ..., or to plotlist
# (as a list of ggplot objects) - cols: Number of columns in layout -
# layout: A matrix specifying the layout. If present, 'cols' is ignored.  If
# the layout is something like matrix(c(1,2,3,3), nrow=2, byrow=TRUE), then
# plot 1 will go in the upper left, 2 will go in the upper right, and 3 will
# go all the way across the bottom.  author: Winston Chang ()
multiplot <- function(..., plotlist = NULL, file, cols = 1, layout = NULL) {
    library(grid)
    
    # Make a list from the ... arguments and plotlist
    plots <- c(list(...), plotlist)
    
    numPlots = length(plots)
    
    # If layout is NULL, then use 'cols' to determine layout
    if (is.null(layout)) {
        # Make the panel ncol: Number of columns of plots nrow: Number of rows
        # needed, calculated from # of cols
        layout <- matrix(seq(1, cols * ceiling(numPlots/cols)), ncol = cols, 
            nrow = ceiling(numPlots/cols))
    }
    
    if (numPlots == 1) {
        print(plots[[1]])
        
    } else {
        # Set up the page
        grid.newpage()
        pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
        
        # Make each plot, in the correct location
        for (i in 1:numPlots) {
            # Get the i,j matrix positions of the regions that contain this subplot
            matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
            
            print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row, layout.pos.col = matchidx$col))
        }
    }
}

filterSample <- function(sig, sig.raw, det, excl) {
    sig <- filter(sig, !sample %in% excl)
    sig.raw <- filter(sig.raw, !sample %in% excl)
    det <- filter(det, !sample %in% excl)
    
    list(sig = sig, sig.raw = sig.raw, det = det)
}

updateSampleList <- function(samples, sampleFile) {
    samples <- sub("_.", "", unique(samples))
    sampleInfo <- readr::read_delim(sampleFile, delim = " ", col_names = c("Family", 
        "Individual", "Father", "Mother", "Sex", "Phenotype"))
    keepSamples <- dplyr::filter(dplyr::select(sampleInfo, Family, Individual), 
        Individual %in% samples)
    write.table(keepSamples, file = "tmp/sample.lst", quote = FALSE, col.names = FALSE, 
        row.names = FALSE)
    invisible(keepSamples)
}

formatQTL <- function(me, limmaTable) {
    eQTL <- me$all$eqtls
    eQTL <- dplyr::left_join(eQTL, dplyr::select(limmaTable, NuID, SymbolReannotated, 
        AveExpr, logFC), by = c(gene = "NuID"))
    eQTL <- dplyr::select(eQTL, snps, SymbolReannotated, gene, beta:FDR, AveExpr, 
        logFC)
    eQTL <- dplyr::rename(eQTL, SNP = snps, Gene = SymbolReannotated, NuID = gene)
    eQTL
}

promoterScan <- function(seqs, pwms, group) {
    res <- motifEnrichment(seqs, pwms)
    report <- lapply(1:length(seqs), function(x, i) sequenceReport(x, i)@d, 
        x = res)
    report <- do.call(rbind, report)
    cbind(report, EntrezReannotated = names(seqs), expr = group)
}

plsLimma <- function(data, cluster) {
    design <- model.matrix(~0 + cluster)
    contrasts <- makeContrasts(cluster3 - cluster1, cluster3 - cluster2, cluster1 - 
        cluster2, levels = design)
    clusterFit <- lmFit(data, design)
    clusterFit2 <- contrasts.fit(clusterFit, contrasts)
    clusterFit2 <- eBayes(clusterFit2)
    
    clust3_1Table <- topTable(clusterFit2, coef = 1, n = dim(clusterFit2)[1], 
        adjust = "BH")
    clust3_2Table <- topTable(clusterFit2, coef = 2, n = dim(clusterFit2)[1], 
        adjust = "BH")
    clust1_2Table <- topTable(clusterFit2, coef = 3, n = dim(clusterFit2)[1], 
        adjust = "BH")
    
    clust3_1Table <- cbind(ArrayAddress = rownames(clust3_1Table), clust3_1Table)
    clust3_1Table <- inner_join(dplyr::select(annot, ArrayAddress, NuID, SymbolReannotated, 
        EntrezReannotated, GenomicLocation), clust3_1Table)
    clust3_1Table <- clust3_1Table[order(clust3_1Table$P.Value), ]
    
    clust3_2Table <- cbind(ArrayAddress = rownames(clust3_2Table), clust3_2Table)
    clust3_2Table <- inner_join(dplyr::select(annot, ArrayAddress, NuID, SymbolReannotated, 
        EntrezReannotated, GenomicLocation), clust3_2Table)
    clust3_2Table <- clust3_2Table[order(clust3_2Table$P.Value), ]
    
    clust1_2Table <- cbind(ArrayAddress = rownames(clust1_2Table), clust1_2Table)
    clust1_2Table <- inner_join(dplyr::select(annot, ArrayAddress, NuID, SymbolReannotated, 
        EntrezReannotated, GenomicLocation), clust1_2Table)
    clust1_2Table <- clust1_2Table[order(clust1_2Table$P.Value), ]
    list(clust3_1Table = clust3_1Table, clust3_2Table = clust3_2Table, clust1_2Table = clust1_2Table)
}
prep_topGO <- function(limmaTable, fdrCut, fcCut) {
    groupUp <- ifelse(limmaTable$adj.P.Val < fdrCut & limmaTable$logFC > fcCut, 
        1, 0)
    groupDown <- ifelse(limmaTable$adj.P.Val < fdrCut & limmaTable$logFC < -fcCut, 
        1, 0)
    groupAll <- ifelse(limmaTable$adj.P.Val < fdrCut & abs(limmaTable$logFC) > 
        fcCut, 1, 0)
    
    universeUp <- factor(groupUp)
    names(universeUp) <- limmaTable$EntrezReannotated
    universeUp <- universeUp[!is.na(names(universeUp))]
    universeDown <- factor(groupDown)
    names(universeDown) <- limmaTable$EntrezReannotated
    universeDown <- universeDown[!is.na(names(universeDown))]
    universeAll <- factor(groupAll)
    names(universeAll) <- limmaTable$EntrezReannotated
    universeAll <- universeAll[!is.na(names(universeAll))]
    
    tgUp <- new("topGOdata", ontology = "BP", allGenes = universeUp, nodeSize = 10, 
        description = "Genes up-regulated after heat shock", annotationFun = annFUN.org, 
        mapping = "org.Hs.eg.db")
    tgDown <- new("topGOdata", ontology = "BP", allGenes = universeDown, nodeSize = 10, 
        description = "Genes down-regulated after heat shock", annotationFun = annFUN.org, 
        mapping = "org.Hs.eg.db")
    tgAny <- new("topGOdata", ontology = "BP", allGenes = universeAll, nodeSize = 10, 
        description = "Genes differentially expressed in response to heat shock", 
        annotationFun = annFUN.org, mapping = "org.Hs.eg.db")
    
    list(up = tgUp, down = tgDown, any = tgAny)
}

figRef <- local({
    tag <- numeric()
    created <- logical()
    used <- logical()
    function(label, caption, prefix = options("figcap.prefix"), sep = options("figcap.sep"), 
        prefix.highlight = options("figcap.prefix.highlight")) {
        i <- which(names(tag) == label)
        if (length(i) == 0) {
            i <- length(tag) + 1
            tag <<- c(tag, i)
            names(tag)[length(tag)] <<- label
            used <<- c(used, FALSE)
            names(used)[length(used)] <<- label
            created <<- c(created, FALSE)
            names(created)[length(created)] <<- label
        }
        if (!missing(caption)) {
            created[label] <<- TRUE
            paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
                " ", caption)
        } else {
            used[label] <<- TRUE
            paste(prefix, tag[label])
        }
    }
})

tabRef <- local({
    tag <- numeric()
    created <- logical()
    used <- logical()
    function(label, caption, prefix = options("tabcap.prefix"), sep = options("tabcap.sep"), 
        prefix.highlight = options("tabcap.prefix.highlight")) {
        i <- which(names(tag) == label)
        if (length(i) == 0) {
            i <- length(tag) + 1
            tag <<- c(tag, i)
            names(tag)[length(tag)] <<- label
            used <<- c(used, FALSE)
            names(used)[length(used)] <<- label
            created <<- c(created, FALSE)
            names(created)[length(created)] <<- label
        }
        if (!missing(caption)) {
            created[label] <<- TRUE
            paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
                " ", caption)
        } else {
            used[label] <<- TRUE
            paste(prefix, tag[label])
        }
    }
})</code></pre></div></div><div id="session-info" class="section level2"><h2>Session Info</h2><div class="row"><p>R version 3.2.1 (2015-06-18) Platform: x86_64-pc-linux-gnu (64-bit) Running under: Debian GNU/Linux stretch/sid</p><p>locale: [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C<br /> [3] LC_TIME=en_US.UTF-8 LC_COLLATE=en_US.UTF-8<br /> [5] LC_MONETARY=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8<br /> [7] LC_PAPER=en_US.UTF-8 LC_NAME=C<br /> [9] LC_ADDRESS=C LC_TELEPHONE=C<br />[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C</p><p>attached base packages: [1] grid stats4 parallel methods stats graphics grDevices [8] utils datasets base</p><p>other attached packages: [1] PWMEnrich.Hsapiens.background_4.2.0<br /> [2] PWMEnrich_4.4.0<br /> [3] biomaRt_2.24.0<br /> [4] RISmed_2.1.5<br /> [5] MultiPhen_2.0.1<br /> [6] meta_4.3-0<br /> [7] epitools_0.5-7<br /> [8] abind_1.4-3<br /> [9] plsdepot_0.1.17<br />[10] MatrixEQTL_2.1.1<br />[11] plink2R_1.0<br />[12] RcppEigen_0.3.2.5.0<br />[13] Rcpp_0.12.0<br />[14] TxDb.Hsapiens.UCSC.hg19.knownGene_3.1.2 [15] GenomicFeatures_1.20.1<br />[16] BSgenome.Hsapiens.UCSC.hg19_1.4.0<br />[17] SNPlocs.Hsapiens.dbSNP142.GRCh37_0.99.5 [18] BSgenome_1.36.2<br />[19] rtracklayer_1.28.6<br />[20] Biostrings_2.36.1<br />[21] XVector_0.8.0<br />[22] GenomicRanges_1.20.5<br />[23] topGO_2.20.0<br />[24] SparseM_1.6<br />[25] GO.db_3.1.2<br />[26] graph_1.46.0<br />[27] ggdendro_0.1-15<br />[28] reshape2_1.4.1<br />[29] pander_0.5.2<br />[30] illuminaHumanv3.db_1.26.0<br />[31] org.Hs.eg.db_3.1.2<br />[32] RSQLite_1.0.0<br />[33] DBI_0.3.1<br />[34] AnnotationDbi_1.30.1<br />[35] GenomeInfoDb_1.4.1<br />[36] IRanges_2.2.5<br />[37] S4Vectors_0.6.2<br />[38] stringr_1.0.0<br />[39] readr_0.1.1.9000<br />[40] tidyr_0.2.0<br />[41] dplyr_0.4.2<br />[42] ggplot2_1.0.1<br />[43] RColorBrewer_1.1-2<br />[44] gplots_2.17.0<br />[45] sva_3.14.0<br />[46] genefilter_1.50.0<br />[47] mgcv_1.8-6<br />[48] nlme_3.1-121<br />[49] scatterplot3d_0.3-36<br />[50] limma_3.24.15<br />[51] vsn_3.36.0<br />[52] affy_1.46.1<br />[53] Biobase_2.28.0<br />[54] BiocGenerics_0.14.0<br />[55] gdata_2.17.0<br />[56] proto_0.3-10<br />[57] MASS_7.3-42<br />[58] pushoverr_0.1.4<br />[59] knitrBootstrap_1.0.0<br />[60] knitr_1.11<br />[61] rmarkdown_0.7<br />[62] BiocInstaller_1.18.4</p>loaded via a namespace (and not attached): [1] minqa_1.2.4 colorspace_1.2-6<br /> [3] evd_2.3-0 markdown_0.7.7<br /> [5] futile.logger_1.4.1 mice_2.22<br /> [7] affyio_1.36.0 codetools_0.2-11<br /> [9] splines_3.2.1 R.methodsS3_1.7.0<br />[11] nloptr_1.0.4 Rsamtools_1.20.4<br />[13] seqLogo_1.34.0 pbkrtest_0.4-2<br />[15] annotate_1.46.1 R.oo_1.19.0<br />[17] httr_1.0.0 assertthat_0.1<br />[19] Matrix_1.2-2 lazyeval_0.1.10<br />[21] formatR_1.2 quantreg_5.11<br />[23] htmltools_0.2.6 tools_3.2.1<br />[25] gtable_0.1.2 preprocessCore_1.30.0<br />[27] lme4_1.1-8 gtools_3.5.0<br />[29] XML_3.98-1.3 zlibbioc_1.14.0<br />[31] scales_0.2.5 lambda.r_1.1.7<br />[33] HardyWeinberg_1.5.5 yaml_2.1.13<br />[35] rpart_4.1-10 stringi_0.5-5<br />[37] randomForest_4.6-10 caTools_1.17.1<br />[39] BiocParallel_1.2.9 bitops_1.0-6<br />[41] evaluate_0.7.2 lattice_0.20-31<br />[43] GenomicAlignments_1.4.1 labeling_0.3<br />[45] plyr_1.8.3 magrittr_1.5<br />[47] R6_2.1.0 sp_1.1-1<br />[49] survival_2.38-3 RCurl_1.95-4.7<br />[51] nnet_7.3-10 car_2.0-25<br />[53] futile.options_1.0.0 KernSmooth_2.23-15<br />[55] digest_0.6.8 xtable_1.7-4<br />[57] R.utils_2.1.0 munsell_0.4.2<br /></div><div class="references"><h1>References</h1><p>Alexa, Adrian, and Jorg Rahnenfuhrer. 2010. <em>TopGO: TopGO: Enrichment Analysis for Gene Ontology</em>.</p><p>Barbosa-Morais, Nuno L., Mark J. Dunning, Shamith A. Samarajiwa, Jeremy F. J. Darot, Matthew E. Ritchie, Andy G. Lynch, and Simon Tavaré. 2010. “A Re-Annotation Pipeline for Illumina BeadArrays: Improving the Interpretation of Gene Expression Data.” <em>Nucleic Acids Research</em> 38 (3): e17. doi:<a href="http://dx.doi.org/10.1093/nar/gkp942">10.1093/nar/gkp942</a>.</p><p>Carlson, Marc. 2015. <em>TxDb.Hsapiens.UCSC.hg19.knownGene: Annotation Package for TxDb Object(s)</em>.</p><p>Chang, Christopher, Carson Chow, Laurent Tellier, Shashaank Vattikuti, Shaun Purcell, and James Lee. 2015. “Second-Generation PLINK: Rising to the Challenge of Larger and Richer Datasets.” <em>GigaScience</em> 4 (1): 7. doi:<a href="http://dx.doi.org/10.1186/s13742-015-0047-8">10.1186/s13742-015-0047-8</a>.</p><p>Gibbs, Richard <em>et al.</em> 2010. “The International HapMap Project.” <em>Nature</em> 426: 789–96.</p><p>Huber, Wolfgang, Anja <span>von Heydebreck</span>, Holger Sueltmann, Annemarie Poustka, and Martin Vingron. 2002. “Variance Stabilization Applied to Microarray Data Calibration and to the Quantification of Differential Expression.” <em>Bioinformatics</em> 18 Suppl. 1: S96–S104.</p><p>Johnson, W. Evan, Cheng Li, and Ariel Rabinovic. 2007. “Adjusting Batch Effects in Microarray Expression Data Using Empirical Bayes Methods.” <em>Biostatistics</em> 8 (1): 118–27. doi:<a href="http://dx.doi.org/10.1093/biostatistics/kxj037">10.1093/biostatistics/kxj037</a>.</p><p>Kee, C., K. Y. Cheong, K. Pham, G. W. Waterer, and S. E. L. Temple. 2008. “Genetic Variation in Heat Shock Protein 70 Is Associated with Septic Shock: Narrowing the Association to a Specific Haplotype.” <em>International Journal of Immunogenetics</em> 35 (6). Blackwell Publishing Ltd: 465–73. doi:<a href="http://dx.doi.org/10.1111/j.1744-313X.2008.00812.x">10.1111/j.1744-313X.2008.00812.x</a>.</p><p>Leek, Jeffrey T., W. Evan Johnson, Hilary S. Parker, Elana J. Fertig, Andrew E. Jaffe, and John D. Storey. 2015. <em>Sva: Surrogate Variable Analysis</em>.</p><p>Martin, Annalise M., David Nolan, Silvana Gaudieri, Coral Ann Almeida, Richard Nolan, Ian James, Filipa Carvalho, et al. 2004. “Predisposition to Abacavir Hypersensitivity Conferred by HLA-B*5701 and a Haplotypic Hsp70-Hom Variant.” <em>Proceedings of the National Academy of Sciences</em> 101 (12): 4180–85. doi:<a href="http://dx.doi.org/10.1073/pnas.0307067101">10.1073/pnas.0307067101</a>.</p><p>O’Reilly, Clive J. AND Pomyen, Paul F. AND Hoggart. 2012. “MultiPhen: Joint Model of Multiple Phenotypes Can Increase Discovery in GWAS.” <em>PLoS ONE</em> 7 (5). Public Library of Science: e34861. doi:<a href="http://dx.doi.org/10.1371/journal.pone.0034861">10.1371/journal.pone.0034861</a>.</p><p>Pachkov, Mikhail, Ionas Erb, Nacho Molina, and Erik van Nimwegen. 2007. “SwissRegulon: A Database of Genome-Wide Annotations of Regulatory Sites.” <em>Nucleic Acids Research</em> 35 (suppl 1): D127–31. doi:<a href="http://dx.doi.org/10.1093/nar/gkl857">10.1093/nar/gkl857</a>.</p><p>Pagès, Hervé. 2014. <em>SNPlocs.Hsapiens.dbSNP142.GRCh37: SNP Locations for Homo Sapiens (DbSNP Build 142)</em>.</p><p>Richter, Klaus, Martin Haslbeck, and Johannes Buchner. 2010. “The Heat Shock Response: Life on the Verge of Death.” <em>Molecular Cell</em> 40 (2). Elsevier: 253–66. doi:<a href="http://dx.doi.org/10.1016/j.molcel.2010.10.006">10.1016/j.molcel.2010.10.006</a>.</p><p>Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law, Wei Shi, and Gordon K Smyth. 2015. “limma Powers Differential Expression Analyses for RNA-Sequencing and Microarray Studies.” <em>Nucleic Acids Research</em> 43: doi: 10.1093/nar/gkv007.</p><p>Ritossa, Ferruccio. 1962. “A New Puffing Pattern Induced by Temperature Shock and DNP in Drosophila.” <em>Experientia</em> 18: 571–73.</p><p>Sfar, Sana, Hamadi Saad, Faouzi Mosbah, and Lotfi Chouchane. 2010. “Synergistic Effect and VEGF/HSP70-Hom Haplotype Analysis: Relationship to Prostate Cancer Risk and Clinical Outcome.” <em>Human Immunology</em> 71 (4): 377–82. doi:<a href="http://dx.doi.org/http://dx.doi.org/10.1016/j.humimm.2010.01.017">http://dx.doi.org/10.1016/j.humimm.2010.01.017</a>.</p><p>Stojnic, Robert, and Diego Diez. 2014. <em>PWMEnrich: PWM Enrichment Analysis</em>.</p><p>Vihervaara, Anniina, Christian Sergelius, Jenni Vasara, Malin A. H. Blom, Alexandra N. Elsing, Pia Roos-Mattjus, and Lea Sistonen. 2013. “Transcriptional Response to Stress in the Dynamic Chromatin Environment of Cycling and Mitotic Cells.” <em>Proceedings of the National Academy of Sciences</em> 110 (36): E3388–97. doi:<a href="http://dx.doi.org/10.1073/pnas.1305275110">10.1073/pnas.1305275110</a>.</p></div></div></div>

          <p class="text-muted" id="credit">
            Styled with <a href="https://github.com/jimhester/knitrBootstrap">knitrBootstrap</a>
          </p>

        </div>

                <div  class="col-md-2">
          <div class="hidden-print hidden-xs hidden-sm sidebar affix">
            <div id="toc" class="sidenav well">
              <ul><li><a href="#introduction">Introduction</a></li><li><a href="#methods">Methods</a><ul><li><a href="#pre-processing-and-quality-control">Pre-processing and Quality Control</a><ul><li><a href="#raw-gene-expression-qc">Raw gene expression QC</a></li><li><a href="#genotype-qc">Genotype QC</a></li><li><a href="#normalising-gene-expression-estimates">Normalising gene expression estimates</a></li></ul></li><li><a href="#differential-expression-analysis">Differential expression analysis</a></li><li><a href="#go-enrichment-analysis">GO enrichment analysis</a></li><li><a href="#gene-annotions">Gene annotions</a><ul><li><a href="#gene-ontology-terms">Gene Ontology terms</a></li><li><a href="#pubmed">PubMed</a></li><li><a href="#heat-shock-factor-binding">Heat shock factor binding</a></li></ul></li><li><a href="#global-heat-shock-response">Global heat shock response</a><ul><li><a href="#interpreting-the-pls-components">Interpreting the PLS components</a></li></ul></li></ul></li><li><a href="#results-1">Results</a><ul><li><a href="#differential-expression-analysis-1">Differential expression analysis</a></li><li><a href="#go-enrichment-analysis-1">GO enrichment analysis</a></li><li><a href="#gene-annotations">Gene annotations</a></li><li><a href="#global-heat-shock-response-1">Global heat shock response</a></li></ul></li><li><a href="#appendix">Appendix</a><ul><li><a href="#custom-functions-used">Custom functions used</a></li><li><a href="#session-info">Session Info</a></li></ul></li></ul>
            </div>
          </div>
        </div>
        
      </div>
    </div>

        <!-- dynamically load mathjax for compatibility with self-contained -->
    <script>
      (function () {
          var script = document.createElement("script");
          script.type = "text/javascript";
          script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
          document.getElementsByTagName("head")[0].appendChild(script);
      })();
    </script>
      </body>
</html>
